---
title: "paper figures"
author: "Benedikt Obermayer"
date: "2024/12/19"
output: 
  html_document:
    toc: true
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, cache.lazy=FALSE, message=FALSE, warning=FALSE,dev=c('pdf'))
library(ggpubr)
library(stats)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(caret)
library(scales)
library(cowplot)
library(ComplexHeatmap)
library(circlize)
library(ggrepel)
library(car)
library(glmnet)
library(pROC)
library(readxl)
library(variancePartition)
library(RColorBrewer)
library(gtools)
library(grid)
library(gtools)
library(lme4)
library(dendextend)
library(openxlsx)
```

setup

```{r definitions}
fig_scale <- 1.
columns <- list('Diversity'=c('clusters',
                              'clusters_eff',
                              'cluster_size_mean',
                              'cluster_size_std',
                              'cluster_gini',
                              'cluster_entropy',
                              'cluster_inverse_simpson',
                              'occupancy_top_cluster',
                              'occupancy_big_clusters'),
                'Isotypes'=c('alpha_ratio_clusters',
                             'pct_clusters_SA1','pct_clusters_SA2','pct_clusters_SE',
                             'pct_clusters_SG1','pct_clusters_SG2','pct_clusters_SG3',
                             'pct_clusters_SG4','pct_clusters_SM',
                             'pct_clusters_Sa','pct_clusters_Sg3',
                             'pct_clusters_Sg2b','pct_clusters_Sg2c',
                             'pct_clusters_Sm','pct_clusters_Sg1'),
                'Structural'=c('pct_templated_inserts',
                               'pct_inversions','pct_duplications',
                               'duplication_size','inversion_size'),
                'Context'=c('homology_score_fw','homology_score_rv',
                            'donor_score_S', 'donor_score_W', 
                            'donor_score_WGCW', 'donor_score_GAGCT',
                            'donor_score_GGGST',
                            'donor_complexity',
                            'acceptor_score_S', 'acceptor_score_W', 
                            'acceptor_score_WGCW', 'acceptor_score_GAGCT',
                            'acceptor_score_GGGST',
                            'acceptor_complexity',
                            'homology', 
                            'untemplated_inserts',
                            'pct_blunt'),
                'Breaks'=c('read_length_mean',
                           'read_length_std',
                           'GC_content_mean',
                           'GC_content_std',
                           'read_length_SA_mean',
                           'read_length_SG_mean',
                           'read_length_SA_std',
                           'read_length_SG_std',
                           'pct_direct_switch',
                           'pct_sequential_switch',
                           'pct_intraswitch_deletion',
                           'intraswitch_size_mean',
                           'intraswitch_size_std',
                           'break_dispersion_SM','break_dispersion_SA','break_dispersion_SG',
                           'break_dispersion_SA1','break_dispersion_SA2',
                           'break_dispersion_SG1','break_dispersion_SG2',
                           'break_dispersion_SG2B','break_dispersion_SG2C',
                           'break_dispersion_SG3','break_dispersion_SG4',
                           'pct_breaks_SM_SM','pct_breaks_SM_SG',
                           'pct_breaks_SG_SG','pct_breaks_SM_SA',
                           'pct_breaks_SG_SA','pct_breaks_SA_SA'),
                'Context_isotype'=c('homology_score_fw_SM_SA','homology_score_rv_SM_SA',
                                    'homology_score_fw_SM_SG','homology_score_rv_SM_SG',
                                    'donor_score_GAGCT_SM_SG','donor_score_GGGST_SM_SA',
                                    'donor_score_GGGST_SM_SG','donor_score_S_SM_SA',
                                    'donor_score_S_SM_SG','donor_score_W_SM_SA',
                                    'donor_score_W_SM_SG','donor_score_WGCW_SM_SA',
                                    'donor_score_WGCW_SM_SG','donor_score_GAGCT_SM_SA',
                                    'donor_complexity_SM_SG','donor_complexity_SM_SA',
                                    'acceptor_score_GAGCT_SM_SA','acceptor_score_GAGCT_SM_SG',
                                    'acceptor_score_GGGST_SM_SA','acceptor_score_GGGST_SM_SG',
                                    'acceptor_score_S_SM_SA','acceptor_score_WGCW_SM_SG',  
                                    'acceptor_score_S_SM_SG','acceptor_score_W_SM_SA',
                                    'acceptor_score_W_SM_SG','acceptor_score_WGCW_SM_SA',
                                    'acceptor_complexity_SM_SA',
                                    'acceptor_complexity_SM_SG',
                                    'homology_SA', 
                                    'homology_SG', 
                                    'untemplated_inserts_SA',
                                    'untemplated_inserts_SG',
                                    'pct_blunt_SA',
                                    'pct_blunt_SG'),
                'Diversity_raw'=c('clusters_raw',
                                  'clusters_eff_raw',
                                  'cluster_size_mean_raw',
                                  'cluster_size_std_raw',
                                  'cluster_gini_raw',
                                  'cluster_entropy_raw',
                                  'cluster_inverse_simpson_raw',
                                  'occupancy_top_cluster_raw',
                                  'occupancy_big_clusters_raw'),
                'Variants'=c('somatic_variants',
                             'frac_variants_transitions',
                             'num_variants')) %>%
  unlist() %>%
  data.frame(col=.) %>%
  tibble::rownames_to_column('category') %>%
  dplyr::mutate(category=factor(gsub('[0-9]*$','',category), 
                                levels=c('Isotypes',
                                         'Diversity',
                                         'Diversity_raw',
                                         'Structural',
                                         'Context',
                                         'Context_isotype',
                                         'Breaks',
                                         'Variants')))
cat_cols <- c('Structural'='#FD9E40',
              'Context'='#FED6AC',
              'Context_isotype'='#FED6AC',
              'Breaks'='#1E8AC4',
              'Isotypes'='#AEF58E',
              'Diversity_raw'='#3A7A2C',
              'Diversity'='#3A7A2C',
              'Variants'='lightpink3',
              'Library'='gray40')

genotype_colors <- c('WT'='gray40','Xlf'='#e1c2f0','Shld1'='#b870dc','Shld1_Xlf'='#a447d2',
                     'Trp53bp1 KO'='darksalmon','Brca1 KO'='darkkhaki','Rif1 KO'='skyblue1','Lig4 KO'='mediumpurple2', 
                     'Trp53bp1'='darksalmon','Atm'='#A65628','Lig4'='mediumpurple2', 'Rif1'='skyblue1', 'Brca1'='darkkhaki')

tissue_colors <- c('Splenocytes'='orange2','Lymph Node'='orangered3','CH12'='forestgreen')
var_colors <- c('Batch'='lavenderblush2','Diagnosis'='tan2','Donor'='firebrick4','Sequencing'='slategray1','unexplained'='gray80',
                'Sex'='steelblue2','Age'='lightgoldenrod1','Treatment'='saddlebrown','IV'='navy')

diagnosis_colors <- c('Control'='gray20','HD'='gray20',
                      'Subclass'='lightslategray','CVID'='coral1',
                      'ICOS'='#8DD3C7','Kabuki'='#BEBADA',
                      'IgG4def'='burlywood4','TNFRSF13B'='palegreen4',
                      'Atypical'='steelblue4',
                      'CVID-like'='#BC80BD','AID'='darkgoldenrod3',
                      'ATM'='#A65628','LIG4'='mediumpurple2',
                      'NIPBL'='#FCCDE5','BRCA1'='darkkhaki',
                      'NFKB2'='#FFED6F','PI3K'='#80B1D3')

batch_colors <- setNames(c(RColorBrewer::brewer.pal('Set1',n=9),RColorBrewer::brewer.pal('Set1',n=6)),
                         c('20230728','20230731','20230904','20230908','20231025',
                           '20231030','20240603','20240711','20240725','20220425',
                           '20221118','20221212','20230120','20230213','20240830'))

col_cols <- columns %>% 
  dplyr::mutate(color=cat_cols[as.character(category)]) %>%
  dplyr::pull(color, col)

my_theme <- function() {
  theme_classic(base_size=7) +
  theme(strip.background=element_blank(),
        strip.placement='outside',
        strip.text = element_text(size=7),
        legend.key.size=unit(2.5, 'mm'),
        legend.spacing=unit(1,'mm'),
        legend.text = element_text(size=7),
        axis.text = element_text(size=7),
        axis.title = element_text(size=7))
}
```

# Figure 1 + supplements

plot # clusters vs. # clones

```{r Fig1D, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
# simulation data for Fig. 1
#synthetic.data <- read.csv('data/synthetic_data.csv',header=1, row.names=1)
synthetic.data <- read.xlsx('data/source_data.xlsx',sheet='synthetic data')
ggplot(synthetic.data %>% 
         dplyr::mutate(nreads=factor(nreads, levels=c('10k','20k','50k','100k'))), 
       aes(x=nclones, y=clusters_raw, color=nreads)) + 
  geom_abline(slope=1,intercept=0, size=.5, color='gray80') +
  geom_point(size=.5) + 
  geom_smooth(method='lm', alpha=.2, size=.5, se=FALSE) +
  scale_x_log10() + 
  scale_y_log10() +
  my_theme() +
  scale_color_brewer(palette='Set2') +
  stat_cor(method = "pearson", aes(label=..rr.label..), size=2, show.legend=FALSE) +
  theme(legend.position=c(.85,.35),
        legend.background = element_rect(fill=NA)) +
  labs(x='# Clones', y='# Filtered clusters', color='# Reads')
```

cell switch: plot # clusters vs. # cells

```{r Fig1E, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
# cell number experiment for  Fig. 1
cellswitch.data <- read.xlsx('data/source_data.xlsx',sheet='cell numbers')

ggplot(cellswitch.data %>%
         dplyr::mutate(Donor=plyr::revalue(Donor,
                                           c('21084'='A',
                                             '21085'='B',
                                             '21086'='C'))),
       aes(x=ncells, y=clusters_raw)) + 
  geom_abline(slope=1,intercept=0, size=.5, color='gray80') +
  geom_point(aes(color=Donor),size=.5) + 
  geom_smooth(size=.5, method='lm', alpha=.2, color='black') +
  stat_cor(method = "pearson", aes(label=..rr.label..), size=2, show.legend=FALSE) +
  scale_x_log10() + 
  scale_y_log10() +
  my_theme() +
  scale_color_brewer(palette='Set1') +
  theme(legend.position=c(.8,.25),
        legend.background = element_rect(fill=NA)) +
  labs(x='# Input cells', y='# Filtered clusters', color='Donor')
```

check reproducibility of clones in the cell switch experiment

```{r Fig1F, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
meta_clustering <- read.xlsx('data/source_data.xlsx', sheet='meta-clustering')

ggplot(meta_clustering %>% 
         dplyr::mutate(ncells=factor(ncells, levels=c('50k','100k')),
                       Donor=ifelse(Donor=='mixed','mixed',paste0(Donor,'-',Donor)),
                       mixing=gsub('-donor','',mixing)), 
       aes(x=ncells,y=100*frac_clusters, fill=mixing)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA) +
  geom_jitter(aes(color=Donor, group=mixing), position=position_jitterdodge(dodge.width = .8, jitter.width=.05), size=1) +
  my_theme() +
  scale_fill_manual(values=c('gray','white'), guide=FALSE) +
  scale_color_manual(values=setNames(c(brewer.pal(3,'Set1'),'gray20'),c('A-A','B-B','C-C','mixed')),na.value='gray20') +
  labs(x='# cells',y='% shared clusters in pair', fill='pair', color='Donors')
```

show VDJ results: isotype frequencies

```{r Fig1G, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
# bulk VDJ results for Fig. 1
VDJ_results <- read.xlsx('data/source_data.xlsx',sheet='VDJ data')

get_gini <- function(x) {
  # calculate Gini coefficient of array x; see https://stackoverflow.com/questions/48999542/more-efficient-weighted-gini-coefficient-in-python
  sorted_x <- sort(x)
  n <- length(x)
  cumx <- cumsum(sorted_x)
  (n + 1 - 2 * sum(cumx) / cumx[n]) / n
}

get_inverse_simpson <- function(x) {
  rel_x <- x/sum(x)
  sum(1 / rel_x**2)
}

vdj_stats <- VDJ_results %>%
  dplyr::mutate(clone=paste0(gsub('\\([0-9]*\\)$','',allVHitsWithScore,'_',str_length(aaSeqCDR3)))) %>%
  dplyr::group_by(Donor,ncells,Sample) %>%
  dplyr::summarise(clusters_raw=n_distinct(clone),
                   cluster_gini_raw=get_gini(uniqueMoleculeCount),
                   cluster_inverse_simpson_raw=get_inverse_simpson(uniqueMoleculeCount),
                   alpha_ratio_clusters=mean(grepl('IGHA',allCHitsWithScore))) %>%
  dplyr::mutate(method='VDJ')

swibrid_stats <- cellswitch.data %>%
  dplyr::select(Donor,ncells,
                clusters_raw,
                cluster_gini_raw,
                cluster_inverse_simpson_raw,
                alpha_ratio_clusters) %>%
  dplyr::filter(ncells %in% c(50000,100000)) %>%
  dplyr::mutate(method='SWIBRID')

vdj_cols <- intersect(colnames(swibrid_stats),colnames(vdj_stats))

vdj_df <- rbind(swibrid_stats[,vdj_cols],vdj_stats[,vdj_cols]) %>%
  gather(metric,value, c(clusters_raw,alpha_ratio_clusters,cluster_gini_raw,cluster_inverse_simpson_raw)) %>%
  dplyr::mutate(ncells=factor(as.character(as.integer(ncells)), levels=c('50000','100000')))

vdj_df %>% 
  dplyr::mutate(ncells=factor(gsub('000$','k',ncells), levels=c('50k','100k')),
                Donor=gsub('Donor ','',Donor)) %>%
  dplyr::filter(metric=='alpha_ratio_clusters') %>%
  dplyr::group_by(ncells, Donor, method) %>%
  dplyr::summarise(mean_alpha_ratio=mean(value),
                   se_alpha_ratio=sd(value)/sqrt(length(value))) %>%
  pivot_wider(id_cols=c(ncells,Donor),names_from=method,values_from=c('mean_alpha_ratio','se_alpha_ratio')) %>%
  ggplot(aes(x=mean_alpha_ratio_VDJ, y=mean_alpha_ratio_SWIBRID,
             xmin=mean_alpha_ratio_VDJ-se_alpha_ratio_VDJ,
             xmax=mean_alpha_ratio_VDJ+se_alpha_ratio_VDJ,
             ymin=mean_alpha_ratio_SWIBRID-se_alpha_ratio_SWIBRID,
             ymax=mean_alpha_ratio_SWIBRID+se_alpha_ratio_SWIBRID)) +
  geom_point(aes(color=Donor,shape=ncells),size=1) +
  geom_errorbar(linewidth=.25) +
  geom_errorbarh(linewidth=.25) +
  scale_x_continuous(breaks=c(.6,.65)) +
  my_theme() +
  scale_color_brewer(palette='Set1') +
  theme(strip.background=element_blank()) +
  labs(y='Fraction IGHA SWIBRID',x='Fraction IGHA VDJ',shape='# cells')
```

# Figure 2 + supplements 

fraction of explained variance for different feature groups in HD cohort

```{r Fig2B,fig.width=7 * fig_scale,fig.height=1.4 * fig_scale}
my_fitExtractVarPartModel <- function(formula, data) {
  # replace fitExtractVarPartModel from variancePartition package by simple function to run over single features one at a time
  # extract relevant parts of code from calcVarPart.R
  # https://github.com/GabrielHoffman/variancePartition/blob/b578069c0e5dab6234516b12e624e3ce995f850d/R/calcVarPart.R#L292-L363
  
  vars <- c(all.vars(formula),'Residuals')
  res <- setNames(rep(NA, length(vars)-1),sort(vars[2:length(vars)]))
  
  try({
    fit <- lmer(formula, data, REML=FALSE)
    w <- weights(fit)
    if (is.null(w)) {
      w <- rep(1, nrow(fit$model))
    }
    varComp <- lapply(lme4::VarCorr(fit), function(fit) attr(fit, "stddev")^2)
    varComp <- varComp[order(names(varComp))]
    idx <- which(colnames(fit@pp$X) != "(Intercept)")
    if (length(idx) > 0) {
      fxeff <- sapply(idx, function(i) {
        fit@pp$X[, i] * lme4::fixef(fit)[i]
      })
      colnames(fxeff) <- colnames(fit@pp$X)[idx]
      N <- nrow(fxeff)
      fixedVar <- apply(fxeff, 2, function(x) {
        weighted.var(x, w) * (N - 1) / N
      })
      varFixedTotal <- weighted.var(rowSums(fxeff), w) * (N - 1) / N
      for (key in names(fixedVar)) {
        varComp[[key]] <- as.array(fixedVar[[key]] / sum(fixedVar) *
                                     varFixedTotal)
      }
    }
    varComp$Residuals <- sigma(fit)^2
    varComp <- unlist(varComp)
    names(varComp) <- gsub("\\.\\(Intercept\\)", "", names(varComp))
    res <- varComp/sum(varComp)
  })
  res
}

# HD data for Fig. 2
HD.data <- read.xlsx('data/source_data.xlsx', sheet='HD data')

HD.scaled <- HD.data %>%
  dplyr::filter(Cohort=='C1') %>%
  dplyr::group_by(feature) %>%
  dplyr::mutate(value=as.vector(scale(value))) %>%
  dplyr::filter(!is.na(value)) %>%
  dplyr::ungroup()

HD.var <- sapply(unique(HD.scaled$feature), 
                 function(ftr)
                   tryCatch(my_fitExtractVarPartModel(value ~ (1 | Donor) + (1 | Sex) + (1 | Age) + (1 | Sequencing) + (1 | Batch),
                                                      HD.scaled %>%
                                                        dplyr::filter(feature==ftr)),
                            error=setNames(rep(NA,5), c('Donor','Sex','Age','Sequencing','Batch')))) %>%
                   t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  gather(variable,variance,-col)

cols.here <- c('cluster_entropy',
               'cluster_gini',
               'cluster_size_mean',
               'cluster_size_std',
               'occupancy_top_cluster',
               'alpha_ratio_clusters',
               'pct_clusters_SA1',
               'pct_clusters_SA2',
               'pct_clusters_SG1',
               'pct_clusters_SG2',
               'pct_templated_inserts',
               'pct_inversions',
               'pct_duplications',
               'duplication_size',
               'inversion_size',
               'homology_score_fw',
               'donor_score_WGCW',
               'acceptor_complexity',
               'homology',
               'untemplated_inserts',
               'pct_direct_switch',
               'pct_sequential_switch',
               'pct_intraswitch_deletion',
               'break_dispersion_SM',
               'break_dispersion_SA')

HD.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(gsub('Residuals','unexplained',variable),
                                levels=rev(c('Donor','Sex','Age','Batch','Sequencing','unexplained'))),
                category=factor(category, levels=c('Structural','Context','Breaks','Diversity','Isotypes'))) %>%
  dplyr::filter(col %in% cols.here) %>%
  dplyr::mutate(col=factor(col, levels=cols.here)) %>%
  ggplot(aes(x=col, y=100*variance, fill=variable)) + 
  geom_col(width=.8) +
  my_theme() +
  facet_wrap(~category,scales='free_y', ncol=5) +
  theme(strip.clip='off',
        legend.position='bottom',
        legend.box='horizontal') +
  guides(fill=guide_legend(ncol=6)) +
  coord_flip() +
  scale_y_continuous(breaks=c(0,50,100)) +
  scale_fill_manual(values=var_colors,
                    breaks=c('Donor','Sex','Age','Batch','Sequencing','unexplained')) +
  labs(x='', y='% Explained variance', fill='')
```

```{r Fig2C, fig.width=1.5 * fig_scale,fig.height=1.75 * fig_scale}
donor_cutoff <- .25
batch_cutoff <- .35
sequencing_cutoff <- .2
selected.cols <- HD.var %>%
  dplyr::left_join(columns, by='col') %>%
  spread(variable, variance) %>%
  dplyr::filter(!(category %in% c('Diversity_raw','Context_isotype','library')), 
                !(col %in% c('log10_nreads','mean_GC')), 
                Donor >= donor_cutoff, 
                Sequencing < sequencing_cutoff, 
                Batch < batch_cutoff) %>%
  dplyr::arrange(category) %>%
  dplyr::pull(col)

HD.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(gsub('Residuals','unexplained',variable),
                                levels=c('Sequencing','Batch','Donor','Sex','Age'))) %>%
  dplyr::filter(variable!='unexplained', category %in% c('Isotypes','Diversity','Structural','Context','Breaks','Variants')) %>%
  ggplot(aes(x=variable, y=100*variance)) + 
  geom_boxplot(aes(fill=variable), outlier.shape=NA, lwd=.5, legend=FALSE) + 
  geom_jitter(aes(color=ifelse(col %in% selected.cols, 'selected','unused')), position=position_jitter(width=.25), size=.25) +
  annotate("rect", xmin = .5, xmax = 1.5, ymax = 100*sequencing_cutoff, ymin = 0, fill='orangered', linetype=1, size=.25, alpha=.25) +
  annotate("rect", xmin = 1.5, xmax = 2.5, ymax = 100*batch_cutoff, ymin = 0, fill='orangered', linetype=1, size=.25, alpha=.25) +
  annotate("rect", xmin = 2.5, xmax = 3.5, ymax = Inf, ymin = 100*donor_cutoff, fill='orangered', linetype=1, size=.25, alpha=.25) +
  annotate('text',.6,78,label='unused',color='black',size=2, hjust=0) +
  annotate('text',.6,73,label='selected',color='orangered',size=2, hjust=0) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='none',
        legend.box='vertical') +
  scale_fill_manual(values=var_colors, guide='none') +
  scale_color_manual(values=c('unused'='black','selected'='orangered')) +
  labs(x='', y='% Explained variance', fill='', color='')
```

we now have `r length(selected.cols)` robust features

```{r FigS2F,fig.width=3.5 * fig_scale,fig.height=1.7 * fig_scale}
HD.var %>%  
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable, 
                                levels=c('Donor','Sex','Age','Batch','Sequencing','Residuals'))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('Isotypes','Diversity','Structural','Context','Breaks','Variants')) %>%
  ggplot(aes(x=category, y=100*variance, fill=variable)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.25) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='top',
        legend.box='horizontal') +
  scale_fill_manual(values=var_colors) +
  labs(x='', y='% Explained variance', fill='')
```

confirm that downsampled diversity measures are better

```{r FigS2I_1, fig.width=1.4 * fig_scale, fig.height=1.7 * fig_scale}
tmp <- HD.var %>%  
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable,
                                levels=rev(c('Donor','Batch','Sequencing','Residuals')))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('Diversity','Diversity_raw')) 

pvals <- tmp %>%
  dplyr::group_by(variable) %>%
  dplyr::do(wc=wilcox.test(variance ~ category, data=.)) %>%
  dplyr::summarise(variable, p=wc$p.value)

ggplot(tmp,aes(x=variable, y=100*variance, fill=category)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.5) +
  geom_text(data=pvals, aes(x=variable, label=stars.pval(p)), y=Inf, inherit.aes=FALSE, size=5) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='top',
        legend.box='horizontal') +
  guides(fill=guide_legend(title.position='top')) +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=c('Diversity'='blue3','Diversity_raw'='lightblue1'),
                    labels=c('downsampled','raw')) +
  labs(x='', y='% Explained variance', fill='Diversity measures')
```

confirm that using isotype-specific context features doesn't help

```{r FigS2I_2, fig.width=1.4 * fig_scale, fig.height=1.7 * fig_scale}
tmp <- HD.var %>%  
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable,
                                levels=rev(c('Donor','Batch','Sequencing','Residuals')))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('Context','Context_isotype'))

pvals <- tmp %>% 
  dplyr::group_by(variable) %>%
  dplyr::do(wc=wilcox.test(variance ~ category, data=., exact=FALSE)) %>%
  dplyr::summarise(variable, p=wc$p.value)

ggplot(tmp, aes(x=variable, y=100*variance, fill=category)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.5) +
  geom_text(data=pvals, aes(x=variable, label=stars.pval(p)), y=Inf, inherit.aes=FALSE, size=5) +
  my_theme() +
  guides(fill=guide_legend(title.position='top')) +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='top',
        legend.box='horizontal') +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=c('Context'='darkorchid3','Context_isotype'='orchid1'),
                    labels=c('global','per isotype')) +
  labs(x='', y='% Explained variance', fill='Context measures')
```

test the effects of averaging over clusters or reads

```{r FigS2I_3, fig.width=1.4 * fig_scale,fig.height=1.7 * fig_scale}
HD.data.br <- read.xlsx('data/source_data.xlsx', sheet='HD data by reads')

HD.scaled.br <- HD.data.br %>%
  dplyr::filter(Cohort=='C1') %>%
  dplyr::group_by(feature) %>%
  dplyr::mutate(value=as.vector(scale(value))) %>%
  dplyr::filter(!is.na(value)) %>%
  dplyr::ungroup()

HD.var.br <- sapply(unique(HD.scaled.br$feature), 
                    function(ftr)
                      my_fitExtractVarPartModel(value ~ (1 | Donor) + (1 | Sex) + (1 | Age) + (1 | Sequencing) + (1 | Batch),
                                                HD.scaled.br %>%
                                                  dplyr::filter(feature==ftr))) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  gather(variable,variance,-col)

fts <- intersect(HD.var$col, HD.var.br$col)

tmp <- rbind(HD.var %>% 
               dplyr::mutate(averaging='Clusters'), 
             HD.var.br %>%
               dplyr::mutate(averaging='Reads')) %>% 
  dplyr::filter(col %in% fts) %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable, 
                                levels=rev(c('Donor','Sex','Age','Batch','Sequencing','Residuals')))) %>%
  dplyr::filter(variable!='Residuals', 
                category %in% c("Context","Breaks"),
                variable %in% c("Donor", "Batch", "Sequencing")) %>% 
  dplyr::arrange(col,averaging,variable) 

pvals <- tmp %>%
  dplyr::group_by(variable) %>%
  dplyr::do(wc=wilcox.test(variance ~ averaging, paired=TRUE, data=., exact=FALSE)) %>%
  dplyr::summarise(variable, p=wc$p.value)

ggplot(tmp, aes(x=variable, y=100*variance, fill=averaging)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.1,dodge.width=.8), size=.5) +
  geom_text(data=pvals, aes(x=variable, label=stars.pval(p)), y=Inf, inherit.aes=FALSE, size=5) +
  my_theme() +
  coord_cartesian(clip='off') +
  guides(fill=guide_legend(title.position='top')) +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='top',
        legend.box='horizontal') +
  scale_fill_manual(values=c('Clusters'='aquamarine4','Reads'='darkseagreen1')) +
  labs(x='', y='% Explained variance', fill='Averages over')
```

show top examples for features related to age or sex

```{r Fig2E, fig.width=.75 * fig_scale, fig.height=1.5 * fig_scale}
col <- 'break_dispersion_SG1'
HD.data %>%
  dplyr::filter(feature == col, Cohort=='C1') %>%
  ggplot(aes(x=str_to_title(Sex), y=value)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_jitter(width=.2, size=.1) + 
  stat_compare_means(comparisons=list(c('Male','Female')),
                     method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2.5) +
  my_theme() + 
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  labs(x='',y=col)

col <- 'occupancy_top_cluster'
HD.data %>%
  dplyr::filter(feature==col, Cohort=='C1') %>%
  ggplot(aes(x=Age, y=value)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_jitter(width=.2, size=.1) + 
  stat_compare_means(comparisons=list(c('<50','>=50')),
                     method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2.5) +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  labs(x='',y=col)
```

HD feature correlation (on FB data)

```{r FigS2H, fig.width=4.25 * fig_scale, fig.height=4.25 * fig_scale}
library(dendextend)
FB.X <- HD.data %>%
  dplyr::filter(Cohort=='C2') %>%
  dplyr::select(Sample, feature, value) %>%
  spread(feature, value) %>%
  dplyr::mutate_if(is.numeric, scale) %>%
  tibble::column_to_rownames('Sample')

R <- cor(FB.X[,selected.cols], use='na.or.complete')
io <- rownames(R)[order.hclust(hclust(dist(R)))]

tmp <- as.data.frame(R) %>% 
  tibble::rownames_to_column('var2') %>%
  gather(var1, R, -var2) %>%
  dplyr::left_join(columns, by=c('var1'='col')) %>%
  dplyr::mutate(cat1=category) %>%
  dplyr::select(-category) %>%
  dplyr::left_join(columns, by=c('var2'='col')) %>%
  dplyr::mutate(cat2=category) %>%
  dplyr::select(-category)  %>%
  dplyr::mutate(var1=factor(var1, levels=io),
                var2=factor(var2, levels=io),
                i1=as.vector(sapply(var1, function(x) which(io==x))),
                i2=as.vector(sapply(var2, function(x) which(io==x)))) 

anno_grob_y <- grid::rectGrob(x=1:length(io), y=0, gp=gpar(col=NA, fill=col_cols[io], alpha=.5))
anno_grob_x <- grid::rectGrob(y=1:length(io), x=0, gp=gpar(col=NA, fill=col_cols[io], alpha=.5))

ggplot(tmp %>%
  dplyr::filter(i1 <= i2),
  aes(x=var1,y=var2,fill=R)) + 
  geom_tile() +
  scale_fill_gradient2(low='blue3',high='red3',mid='gray',limits=c(-1,1)) + 
  my_theme() +
  scale_x_discrete(position = "top") +
  coord_cartesian(clip='off') +
  theme(axis.text.x.top=element_text(angle=90, hjust=0., vjust=.5), #, colour=col_cols[io]),
        legend.position=c(-.1,1.2)) +
  labs(x='',y='') +
  annotation_custom(grob=anno_grob_y, xmin = 0, xmax = 1, ymin = length(io) + 10, ymax = length(io) + 28) +
  annotation_custom(grob=anno_grob_x, xmin = -8, xmax = 9, ymin = 0, ymax = 1)

```

distributions of features in the two HD cohorts

```{r Fig2D, fig.width=4.25 * fig_scale, fig.height=1.5 * fig_scale}
cols.here.human <- c('pct_templated_inserts',
                     'untemplated_inserts',
                     'pct_sequential_switch',
                     'pct_intraswitch_deletion',
                     'cluster_gini',
                     'pct_clusters_SG1')

cohort.data <- HD.data %>%
  dplyr::filter(Sequencing=='minION') %>% 
  dplyr::filter(feature %in% cols.here.human) %>%
  dplyr::left_join(columns, by=c('feature'='col')) %>%
  dplyr::mutate(feature=factor(feature, levels=cols.here.human)) %>%
  dplyr::filter(category %in% c('Diversity','Isotypes','Breaks','Context','Structural')) 
  
ggplot(cohort.data, aes(x=Cohort, y=value)) + 
  geom_boxplot(aes(group=Cohort), position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25, fill='gray80') + 
  geom_pointrange(data=cohort.data %>%
                    dplyr::group_by(Donor,feature,category,Cohort) %>%
                    dplyr::summarise(se=sd(value)/sqrt(n()),
                                     value=mean(value),
                                     n=n()) %>% dplyr::filter(n > 1),
                  aes(fill=Cohort, ymin=value-se,ymax=value+se, group=Cohort),
                  position=position_jitterdodge(jitter.width=.5, dodge.width=.75, jitter.height = 0), 
                  color='black',size=1,stroke=0.1,
                  linewidth=.1,fatten=1) + 
  my_theme() +
  coord_cartesian(clip='off') +
  facet_wrap(~feature, switch='y', scales='free_y', ncol=7) +
  theme(legend.position='none',
        axis.text.x=element_text(angle=90, hjust=1, vjust=.5)) +
  labs(x='',y='')
```

same thing for mouse tissue and Vincendeau

```{r Fig2F, fig.width=4. * fig_scale, fig.height=1.75 * fig_scale}
mouse.data <- read.xlsx('data/source_data.xlsx', sheet='mouse tissues') %>%
  tibble::column_to_rownames('Sample')
vincendeau.data <- read.xlsx('data/source_data.xlsx', sheet='Vincendeau data') %>%
  tibble::column_to_rownames('Sample')

cols.here.mouse <- c('pct_inversions',
                     'pct_intraswitch_deletion',
                     'cluster_size_mean')

cohort.data <- rbind(mouse.data %>%
                       dplyr::mutate(species='mouse tissue',
                                     Cohort=material) %>%
                       dplyr::select(species, Cohort, cols.here.mouse),
                     vincendeau.data %>%
                       dplyr::mutate(species='Vincendeau',
                                     Cohort=Genotype) %>%
                       dplyr::select(species, Cohort, cols.here.mouse)) %>%
  dplyr::mutate(Cohort=factor(Cohort, levels=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'))) %>%
  gather(variable, value, cols.here.mouse) %>%
  dplyr::select(variable, value, species, Cohort) %>%
  dplyr::left_join(columns, by=c('variable'='col')) %>%
  dplyr::mutate(variable=factor(variable, levels=cols.here.mouse)) %>%
  dplyr::filter(category %in% c('Diversity','Isotypes','Breaks','Context','Structural'))

cohort.data %>%
  ggplot(aes(x=Cohort, y=value, fill=species)) + 
  geom_boxplot(position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25) + 
  geom_point(position=position_jitterdodge(jitter.width=.2, dodge.width=.75, jitter.height = 0), size=.5) + 
  my_theme() +
  coord_cartesian(clip='off') +
  facet_wrap(~variable, ncol=3, switch='y', scales='free_y') +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position=c(.2,.99)) +
  scale_fill_manual(values=c('mouse tissue'='gray80','Vincendeau'='#cca9e6')) +
  scale_x_discrete(breaks=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
                   labels=c('Lymph node','Splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
                   limits=c('lymph node','splenocytes','','WT','Xlf','Shld1','Shld1_Xlf'), drop=FALSE) +
  labs(x='',y='', fill='') 
```

check length dependence of Vincendeau data

```{r FigS2J, fig.width=1.5 * fig_scale, fig.height=1.75 * fig_scale}
cols.here.mouse <- c('read_length_SG_mean')
cohort.data <- rbind(mouse.data %>%
                       dplyr::mutate(species='mouse tissue',
                                     Cohort=material) %>%
                       dplyr::select(species, Cohort, cols.here.mouse),
                     vincendeau.data %>%
                       dplyr::mutate(species='Vincendeau',
                                     Cohort=Genotype) %>%
                       dplyr::select(species, Cohort, cols.here.mouse)) %>%
  dplyr::mutate(Cohort=factor(Cohort, levels=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'))) %>%
  gather(variable, value, cols.here.mouse) %>%
  dplyr::select(variable, value, species, Cohort) %>%
  dplyr::left_join(columns, by=c('variable'='col')) %>%
  dplyr::filter(category %in% c('Diversity','Isotypes','Breaks','Context','Structural'))

cohort.data %>%
  ggplot(aes(x=Cohort, y=value, fill=species)) + 
  geom_boxplot(position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25) + 
  geom_point(position=position_jitterdodge(jitter.width=.2, dodge.width=.75, jitter.height = 0), size=.5) + 
  my_theme() +
  coord_cartesian(clip='off') +
  facet_wrap(~variable, ncol=3, switch='y', scales='free_y') +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position=c(.2,.99)) +
  scale_fill_manual(values=c('mouse tissue'='gray80','Vincendeau'='#cca9e6')) +
  scale_x_discrete(breaks=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
                   labels=c('Lymph node','Splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
                   limits=c('lymph node','splenocytes','','WT','Xlf','Shld1','Shld1_Xlf'), drop=FALSE) +
  labs(x='',y='', fill='')
```

compare minION against pacBio in HD cohort for selected features

```{r FigS2G, fig.width=2.75 * fig_scale, fig.height=5 * fig_scale}
tmp <- HD.scaled %>%
  dplyr::filter(feature %in% selected.cols, Batch=='20230313') %>%
  dplyr::select(feature,value,Sequencing,Donor) %>%
  spread(Sequencing,value) %>% 
  dplyr::left_join(columns,by=c('feature'='col')) %>% 
  dplyr::group_by(feature) %>%
  dplyr::mutate(R=cor(minION,pacBio)) %>%
  dplyr::mutate(feature=factor(feature, levels=columns %>% 
                                 dplyr::filter(col %in% selected.cols) %>% 
                                 dplyr::pull(col))) %>%
  dplyr::arrange(feature)

cols_shapes <- tmp %>% 
  dplyr::distinct(category, feature) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(feature=factor(feature, levels=levels(tmp$feature))) %>% 
  dplyr::arrange(feature)

ggplot(tmp,aes(x=minION,y=pacBio)) +
  geom_abline(slope=1,intercept=0,color='gray',linewidth=.5) +
  geom_point(aes(color=feature,shape=feature),size=1) + 
  my_theme() +
  stat_cor(aes(label = ..r.label..), size=2.5) +
  labs(color='',shape='') +
  scale_color_manual(values=cols_shapes %>%
                       dplyr::pull(var_col,feature),
                     labels=cols_shapes %>%
                       dplyr::pull(feature)) +
  scale_shape_manual(values=cols_shapes %>%
                       dplyr::pull(var_ind,feature),
                     labels=cols_shapes %>%
                       dplyr::pull(feature)) +
  guides(color=guide_legend(ncol=2),
         shape=guide_legend(ncol=2))  +
  theme(legend.position='bottom',
        legend.key.size=unit(1.5,'mm'),
        legend.key.spacing.y=unit(0, 'mm'))
```

show a PCA of the Vincendeau data

```{r Fig2G, fig.width=2.5 * fig_scale, fig.height=1.7 * fig_scale}
selected.cols.mouse <- gsub('clusters_SG1','clusters_Sg2b',
                            gsub('clusters_SG2','clusters_Sg2c',
                                 gsub('clusters_SG3','clusters_Sg3',
                                      gsub('clusters_SA1','clusters_Sa',selected.cols))))
vincendeau.cols <- intersect(colnames(vincendeau.data),selected.cols.mouse)
vincendeau.X <- vincendeau.data[,vincendeau.cols] %>%
  dplyr::mutate_if(is.numeric, scale) 
vincendeau.X <- vincendeau.X[,colSums(is.na(vincendeau.X))==0]
vincendeau.meta <- vincendeau.data %>%
  dplyr::mutate(Genotype=factor(Genotype,levels=c('WT','Xlf','Shld1','Shld1_Xlf'))) %>%
  dplyr::select(Genotype)

pca <- prcomp(vincendeau.X)
pcadat <- cbind(pca$x,vincendeau.meta)
pcarot <- data.frame(coef=pca$rot[,'PC1'],
                     col=row.names(pca$rot))  %>%
  dplyr::arrange(coef) %>%
  dplyr::mutate(col=factor(col, levels=col)) 

ggplot(pcadat,aes(x=PC1,y=PC2, color=Genotype)) +
  geom_point(size=1) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(Genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=Genotype)) +
  my_theme() +
  theme(aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position=c(.15,.9),
        legend.background = element_blank()) +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='')  +
  scale_color_manual(values=genotype_colors) +
  scale_fill_manual(values=genotype_colors)
```

(this PCA uses `r dim(vincendeau.X)[2]` features)

show mean scaled values of Vincendeau in a scatter plot (big dots are significant, small ones are not)

```{r Fig2H, fig.width=7 * fig_scale, fig.height=2.25 * fig_scale}
tmp <- vincendeau.data %>%
  gather(variable, value, vincendeau.cols) %>%
  dplyr::group_by(variable) %>%
  dplyr::mutate(value=scale(value)) %>%
  dplyr::group_by(Genotype, variable) %>%
  dplyr::summarise(mean=mean(value),
                   se=sd(value)/sqrt(n())) %>%
  tidyr::pivot_wider(id_cols=c('variable'),names_from='Genotype',values_from=c('mean','se')) %>%
  dplyr::left_join(columns,by=c('variable'='col')) %>%
  dplyr::group_by(variable) %>%
  dplyr::mutate(variable=factor(variable, levels=columns %>% 
                                  dplyr::filter(col %in% vincendeau.cols) %>% 
                                  dplyr::pull(col)),
                var_col=cat_cols[category]) %>%
  dplyr::arrange(variable)

cols_shapes <- tmp %>% 
  dplyr::distinct(category, variable) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(variable=factor(variable, levels=levels(tmp$variable))) %>% 
  dplyr::arrange(variable)

comparisons <- list(c('WT','Xlf'),
                    c('WT','Shld1'),
                    c('WT','Shld1_Xlf'),
                    c('Shld1','Xlf'))

plots <- lapply(comparisons, function(x) 
  tmp %>%
    dplyr::mutate_(mean1=paste0('mean_',x[[1]]),
                   mean2=paste0('mean_',x[[2]]),
                   se1=paste0('se_',x[[1]]),
                   se2=paste0('se_',x[[2]])) %>%
    dplyr::mutate(tstat=abs(mean2-mean1)/(sqrt(se2**2+se1**2)),
                  pval=2*pt(-abs(tstat),4)) %>%
    dplyr::mutate(padj=p.adjust(pval, method='BH'),
                  sig=ifelse(padj < .05, 'sig','ns')) %>%
    ggplot(aes(x=mean1,y=mean2)) +
    geom_abline(slope=1,intercept=0,color='gray',linewidth=.5) +
    geom_point(aes(color=variable, shape=variable, size=sig, alpha=sig)) + 
    stat_cor(method = "pearson", aes(label=..r.label..), size=2.5, show.legend=FALSE) +
    geom_errorbar(aes(ymin=mean2-se2,
                      ymax=mean2+se2,
                      color=variable,
                      alpha=sig),
                  size=.25) +
    geom_errorbarh(aes(xmin=mean1-se1,
                       xmax=mean1+se1,
                       color=variable,
                       alpha=sig),
                   size=.25) +
    my_theme() +
    labs(color='',shape='') +
    scale_color_manual(values=cols_shapes %>% 
                         dplyr::pull(var_col,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable)) +
    scale_shape_manual(values=cols_shapes %>%
                         dplyr::pull(var_ind,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable)) +
    scale_size_manual(values=c('sig'=2,'ns'=.5),guide='none') +
    scale_alpha_manual(values=c('sig'=1,'ns'=.25),guide='none') +
    guides(color=guide_legend(ncol=5),
           shape=guide_legend(ncol=5)) +
    labs(x=x[[1]],y=x[[2]],fill='',color='',size=''))

plot_grid(plot_grid(plotlist=lapply(plots, function(x) x + theme(legend.position='none')), 
                    ncol=4, align='vh'),
          get_legend(plots[[1]] + theme(legend.margin=margin(0,0,0,0,'mm'))), ncol=1, rel_heights=c(.65,.35))
```
# Figure 3 + supplement (CH12)

```{r FigS3A, fig.width=5.5 * fig_scale, fig.height=4.5 * fig_scale}
HTGTS.data <- read.xlsx("data/source_data.xlsx", sheet='HTGTS data') %>%
  tibble::column_to_rownames('Sample')

HTGTS.cols <- intersect(colnames(HTGTS.data),selected.cols.mouse)

tmp <- HTGTS.data %>%
  dplyr::mutate(celltype=gsub('spleen','Spleen',celltype)) %>%
  gather(variable, value, HTGTS.cols) %>%
  dplyr::group_by(variable, celltype) %>%
  dplyr::mutate(value=scale(value)) %>%
  dplyr::group_by(celltype, Genotype, variable) %>%
  dplyr::summarise(mean=mean(value),
                   se=sd(value)/sqrt(n())) %>%
  tidyr::pivot_wider(id_cols=c('variable'),names_from=c('celltype','Genotype'),values_from=c('mean','se')) %>%
  dplyr::left_join(columns,by=c('variable'='col')) %>%
  dplyr::group_by(variable) %>%
  dplyr::mutate(variable=factor(variable, levels=columns %>% 
                                  dplyr::filter(col %in% HTGTS.cols) %>% 
                                  dplyr::pull(col)),
                var_col=cat_cols[category]) %>%
  dplyr::arrange(variable)

cols_shapes <- tmp %>% 
  dplyr::distinct(category, variable) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(variable=factor(variable, levels=levels(tmp$variable))) %>% 
  dplyr::arrange(variable)

comparisons <- list(c('CH12','WT','Atm'),
                    c('CH12','WT','Trp53bp1'),
                    c('CH12','WT','Lig4'),
                    c('Spleen','WT','Atm'),
                    c('Spleen','WT','Trp53bp1'))

plots <- lapply(comparisons, function(x) 
  tmp %>%
    dplyr::mutate_(mean1=paste0('mean_',x[[1]],'_',x[[2]]),
                   mean2=paste0('mean_',x[[1]],'_',x[[3]]),
                   se1=paste0('se_',x[[1]],'_',x[[2]]),
                   se2=paste0('se_',x[[1]],'_',x[[3]])) %>%
    dplyr::mutate(tstat=abs(mean2-mean1)/(sqrt(se2**2+se1**2)),
                  pval=2*pt(-abs(tstat),4),
                  padj=p.adjust(pval, method='BH'),
                  sig=ifelse(padj < .05, 'sig','ns')) %>%
    ggplot(aes(x=mean1,y=mean2)) +
    geom_abline(slope=1,intercept=0,color='gray',linewidth=.5) +
    geom_point(aes(color=variable,shape=variable,size=sig, alpha=sig)) + 
    stat_cor(method = "pearson", aes(label=..r.label..), size=2.5, show.legend=FALSE, label.y.npc = 'bottom') +
    geom_errorbar(aes(ymin=mean2-se2,
                      ymax=mean2+se2,
                      color=variable,
                      alpha=sig),
                  size=.25) +
    geom_errorbarh(aes(xmin=mean1-se1,
                       xmax=mean1+se1,
                       color=variable,
                       alpha=sig),
                   size=.25) +
    my_theme() +
    labs(color='',shape='') +
    scale_color_manual(values=cols_shapes %>% 
                         dplyr::pull(var_col,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable)) +
    scale_shape_manual(values=cols_shapes %>%
                         dplyr::pull(var_ind,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable)) +
    scale_size_manual(values=c('sig'=2,'ns'=.5),guide='none') +
    scale_alpha_manual(values=c('sig'=1,'ns'=.2),guide='none') +
    guides(color=guide_legend(ncol=4),
           shape=guide_legend(ncol=4)) +
    labs(x=x[[2]],y=x[[3]],title=x[[1]],fill='',color='',size=''))

plot_grid(plot_grid(plotlist=lapply(plots, function(x) x + theme(legend.position='none')), 
                    ncol=3, align='vh'),
          get_legend(plots[[1]] + theme(legend.margin=margin(0,0,0,0,'mm'))), ncol=1, rel_heights=c(.65,.35))
```

```{r FigS3C, fig.width=3.75 * fig_scale, fig.height=2.5 * fig_scale}
cols.here.HTGTS <- c('homology',
                     'pct_inversions',
                     'acceptor_score_W',
                     'donor_score_W')

HTGTS.data %>%
  dplyr::mutate(celltype=factor(gsub('spleen','Spleen',celltype), levels=c('CH12','Spleen')),
                Genotype=factor(Genotype, levels=c('WT','Atm','Lig4','Trp53bp1'))) %>%
  gather(variable, value, cols.here.HTGTS) %>%
  dplyr::select(variable, value, Genotype, celltype) %>%
  dplyr::left_join(columns, by=c('variable'='col')) %>%
  dplyr::mutate(variable=factor(variable, levels=cols.here.HTGTS)) %>%
  dplyr::filter(category %in% c('Diversity','Isotypes','Breaks','Context','Structural')) %>%
  ggplot(aes(x=Genotype, y=value, fill=Genotype)) + 
  geom_boxplot(position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25) + 
  geom_point(position=position_jitterdodge(jitter.width=.2, dodge.width=.75, jitter.height = 0), size=.5) + 
  my_theme() +
  coord_cartesian(clip='off') +
  stat_compare_means(ref.group='WT',method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2) +
  ggh4x::facet_grid2(celltype~variable, switch='y', scales='free_y', independent='y', strip=ggh4x::strip_vanilla(clip='off')) +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='none') +
  scale_fill_manual(values=genotype_colors) +
  labs(x='',y='', fill='') 
```

show homology values for HTGTS data

```{r FigS3D, fig.width=3.25 * fig_scale, fig.height=1.25 * fig_scale}
read.xlsx('data/source_data.xlsx', sheet='HTGTS homology') %>%
  dplyr::mutate(Genotype=factor(Genotype, levels=c('WT','Atm','Lig4','Trp53bp1')),
                celltype=gsub('spleen','Spleen',celltype),
                n_homology=as.numeric(n_homology)) %>%
  dplyr::filter(xor(isotype=='Sa', celltype=='Spleen')) %>%
  dplyr::group_by( celltype, Genotype, isotype, n_homology) %>%
  dplyr::summarise(mean_frac=mean(frac),
                   se_frac=sd(frac)/sqrt(n())) %>%
  ggplot(aes(x=n_homology, y=mean_frac, color=Genotype)) +
  geom_line(aes(group=Genotype)) +
  geom_errorbar(aes(ymin=mean_frac-se_frac,ymax=mean_frac+se_frac),width=0) +
  ggh4x::facet_nested(. ~ celltype  + isotype , scales='free', space='free_x', nest_line = TRUE, switch='y') +
  my_theme() +
  scale_color_manual(values=genotype_colors) +
  labs(color='', y='Fraction junctions') +
  coord_cartesian(clip='off') 
```

```{r FigS3B, fig.width=2.5 * fig_scale, fig.height=3. * fig_scale}
HTGTS.CH12.X <- HTGTS.data[HTGTS.data$celltype=='CH12',HTGTS.cols] %>%
  dplyr::mutate_if(is.numeric, scale) 
HTGTS.spleen.X <- HTGTS.data[HTGTS.data$celltype=='spleen',HTGTS.cols] %>%
  dplyr::mutate_if(is.numeric, scale) 

HTGTS.meta <- HTGTS.data %>%
  dplyr::mutate(Genotype=factor(Genotype,levels=c('WT','Atm','Lig4','Trp53bp1'))) %>%
  dplyr::select(celltype,Genotype)

pca.CH12 <- prcomp(HTGTS.CH12.X)
pcadat.CH12 <- cbind(pca.CH12$x,HTGTS.meta[row.names(HTGTS.CH12.X),])
pca.spleen <- prcomp(HTGTS.spleen.X)
pcadat.spleen <- cbind(pca.spleen$x,HTGTS.meta[row.names(HTGTS.spleen.X),])

p1 <- ggplot(pcadat.CH12,aes(x=PC1,y=PC2, color=Genotype)) +
  geom_point(size=1) +
  geom_polygon(data = pcadat.CH12 %>%
                 dplyr::group_by(Genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=Genotype)) +
  my_theme() +
  theme(aspect.ratio=pca.CH12$sdev[2]**2/pca.CH12$sdev[1]**2,
        legend.position=c(.92,.3),
        legend.background = element_blank()) +
  labs(x=paste0('PC 1 (',100*signif(pca.CH12$sdev[1]**2/sum(pca.CH12$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.CH12$sdev[2]**2/sum(pca.CH12$sdev**2),3),'%)'),
       color='',shape='',fill='', title='CH12')  +
  scale_color_manual(values=genotype_colors) +
  scale_fill_manual(values=genotype_colors)

p2 <- ggplot(pcadat.spleen,aes(x=PC1,y=PC2, color=Genotype)) +
  geom_point(size=1) +
  geom_polygon(data = pcadat.spleen %>%
                 dplyr::group_by(Genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=Genotype)) +
  my_theme() +
  theme(aspect.ratio=pca.spleen$sdev[2]**2/pca.spleen$sdev[1]**2,
        legend.position='none',
        legend.background = element_blank()) +
  labs(x=paste0('PC 1 (',100*signif(pca.spleen$sdev[1]**2/sum(pca.spleen$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.spleen$sdev[2]**2/sum(pca.spleen$sdev**2),3),'%)'),
       color='',shape='',fill='', title='Spleen')  +
  scale_color_manual(values=genotype_colors) +
  scale_fill_manual(values=genotype_colors)

plot_grid(p1,p2, ncol=1, align='vh')
```

these PCAs use `r dim(HTGTS.spleen.X)[2]` (spleen) and `r dim(HTGTS.CH12.X)[2]` (CH12) features

show a PCA of spleen and lymph node data together with PP and CH12 using all switch regions or only Sm+Sa

```{r Fig3B, fig.width=3.5 * fig_scale, fig.height=1.75 * fig_scale}
CH12.data <- read.xlsx('data/source_data.xlsx',sheet='CH12 data') %>%
  tibble::column_to_rownames('Sample')
CH12.data.noSg <- read.xlsx('data/source_data.xlsx',sheet='CH12 data (no Sg)') %>%
  tibble::column_to_rownames('Sample')
mouse.data.noSg <- read.xlsx('data/source_data.xlsx',sheet='mouse tissues (no Sg)')

mouse.cols <-  intersect(selected.cols.mouse, intersect(colnames(CH12.data),
                                                        colnames(mouse.data))) 

CH12.take <- CH12.data$Batch != '20240830'
mouse.take <- rep(TRUE, dim(mouse.data)[1]) 
mouse.X <- rbind(CH12.data[CH12.take,mouse.cols],
                 mouse.data[mouse.take,mouse.cols]) %>%
  dplyr::mutate_if(is.numeric, scale)
mouse.cols <- mouse.cols[colSums(is.na(mouse.X))==0]

pca <- prcomp(mouse.X[,mouse.cols])
pcadat <- cbind(pca$x,data.frame(material=c(rep('CH12',sum(CH12.take)),
                                            str_to_title(mouse.data$material[mouse.take])),
                                 Batch=c(CH12.data$Batch[CH12.take],
                                         mouse.data$Batch[mouse.take]),
                                 Genotype=c(CH12.data$Genotype[CH12.take],
                                            rep('WT',sum(mouse.take)))))

p1 <- ggplot(pcadat,aes(x=PC1,y=PC2, color=material)) +
  geom_point(aes(shape=Genotype),size=1) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(material) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=material)) +
  my_theme() +
  scale_fill_manual(values=tissue_colors) +
  scale_color_manual(values=tissue_colors) +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='', title='all switch regions') 

CH12.take.noSg <- rep(TRUE, dim(CH12.data.noSg)[1]) 
mouse.take.noSg <- rep(TRUE, dim(mouse.data.noSg)[1]) 
mouse.cols.noSg <-  grep('SG|Sg|sequential',mouse.cols,
                         value=TRUE,invert=TRUE)
mouse.X.noSg <- rbind(CH12.data.noSg[CH12.take.noSg,mouse.cols.noSg],
                     mouse.data.noSg[mouse.take.noSg,mouse.cols.noSg]) %>%
  dplyr::mutate_if(is.numeric, scale)

mouse.cols.noSg <- mouse.cols.noSg[colSums(is.na(mouse.X.noSg))==0]

pca.noSg <- prcomp(mouse.X.noSg[,mouse.cols.noSg])
pcadat.noSg <- cbind(pca.noSg$x,data.frame(material=c(rep('CH12',sum(CH12.take.noSg)),
                                                      str_to_title(mouse.data.noSg$material[mouse.take.noSg])),
                                           Batch=c(CH12.data.noSg$Batch[CH12.take.noSg],
                                                   mouse.data.noSg$Batch[mouse.take.noSg]),
                                           Genotype=c(CH12.data.noSg$Genotype[CH12.take.noSg],
                                                      rep('WT',sum(mouse.take.noSg)))))

p2 <- ggplot(pcadat.noSg,aes(x=PC1,y=PC2, color=material)) +
  geom_point(aes(shape=Genotype), size=1) +
  geom_polygon(data = pcadat.noSg %>%
                 dplyr::group_by(material) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=material)) +
  my_theme() +
  scale_fill_manual(values=tissue_colors) +
  scale_color_manual(values=tissue_colors) +
  labs(x=paste0('PC 1 (',100*signif(pca.noSg$sdev[1]**2/sum(pca.noSg$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.noSg$sdev[2]**2/sum(pca.noSg$sdev**2),3),'%)'),
       color='',shape='',fill='', title='only Sm + Sa') 

plot_grid(p1 + theme(legend.position='none'),
          p2 + theme(legend.position='none'),
          get_legend(p1 + theme(legend.box='vertical')),
          rel_widths=c(.4,.4,.2),
          ncol=3)
```

(this PCA uses `r dim(mouse.X)[2]` and `r dim(mouse.X.noSg)[2]` features)

selected features in the CH12 data (2023 batches)

```{r Fig3D,fig.width=3.5 * fig_scale,fig.height=2.5 * fig_scale}
CH12.features <- c('pct_direct_switch',
                   'read_length_mean',
                   'pct_duplications',
                   'pct_inversions',
                   'pct_templated_inserts',
                   'cluster_size_mean')

CH12.genotypes <- c('WT','Trp53bp1 KO','Rif1 KO','Lig4 KO','Brca1 KO')

CH12.meta <- CH12.data %>%
  dplyr::mutate(Batch=as.character(Batch),
                Genotype=factor(Genotype,levels=CH12.genotypes)) %>%
  dplyr::select(Batch,Genotype)

tmp <- CH12.data %>% 
  dplyr::filter(Batch != '20240830') %>%
  dplyr::select(Genotype,CH12.features) %>%
  dplyr::mutate(Genotype=factor(Genotype,levels=CH12.genotypes)) %>%
  dplyr::group_by(Genotype) %>%
  dplyr::mutate(ng=n(),
                Genotype_new=paste0(Genotype,' (n=',ng,')')) %>%
  gather(col,value,-c(Genotype,ng,Genotype_new)) %>% 
  dplyr::mutate(col=factor(col, levels=CH12.features))

ggplot(tmp, aes(x=Genotype,y=value,fill=Genotype)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1,height=0,size=.5) + 
  stat_compare_means(ref.group='WT',method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2) +
  facet_wrap(~col,ncol=3,scales='free_y') +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        strip.clip='off',
        axis.text.x=element_text(angle=90,hjust=1,vjust=.5)) +
  scale_fill_manual(values=genotype_colors) +
  scale_x_discrete(breaks=tmp$Genotype,labels=tmp$Genotype_new) +
  labs(x='',y='', color='# Reads')
```

a PCA with all selected features
  
```{r Fig3E, fig.width=2. * fig_scale, fig.height=1.7 * fig_scale}
CH12.samples <- row.names(CH12.data)[CH12.data$Batch != '20240830']
CH12.cols <- setdiff(grep('SG|Sg',intersect(selected.cols.mouse,colnames(CH12.data)),value=TRUE,invert=TRUE),
                     c('pct_sequential_switch','pct_clusters_Sm'))
CH12.X <- CH12.data[CH12.samples,CH12.cols] %>%
  dplyr::mutate_if(is.numeric, scale)
CH12.cols <- CH12.cols[colSums(is.na(CH12.X))==0]

pca <- prcomp(CH12.X[CH12.samples,CH12.cols])
pcadat <- cbind(pca$x,CH12.meta[CH12.samples,]) 

ggplot(pcadat,aes(x=PC1,y=PC2, color=Genotype)) +
  geom_point(size=2) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(Genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=Genotype)) +
  my_theme() +
  theme(aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position='none') +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='') +
  scale_fill_manual(values=genotype_colors) +
  scale_color_manual(values=genotype_colors) 
```

(this PCA uses `r dim(CH12.X)[2]` features)

heatmap of all selected features

```{r Fig3C, fig.width=3.5 * fig_scale, fig.height=3.75 * fig_scale}
breaks <- seq(-2,2,.2)
samples <- CH12.meta %>%
  tibble::rownames_to_column('sample') %>%
  dplyr::filter(Batch != '20240830') %>%
  dplyr::arrange(Genotype) %>%
  dplyr::pull(sample)

dend <- hclust(dist(CH12.X[samples,]))
dend <- reorder(as.dendrogram(dend),wts=order(match(CH12.meta %>%
                                                      tibble::rownames_to_column('sample') %>%
                                                      dplyr::mutate(Genotype=factor(Genotype, levels=unique(Genotype))) %>%
                                                      dplyr::arrange(Genotype) %>%
                                                      dplyr::pull(sample),
                                                    samples)),agglo.FUN=mean)

hm1 <- Heatmap(t(CH12.X[samples,CH12.cols]),
               col=colorRamp2(c(-2,0, 2), c("blue",'gray90', "red")),
               cluster_columns=dend,
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=FALSE,
               show_row_dend=TRUE,
               column_dend_height = unit(5, "mm"),
               top_annotation=HeatmapAnnotation(df=CH12.meta[samples,c('Genotype','Batch')],
                                                show_annotation_name=TRUE,
                                                show_legend=c(TRUE,FALSE),
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_name_gp=gpar(fontsize=7),
                                                annotation_legend_param=list(labels_gp=gpar(fontsize=7),
                                                                             title_gp=gpar(fontsize=7),
                                                                             legend_width=unit(2,'mm'),
                                                                             legend_height=unit(2,'mm'),
                                                                             legend_direction="horizontal",
                                                                             nrow=1),
                                                col=list('Genotype'=genotype_colors[levels(CH12.meta$Genotype)],
                                                         'Batch'=batch_colors[unique(CH12.meta[samples,'Batch'])])),
               column_names_gp=gpar(fontsize=7),
               row_names_gp =gpar(fontsize = 7),
               row_split = columns %>% tibble::column_to_rownames('col') %>% .[CH12.cols, 'category'], 
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=7),
                                         legend_direction = "horizontal", 
                                         legend_width=unit(20,'mm'),
                                         legend_height=unit(2,'mm')),
               row_title=NULL,
               name='z-score')

draw(hm1, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")

```

do ridge regression

```{r Fig3G, fig.width=1.25 * fig_scale, fig.height=3.5 * fig_scale}
train_batches <- c('20220425','20221118','20221212','20230120','20230213')
test_batches <- '20240830'
train <- CH12.meta$Batch %in% train_batches
test <- CH12.meta$Batch %in% test_batches

CH12.X <- CH12.data[, CH12.cols] %>%
  dplyr::mutate_if(is.numeric, function(x) { 
    s <- numeric(sum(train | test)); 
    s[train] <- (x[train]-mean(x[train]))/sd(x[train]);
    s[test] <- (x[test]-mean(x[test]))/sd(x[test]);
    s
  }) 

set.seed(3)
cv_model <- cv.glmnet(as.matrix(CH12.X[train,]), CH12.meta$Genotype[train], alpha = 0, family='multinomial')

mod <- glmnet(as.matrix(CH12.X[train,]), CH12.meta$Genotype[train], alpha = 0, family = "multinomial",
              lambda = cv_model$lambda.min)

tmp <- as.data.frame(predict(mod, type = "response", 
                              newx=as.matrix(CH12.X[test,]))) %>%
  tibble::rownames_to_column('sample') %>% 
  gather(gt,p,-sample) %>% 
  dplyr::mutate(prediction=factor(gsub('.s0','',gt),levels=CH12.genotypes),
                p=as.numeric(p)) %>%
  dplyr::left_join(CH12.meta[test,] %>% 
                     tibble::rownames_to_column('sample') %>% 
                     dplyr::select(sample,Genotype),by='sample') %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(c=ifelse(p == max(p),ifelse(Genotype==prediction, 'green','red'),'black'),
                Genotype=factor(Genotype,levels=CH12.genotypes),
                sample=gsub('[0-9]*_(BC[0-9]*)_switch','\\1',sample)) %>%
  dplyr::ungroup() %>%
  dplyr::arrange(Genotype) %>%
  dplyr::mutate(sample=factor(gsub('20240830_','',sample),
                              levels=gsub('20240830_','',unique(sample)))) 
ggplot(tmp, aes(x=prediction,y=sample,
                label=round(p,2),fill=p,color=c)) + 
  geom_tile() + 
  geom_text(data=. %>% dplyr::filter(p > .25),
            size=2) +
  scale_fill_gradient(low='white',high='steelblue3') +
  scale_color_manual(values=c(black='black',green='darkgreen',red='red3')) +
  my_theme() +
  theme(legend.position='none',
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x=element_text(angle=90, hjust=1, vjust=.5)) +
  labs(x='',y='36 new samples') 
```

(here we have `r sum(train)` samples in the training data and `r sum(test)` in the testing)

show the ridge coefficients 

```{r Fig3F, fig.width=2. * fig_scale, fig.height=2.5 * fig_scale}
tmp <- do.call(cbind,lapply(coef(mod), function(x) x[,'s0'])) %>%
  as.data.frame() %>%
  dplyr::filter(rowSums(abs(.)) > 0) %>%
  dplyr::slice_max(rowSums(abs(.)),n=20) %>%
  dplyr::arrange(rowSums(abs(.))) %>%
  tibble::rownames_to_column('col') %>%
  dplyr::filter(col %in% columns$col) 

col.order <- tmp$col[order.hclust(hclust(dist(tmp[,2:6])))]

tmp <- tmp %>%
  dplyr::mutate(col=factor(col, levels=col.order)) 

anno_grob_y <- grid::rectGrob(y=1:length(tmp$col), x=0, gp=gpar(col=NA, fill=col_cols[rev(levels(tmp$col))], alpha=.5))

tmp %>%
  gather(Genotype,coef,-col) %>%
  dplyr::mutate(Genotype=factor(Genotype, levels=CH12.genotypes),
                col=factor(col, levels=rev(levels(tmp$col)))) %>%
  ggplot(aes(x=Genotype, y=col, fill=coef, label=round(coef,2))) + 
  geom_tile() +
  scale_fill_gradient2(low='pink2',high='steelblue3', mid='white', breaks=c(-.3, 0, .3)) +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position=c(-1,-.15),
        legend.text=element_text(angle=90, hjust=.5, vjust=1),
        legend.direction='horizontal') +
  annotation_custom(grob=anno_grob_y, ymin = 0, ymax = 1, xmin = -4., xmax = 4.5) +
  labs(x='',y='') 
```

evaluate ridge regression by cross-validation

```{r FigS3X, fig.width=1. * fig_scale, fig.height=1.75 * fig_scale, include=FALSE}
batch.CV <- c()
for (batch in unique(CH12.meta$Batch)) {
  train <- CH12.meta$Batch != batch
  test <- CH12.meta$Batch == batch
  X <- CH12.data[, CH12.cols] %>%
    dplyr::mutate_if(is.numeric, function(x) { 
      s <- numeric(sum(train | test)); 
      s[train] <- (x[train]-mean(x[train]))/sd(x[train]);
      s[test] <- (x[test]-mean(x[test]))/sd(x[test]);
      s
      }) %>%
    as.matrix()
  
  cv_model <- cv.glmnet(X[train,], CH12.meta$Genotype[train], alpha = 0, family='multinomial')
  mod <- glmnet(X[train,], CH12.meta$Genotype[train], alpha = 0, family = "multinomial",
                lambda = cv_model$lambda.min)
  batch.CV[[batch]] <- mean(predict(mod, type = "class", newx=X[test,])[,'s0'] == CH12.meta[test,'Genotype'])
}  

set.seed(4)
sample.CV <- c()
for (k in seq(1,10)) {
  train <- runif(dim(CH12.meta)[1]) < .75
  test <- !train
  X <- CH12.data[, CH12.cols] %>%
    dplyr::mutate_if(is.numeric, function(x) { 
      s <- numeric(sum(train | test)); 
      s[train] <- (x[train]-mean(x[train]))/sd(x[train]);
      s[test] <- (x[test]-mean(x[test]))/sd(x[test]);
      s
      }) %>%
    as.matrix()

  cv_model <- cv.glmnet(X[train,], CH12.meta$Genotype[train], alpha = 0, family='multinomial')
  mod <- glmnet(X[train,], CH12.meta$Genotype[train], alpha = 0, family = "multinomial",
                lambda = cv_model$lambda.min)
  sample.CV[[as.character(k)]] <- mean(predict(mod, type = "class", newx=X[test,])[,'s0'] == CH12.meta[test,'Genotype'])
}

data.frame(acc=c(unlist(batch.CV),unlist(sample.CV)),
           fold=c(names(batch.CV),names(sample.CV))) %>%
  dplyr::filter(!is.na(acc)) %>%
  dplyr::mutate(method=ifelse(fold %in% CH12.meta$Batch,'Batch','Sample')) %>%
  dplyr::group_by(method) %>%
  dplyr::mutate(mean_acc=mean(acc)) %>%
  ggplot(aes(x=method,y=acc)) +
  geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1, height=0) + 
  scale_y_continuous(limits=c(0,1)) +
  my_theme() +
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=.5)) +
  labs(x='',y='Accuracy')
```

(average accuracy is `r mean(unlist(batch.CV))` for cross-validation across batches and `r mean(unlist(sample.CV))` for cross-validation across samples)


# Figure 4 + supplement (FB data)

explained variance for FB cohort

```{r Fig4B,fig.width=1.25 * fig_scale, fig.height=1.7 * fig_scale}
FB.data <- read.xlsx('data/source_data.xlsx', sheet='CVID data') %>%
  dplyr::mutate(Batch=as.character(Batch)) %>%
  tibble::column_to_rownames('Sample')
FB.X <- FB.data[grepl('2023',row.names(FB.data)),intersect(colnames(FB.data),columns$col)] %>%
  dplyr::mutate_if(is.numeric, scale)
FB.meta <- FB.data %>%
  dplyr::filter(grepl('2023',Batch)) %>%
  dplyr::select(Donor,Batch,Sex,Age,Diagnosis,Timepoint,Treatment,IV)

FB.var <- fitExtractVarPartModel(t(FB.X[,colSums(is.na(FB.X))==0]), 
                                 ~ (1 | Donor) + (1 | Diagnosis) + (1 | Sex) + Age + (1 | Batch) + (1 | Treatment) + (1 | IV),
                                 FB.meta) %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  gather(variable,variance,-col)

FB.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable,levels=c('Donor','Diagnosis','Sex','Age','Batch','Treatment','IV',
                                                  'Complication','Residuals'))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('Isotypes','Diversity','Structural','Context','Breaks','Variants')) %>%
  ggplot(aes(x=variable, y=100*variance)) + 
  geom_boxplot(aes(fill=variable), outlier.shape=NA, lwd=.5, legend=FALSE) + 
  geom_jitter(position=position_jitter(width=.25), size=.25) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='none',
        legend.box='vertical') +
  scale_fill_manual(values=var_colors, guide='none') +
  labs(x='', y='% Explained variance', fill='', color='')
```

```{r FigS4A,fig.width=3.5 * fig_scale,fig.height=1.7 * fig_scale}
FB.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable, levels=c('Donor','Diagnosis','Sex','Age','Batch','Treatment','IV','Residuals'))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('Isotypes','Diversity','Structural','Context','Breaks','Variants')) %>%
  ggplot(aes(x=category, y=100*variance, fill=variable)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.5) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='top',
        legend.box='horizontal') +
  guides(fill=guide_legend(override.aes = list(shape=22),ncol=7)) +
  scale_fill_manual(values=var_colors) +
  labs(x='', y='% Explained variance', fill='')
```

a PCA with all selected features on pooled FB data from 2023
 
```{r Fig4D, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale} 
FB.data.pooled <- read.xlsx('data/source_data.xlsx',sheet='CVID data pooled') %>%
  dplyr::mutate(status=ifelse(Diagnosis %in% c('Control','Subclass','Atypical'), Diagnosis, 'CVID'),
                Treatment=ifelse((Donor=='FB_35') | (Treatment==''),'none',Treatment),
                IV=ifelse(IV=='','none',IV)) %>%
  tibble::column_to_rownames('Sample')
QP.data <- read.xlsx('data/source_data.xlsx', sheet='DNA repair data') %>%
  dplyr::mutate(Diagnosis=plyr::revalue(Diagnosis,c('control'='Control'))) %>%
  tibble::column_to_rownames('Sample')

FB_train_donors <- paste0('FB_',sprintf('%0.2d', 1:35))
human.meta <- rbind(FB.data.pooled %>%
                      dplyr::mutate(dataset=ifelse(grepl('FB',Donor),
                                                   ifelse(Sample_donor %in% FB_train_donors,'training','testing'),
                                                   'validation')) %>%
                      dplyr::select(Donor,Diagnosis,status,Timepoint, dataset,Sex,Age, pct_B, pct_IgA, pct_IgG),
                    QP.data %>%
                      dplyr::mutate(dataset='validation',
                                    Timepoint='t1',
                                    pct_B=NA,
                                    pct_IgA=NA,
                                    pct_IgG=NA,
                                    status=ifelse(Diagnosis=='Control','Control','CVID')) %>%
                      dplyr::select(Donor,Diagnosis,status,Timepoint,dataset,Sex,Age, pct_B, pct_IgA, pct_IgG)) %>%
  dplyr::mutate(Diagnosis=factor(Diagnosis,levels=c('Control','Subclass','Atypical','CVID','ICOS','Kabuki','IgG4def',
                                                    'AID','TNFRSF13B',
                                                    'NFKB2','PI3K','ATM','BRCA1','LIG4','NIPBL')),
                status=factor(status, levels=c('Control','Subclass','Atypical','CVID')))

human.X <- cbind(rbind(FB.data.pooled[,selected.cols],
                       QP.data[,selected.cols]),
                 human.meta[,'dataset',drop=FALSE]) %>%
  tibble::rownames_to_column('sample') %>%
  dplyr::mutate(sample=gsub('[0-9]*_|pooled_','',sample)) %>%
  dplyr::group_by(dataset) %>%
  dplyr::mutate_if(is.numeric,scale) %>%
  dplyr::ungroup() %>%
  dplyr::select(c('sample',selected.cols)) %>%
  tibble::column_to_rownames('sample')

train <- human.meta$dataset=='training'
test <- human.meta$dataset=='testing'
validation <- human.meta$dataset=='validation'
pca <- prcomp(human.X[train,])
pcadat <- cbind(pca$x,human.meta[train,])

ggplot(pcadat,aes(x=PC1,y=PC2)) +
  geom_point(aes(color=Diagnosis),size=1) +
  geom_line(aes(group=Donor, color=Diagnosis)) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.1, aes(fill=status), legend=FALSE) +
  my_theme() +
  scale_fill_manual(values=diagnosis_colors, guide="none") +
  scale_color_manual(values=diagnosis_colors) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=3)) +
  coord_cartesian(xlim=c(-8,10),
                  ylim=c(-5,8),
                  clip='off') +
  theme(aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.key.size=unit(1.5,'mm'),
        legend.key.spacing.y=unit(0,'mm'),
        legend.position='bottom',
        legend.box='horizontal') +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='')
```

(this PCA uses `r dim(human.X)[2]` features)

show subclass features for HD and subclass-deficient people

```{r FigS4D, fig.width=2.5 * fig_scale, fig.height=1.5 * fig_scale}
subclass.cols <- c("pct_clusters_SA1","pct_clusters_SA2",
                   "pct_clusters_SG1","pct_clusters_SG2","pct_clusters_SG3","pct_clusters_SG4")

tmp <- FB.data.pooled %>%
  dplyr::filter(Sample_donor %in% FB_train_donors, Diagnosis %in% c('Control','Subclass')) %>%
  gather(feature, frac_clusters, subclass.cols) %>%
  dplyr::mutate(feature=factor(gsub('pct_clusters_','',feature),
                               levels=gsub('pct_clusters_','',subclass.cols)))

pvals <- tmp %>%
  dplyr::group_by(feature) %>%
  dplyr::do(wc=wilcox.test(frac_clusters ~ Diagnosis, data=.),
            y=max(.$frac_clusters)) %>%
  dplyr::summarise(feature, p.value=wc$p.value, y)

ggplot(tmp, aes(x=feature, y=frac_clusters)) +
  geom_boxplot(aes(color=Diagnosis), position=position_dodge(width=.8),width=.7,outlier.shape=NA) +
  geom_point(aes(color=Diagnosis), position=position_dodge(width=.8), size=.5) +
  geom_line(data=tmp %>% dplyr::filter(Diagnosis=='Control'),
            aes(group=Donor), position=position_nudge(x=-.2), alpha=.25, size=.25) +
  geom_line(data=tmp %>% dplyr::filter(Diagnosis=='Ssubclass'),
            aes(group=Donor), position=position_nudge(x=+.2), alpha=.25, size=.25) +
  geom_text(data=pvals, aes(label=stars.pval(p.value), y=1.1*y)) +
  scale_color_manual(values=diagnosis_colors) +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(legend.position=c(.8,.8)) +
  labs(color='',x='', y='Fraction clusters')
```

correlate FACS against SWIBRID

```{r FigS4E, fig.width=2.5 * fig_scale, fig.height=1.5 * fig_scale}
tmp <- FB.data.pooled %>%
  dplyr::mutate(alpha_ratio_FACS=pct_IgA/(pct_IgA + pct_IgG)) %>%
  dplyr::filter(Diagnosis %in% c('Control','CVID','CVID-like')) %>%
  dplyr::group_by(status) %>%
  dplyr::mutate(status_R=paste0(status," (R=",signif(cor(alpha_ratio_FACS,alpha_ratio_clusters,use='na.or.complete'),2),")"))
ggplot(tmp, aes(x=alpha_ratio_FACS, y=alpha_ratio_clusters, color=status)) + 
  geom_abline(slope=1, intercept=0, size=.5) +
  geom_point() + 
  my_theme() +
  scale_color_manual(values=diagnosis_colors, breaks=unique(tmp$status), labels=unique(tmp$status_R)) +
  labs(x='FACS', y='SWIBRID', title='IgA / (IgA + IgG)') +
  theme(legend.position=c(.8,.15),
        legend.box=element_blank(),
        legend.title=element_blank())
```


box plots for selected features

```{r Fig4C, fig.width=4 * fig_scale, fig.height=2.7 * fig_scale}
FB.features <- c('pct_clusters_SG2',
                 'pct_clusters_SG3',
                 'read_length_mean',
                 'pct_direct_switch',
                 'homology',
                 'pct_inversions',
                 'cluster_size_mean',
                 'somatic_variants')

tmp <- FB.data.pooled[FB.data.pooled$Sample_donor %in% FB_train_donors,] %>% 
  dplyr::select(Donor,Diagnosis,FB.features) %>%
  dplyr::mutate(Diagnosis_simple=factor(plyr::revalue(Diagnosis, c('AID'='CVID-like',
                                                                   'ICOS'='CVID-like',
                                                                   'IgG4def'='CVID-like',
                                                                   'Kabuki'='CVID-like',
                                                                   'NFKB2'='CVID-like',
                                                                   'PI3K'='CVID-like',
                                                                   'TNFRSF13B'='CVID-like')),
                                        levels=c('Control','Subclass','Atypical','CVID','CVID-like'))) %>%
  dplyr::group_by(Diagnosis_simple) %>%
  dplyr::mutate(ng=n_distinct(Donor),
                Diagnosis_new=paste0(Diagnosis_simple,' (n=',ng,')')) %>%
  gather(col,value,-c(Diagnosis_simple,ng,Diagnosis_new,Donor,Diagnosis)) %>% 
  dplyr::mutate(col=factor(col, levels=FB.features))

ggplot(tmp, aes(x=Diagnosis_simple,y=value,fill=Diagnosis_simple)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1,height=0,size=.5) + 
  stat_compare_means(ref.group='Control',method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2) +
  facet_wrap(~col,ncol=4,scales='free_y') +
  my_theme() +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=diagnosis_colors) +
  theme(legend.position='none',
        axis.text.x=element_text(angle=90,hjust=1,vjust=.5)) +
  scale_x_discrete(breaks=tmp$Diagnosis_simple,labels=tmp$Diagnosis_new) +
  labs(x='',y='', color='# Reads')
```

project test data into the training data PCA

```{r Fig4E, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale}
pcadat.combined <- rbind(cbind(as.matrix(human.X[train,]) %*% pca$rotation, human.meta[train,]),
                         cbind(as.matrix(human.X[test,]) %*% pca$rotation, human.meta[test,]),
                         cbind(as.matrix(human.X[validation,]) %*% pca$rotation, human.meta[validation,])) %>%
  dplyr::mutate(dataset=factor(dataset, levels=c('training','testing','validation')))

ggplot(pcadat.combined %>% dplyr::filter(dataset!='validation'),
       aes(x=PC1,y=PC2)) +
  geom_polygon(data = pcadat.combined %>% dplyr::filter(dataset=='training') %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) +
  geom_point(aes(color=Diagnosis,alpha=dataset, shape=dataset), size=1) +
  geom_line(aes(group=Donor, color=Diagnosis,shape=dataset)) +
  my_theme() +
  scale_fill_manual(values=diagnosis_colors, guide='none') +
  scale_color_manual(values=diagnosis_colors) +
  scale_alpha_manual(values=c('testing'=1,'training'=.25)) +
  scale_shape_manual(values=c('testing'=15,'training'=16)) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=3),
         shape=guide_legend(ncol=1),
         alpha=guide_legend(ncol=1)) +
  theme(legend.key.height=unit(1.5,'mm'),
        legend.key.spacing=unit(0,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position='bottom',
        legend.box='horizontal') +
  coord_cartesian(xlim=c(-8,10),
                  ylim=c(-5,8),
                  clip='off') +
  labs(x='PC 1',
       y='PC 2',
       color='',shape='',fill='',alpha='')
```


project validation data into training PCA

```{r Fig4F, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale}
ggplot(pcadat.combined %>% dplyr::filter(dataset!='testing'),
       aes(x=PC1,y=PC2)) +
  geom_polygon(data = pcadat.combined %>% dplyr::filter(dataset=='training') %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) +
  geom_point(aes(color=Diagnosis,alpha=dataset, shape=dataset), size=1) +
  geom_line(aes(group=Donor, color=Diagnosis,shape=dataset)) +
  my_theme() +
  scale_fill_manual(values=diagnosis_colors, guide='none') +
  scale_color_manual(values=diagnosis_colors) +
  scale_alpha_manual(values=c('training'=.25,'validation'=1)) +
  scale_shape_manual(values=c('training'=16,'validation'=15)) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=3),
         shape=guide_legend(ncol=1),
         alpha=guide_legend(ncol=1)) +
  theme(legend.key.height=unit(1.5,'mm'),
        legend.key.spacing=unit(0,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position='bottom',
        legend.box='horizontal') +
  coord_cartesian(xlim=c(-8,10),
                  ylim=c(-5,8),
                  clip='off') +
  labs(x='PC 1', 
       y='PC 2', 
       color='',shape='',fill='',alpha='')
```

do ridge regression from training to testing or validation data (take out subclass and atypical, as well as biol replicates of training)

```{r Fig4G, fig.width=2. * fig_scale, fig.height=1.5 * fig_scale}
set.seed(1)
train.here <- train & !(human.meta[,'Diagnosis'] %in% c('Subclass','Atypical')) 
test.here <- test & !(human.meta[, 'Donor'] %in% FB_train_donors)
cv_model <- cv.glmnet(as.matrix(human.X[train.here,]), human.meta[train.here,'status']!='Control', alpha = 0, family='binomial')
mod <- glmnet(as.matrix(human.X[train.here,]), human.meta[train.here,'status']!='Control', alpha = 0, family = "binomial",
                lambda = cv_model$lambda.min)
pred_1 <- predict(mod, type = "response", newx=as.matrix(human.X[test.here,]))
ROC_1 <- roc(human.meta[test.here,'status']!='Control', pred_1)

pred_2 <- predict(mod, type = "response", newx=as.matrix(human.X[validation,]))
ROC_2 <- roc(human.meta[validation,'status']!='Control', pred_2)

FB_res <- data.frame(specificity=c(ROC_1$specificities,ROC_2$specificities),
                     sensitivity=c(ROC_1$sensitivities,ROC_2$sensitivities),
                     thresholds=c(ROC_1$thresholds,ROC_2$thresholds),
                     model=c(rep(paste0('training -> testing; AUC=',signif(ROC_1$auc,3)),length(ROC_1$specificities)),
                             rep(paste0('training -> validation; AUC=',signif(ROC_2$auc,3)),length(ROC_2$specificities)))) %>%
  dplyr::mutate(model=factor(model, levels=unique(model))) %>%
  dplyr::arrange(model,sensitivity) 

ggplot(FB_res, aes(x=specificity,y=sensitivity,color=model)) + 
  geom_step() +
  scale_x_reverse() +  
  geom_abline(slope=1,intercept=1,linewidth=.25,linetype='dashed') +
  my_theme() +
  guides(color=guide_legend(ncol=1)) +
  scale_color_manual(values=c('steelblue4','lightsteelblue')) +
  theme(legend.position=c(.5,.2)) +
  labs(color='',linetype='')
```

here we have `r sum(train.here)` samples in the training data set, `r sum(test.here)` in the testing dataset, and `r sum(validation)` in the validation cohort

show the ridge coefficients

```{r FigS4G, fig.width=2.5 * fig_scale, fig.height=1.5 * fig_scale}
tmp <- data.frame(coef=coef(mod)[,'s0']) %>%
  dplyr::filter(rowSums(abs(.)) > 0) %>%
  dplyr::slice_max(rowSums(abs(.)),n=20) %>%
  dplyr::arrange(rowSums(abs(.))) %>%
  tibble::rownames_to_column('col') %>%
  dplyr::filter(col %in% columns$col) 

col.order <- tmp$col[order(tmp$coef)]

tmp <- tmp %>%
  dplyr::mutate(col=factor(col, levels=col.order)) 

anno_grob_x <- grid::rectGrob(x=1:length(tmp$col), y=0, gp=gpar(col=NA, fill=col_cols[rev(levels(tmp$col))], alpha=.5))

tmp %>%
  dplyr::mutate(col=factor(col, levels=rev(levels(tmp$col)))) %>%
  ggplot(aes(x=col, y=1, fill=coef, label=round(coef,2))) + 
  geom_tile() +
  scale_fill_gradient2(low='pink2',high='steelblue3', mid='white', breaks=c(-.05, 0, .05), labels=c('-.05 (HD)','0','+.05 (CVID)')) +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(axis.text.y=element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x=element_text(angle=90, hjust=1, vjust=.5),
        legend.position='right',
        legend.direction='vertical') +
  annotation_custom(grob=anno_grob_x, xmin = 0, xmax = 1, ymin = -2.8, ymax = 3) +
  labs(x='',y='') 
```

evaluate ridge regression by cross-validation

```{r FigS4X, fig.width=.8 * fig_scale, fig.height=1 * fig_scale, include=FALSE}
set.seed(4)
sample.CV <- c()
FB.samples <- row.names(FB.data.pooled)[!(FB.data.pooled$Diagnosis %in% c('Subclass','Atypical'))]
for (k in seq(1,20)) {
  tr <- runif(length(FB.samples)) < .75
  te <- !tr
  X <- FB.data.pooled[FB.samples, selected.cols] %>%
    dplyr::mutate_if(is.numeric, function(x) {
      s <- numeric(sum(tr | te));
      s[tr] <- (x[tr]-mean(x[tr]))/sd(x[tr]);
      s[te] <- (x[te]-mean(x[te]))/sd(x[te]);
      s
      }) %>%
    as.matrix()
  
  y <- FB.data.pooled[FB.samples, 'status'] != 'Control'
  cv_model <- cv.glmnet(X[tr,], y[tr], alpha = 0, family='binomial')
  mod <- glmnet(X[tr,], y[tr], alpha = 0, family = "binomial",lambda = cv_model$lambda.min)
  sample.CV[[as.character(k)]] <- mean(predict(mod, type = "class", newx=X[te,])[,'s0'] == y[te])
}

data.frame(acc=unlist(sample.CV),
           fold=names(sample.CV)) %>%
  dplyr::filter(!is.na(acc)) %>%
  dplyr::mutate(mean_acc=mean(acc)) %>%
  ggplot(aes(x=1,y=acc)) +
  geom_jitter(width=.1, height=0, size=.25) + 
  scale_y_continuous(limits=c(0,1)) +
  scale_x_continuous(limits=c(.85,1.15)) +
  my_theme() +
  theme(axis.text.x=element_blank(),
        axis.ticks = element_blank()) +
  labs(x='',y='Accuracy')
```

(average accuracy is `r mean(unlist(sample.CV))`) 

heatmap

```{r Fig4H, fig.width=5 * fig_scale, fig.height=5.5 * fig_scale}
FB.samples <- row.names(FB.data.pooled)[grepl('FB',FB.data.pooled$Donor)]
FB.X <- FB.data.pooled[FB.samples, selected.cols] %>%
  dplyr::mutate_if(is.numeric, scale)

row.names(FB.X) <- FB.data.pooled[FB.samples, c('Donor','Timepoint')] %>% 
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(Donor,'_',Timepoint)))) %>% 
  dplyr::pull(name)
pca_hm <- pcadat.combined[gsub('pooled_(FB)_([0-9]*)','\\1\\2',FB.samples), c('PC1','PC2','dataset','Donor')]
row.names(pca_hm) <- row.names(FB.X)

FB.meta <- FB.data.pooled[FB.samples,] %>%
  dplyr::mutate(Sample_donor=gsub("FB_","FB",Sample_donor),
                dataset=ifelse(grepl('FB',Donor),
                               ifelse(Sample_donor %in% gsub('_','',FB_train_donors),'training','testing'),
                               'validation')) %>%
  dplyr::select(Donor,Sample_donor,Diagnosis,dataset,Sex,Age,Timepoint,Complication)

dend <- hclust(dist(FB.X))

class_df <- data.frame(Class=plyr::revalue(as.character(cutree(dend, k = 4)),
                                           c('2'='Cluster1',
                                             '3'='Cluster2',
                                             '1'='Cluster3',
                                             '4'='Cluster3')),
                       name=row.names(FB.X))

class_colors <- c('Cluster1'=diagnosis_colors[['Control']],
                  'Cluster2'=cat_cols[['Structural']],
                  'Cluster3'=cat_cols[['Diversity']])

hm_meta <- FB.meta[,c('Diagnosis','Sex','Age','Donor','Timepoint' ,'Complication'),drop=FALSE] %>% 
  tibble::remove_rownames() %>%
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(Donor,'_',Timepoint))),
                Diagnosis=factor(Diagnosis,
                                 levels=c('Control','Subclass','Atypical',
                                          'CVID','ICOS','Kabuki',
                                          'AID','IgG4def','NFKB2','PI3K','TNFRSF13B')),
                Complication=factor(Complication, levels=c('NA','io','c'))) %>%
  dplyr::left_join(class_df,by='name') %>%
  dplyr::select(-c(Donor,Timepoint)) %>%
  tibble::column_to_rownames('name')

hm1 <- Heatmap(t(FB.X),
               col=colorRamp2(c(-2,0, 2), c("blue",'gray90', "red")),
               cluster_columns=dend,
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               top_annotation=HeatmapAnnotation(df=hm_meta,
                                                show_annotation_name=TRUE,
                                                show_legend=TRUE,
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_legend_param=list(title_gp=gpar(fontsize=7),
                                                                             labels_gp=gpar(fontsize=7),
                                                                             legend_height = unit(5, "mm"),
                                                                             legend_width = unit(3, "mm"),
                                                                             grid_height = unit(3, "mm"),
                                                                             grid_width = unit(3, "mm")),
                                                annotation_name_gp=gpar(fontsize=7),
                                                col=list('Diagnosis'=diagnosis_colors,
                                                         'Class'=class_colors,
                                                         'Age'=colorRamp2(c(20,50,80),c("yellow2","orange2","coral3")),
                                                         'Sex'=c('male'='mistyrose2',
                                                                 'female'='lightseagreen'),
                                                         'Complication'=c('NA'='gray80',
                                                                          'io'='blue3',
                                                                          'c'='red3'))),
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize = 7),
               row_dend_width = unit(7, "mm"),
               row_split = columns %>% tibble::column_to_rownames('col') %>% .[selected.cols, 'category'], 
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=7),
                                         at=c(-2,0,2),
                                         grid_height=unit(4, 'mm'),
                                         grid_width=unit(3, 'mm'),
                                         legend_height=unit(5,'mm'),
                                         legend_width=unit(3,'mm')),
               row_title=NULL,
               name='z-score')  


hm2 <- Heatmap(t(FB.X),
               col=colorRamp2(c(-2,0, 2), c("blue",'gray90', "red")),
               column_split = grepl('09|32', row.names(FB.X)), 
               column_title = NULL,
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               top_annotation=HeatmapAnnotation(df=hm_meta,
                                                show_annotation_name=TRUE,
                                                show_legend=TRUE,
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_legend_param=list(title_gp=gpar(fontsize=7),
                                                                             labels_gp=gpar(fontsize=7),
                                                                             legend_height = unit(5, "mm"),
                                                                             legend_width = unit(3, "mm"),
                                                                             grid_height = unit(3, "mm"),
                                                                             grid_width = unit(3, "mm")),
                                                annotation_name_gp=gpar(fontsize=7),
                                                col=list('Diagnosis'=diagnosis_colors,
                                                         'Class'=class_colors,
                                                         'Age'=colorRamp2(c(20,50,80),c("yellow2","orange2","coral3")),
                                                         'Sex'=c('male'='mistyrose2',
                                                                 'female'='lightseagreen'))),
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize = 7),
               row_dend_width = unit(7, "mm"),
               row_split = columns %>% tibble::column_to_rownames('col') %>% .[selected.cols, 'category'], 
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=7),
                                         at=c(-2,0,2),
                                         grid_height=unit(4, 'mm'),
                                         grid_width=unit(3, 'mm'),
                                         legend_height=unit(5,'mm'),
                                         legend_width=unit(3,'mm')),
               row_title=NULL,
               name='z-score')  

hm3 <- Heatmap(t(pca_hm[,c('PC1','PC2')]),
               cluster_columns=dend,
               cluster_rows=TRUE,
               col=colorRamp2(c(-5, 0, 5), c("darkorchid3","gray90","darkseagreen3")),
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize=7),
               row_dend_width = unit(7, "mm"),
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=7),
                                         grid_height=unit(5, 'mm'),
                                         grid_width=unit(3, 'mm'),
                                         legend_height = unit(5, "mm"),
                                         legend_width = unit(3, "mm")),
               row_title=NULL,
               name='PC')

hm1 %v% hm3

hm2 %v% hm3

```

show PCA loadings per category

```{r FigS4C, fig.width=2.5 * fig_scale, fig.height=4 * fig_scale}
tmp <- data.frame(pca$rotation) %>%
  tibble::rownames_to_column('col') %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(col=factor(col, levels=columns %>% 
                             dplyr::filter(col %in% selected.cols) %>% 
                             dplyr::pull(col)))

cols_shapes <- tmp %>% 
  dplyr::distinct(category, col) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(col=factor(col, levels=levels(tmp$col))) %>% 
  dplyr::arrange(col)

ggplot(tmp,aes(x=abs(PC1),y=abs(PC2), color=col, shape=col)) +
  geom_point(size=1) +
  my_theme() +
  scale_color_manual(values=cols_shapes %>%
                       dplyr::pull(var_col,col),
                     labels=cols_shapes %>%
                       dplyr::pull(col)) +
  scale_shape_manual(values=cols_shapes %>%
                       dplyr::pull(var_ind,col),
                     labels=cols_shapes %>%
                       dplyr::pull(col)) +
  theme(legend.key.size=unit(1.5,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.spacing=unit(0,'mm'),
        legend.key.spacing.y=unit(0,'mm'),
        legend.position='bottom',
        legend.box='horizontal') +
  guides(color=guide_legend(ncol=2),
         shape=guide_legend(ncol=2)) +
  labs(x='Abs. PC 1', y='Abs. PC 2',
       color='',shape='',fill='',alpha='')
```

categorize the samples

```{r FigS4H, fig.width=2.5 * fig_scale, fig.height=2 * fig_scale}
ggplot(pca_hm %>% tibble::rownames_to_column('name') %>%
         dplyr::filter(dataset!='validation') %>%
         dplyr::left_join(class_df, by='name'), 
       aes(x=PC1,y=PC2, color=Class, shape=dataset)) +
  geom_point(size=1) +
  geom_point(data=pcadat.combined %>% dplyr::filter(dataset=='validation'),
             aes(color=Diagnosis), size=1) +
  geom_line(aes(group=Donor)) +
  my_theme() +
  scale_shape_manual(values=c('testing'=16,'training'=16,'validation'=15), guide='none') +
  scale_color_manual(values=c(class_colors,diagnosis_colors), 
                     breaks=c(names(class_colors),
                              names(diagnosis_colors))) +
  guides(color=guide_legend(override.aes = list(shape=c(rep(16,3),rep(15,6)),size=2), ncol=4)) +
  theme(legend.key.size=unit(1.5,'mm'),
        legend.position='bottom',
        legend.key.spacing.y=unit(0,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2) +
  labs(x='PC 1', 
       y='PC 2', 
       color='')
```

box plots for selected features

```{r Fig4J, fig.width=5.25 * fig_scale, fig.height=1.7 * fig_scale}
features.here <- c('cluster_size_mean',
                   'pct_intraswitch_deletion',
                   'homology',
                   'untemplated_inserts',
                   'pct_templated_inserts')

groups <- c('Cluster1','Cluster2','Cluster3','Control','BRCA1','ATM')

tmp <- cbind(rbind(FB.data.pooled[,features.here],
                   QP.data[,features.here]),
             human.meta[,c('Diagnosis','dataset','Donor','Timepoint')]) %>%
  tibble::rownames_to_column('Sample_donor') %>%
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(Donor,'_',Timepoint)))) %>%
  dplyr::left_join(class_df,by='name') %>%
  dplyr::mutate(group=ifelse(dataset=='validation',as.character(Diagnosis),as.character(Class))) %>%
  dplyr::filter(group %in% groups) %>%
  dplyr::mutate(group=factor(group,levels=groups)) %>%
  dplyr::group_by(group) %>%
  dplyr::mutate(ng=n_distinct(Donor),
                group_new=paste0(group,' (n=',ng,')')) %>%
  gather(col,value,features.here) %>%
  dplyr::mutate(col=factor(col, levels=features.here))

ggplot(tmp, aes(x=ifelse(grepl('Cluster',group),as.numeric(group),1+as.numeric(group)),
                y=value,fill=group)) +
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_jitter(width=.1,height=0,size=.5) +
  stat_compare_means(comparisons=list(c(1,2),c(1,3),c(2,3)),
                     method='wilcox.test',label = "p.signif", size=2, vjust=.5, hide.ns=TRUE) +
  stat_compare_means(comparisons=list(c(5,6),c(5,7),c(6,7)),
                     method='wilcox.test',label = "p.signif", size=2, vjust=.5, hide.ns=TRUE) +
  facet_wrap(.~col,ncol=length(features.here),scales='free_y') +
  my_theme() +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=setNames(c(rep('gray',3),rep('darkseagreen3',3)),groups)) +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5),
        legend.position='none') +
  scale_x_continuous(breaks=c(1,2,3,5,6,7),
                     labels=groups) +
  labs(x='',y='', color='Diagnosis')
```

check association of clinics with features

```{r FigS4B, fig.width=3 * fig_scale, fig.height=2.5 * fig_scale}
pvals <- list()
for (x in c("Tetanus",
            "Diphteria",
            "PnPS")) {
  for (f in selected.cols) {
    data <- FB.data.pooled[FB.samples,c(x,f)] %>%
      tidyr::drop_na()
    form <- as.formula(paste0(x,'=="negative" ~ ',f))
    res <- glm(form, family='binomial', data=data) %>% summary()
    pvals[[paste0(x,'_',f)]] <- data.frame(pval=coef(res)[2,4],
                                           vacc=x,
                                           col=f)
  }
}

FB.data.pooled[FB.samples, ] %>%
  dplyr::mutate(Tetanus=ifelse(status=='Control','Positive',str_to_title(Tetanus)),
                Diphteria=ifelse(status=='Control','Positive',str_to_title(Diphteria)),
                PnPS=ifelse(status=='Control','Positive',str_to_title(PnPS))) %>%
  gather(feature, value, c('pct_clusters_SA1','pct_clusters_SG3')) %>%
  gather(vacc, vacc_status, c('Tetanus','Diphteria','PnPS')) %>%
  dplyr::mutate(vacc_status=factor(ifelse(vacc_status=='','No data',vacc_status),
                                   c('Negative','Positive','No data'))) %>%
  ggplot(aes(x=vacc_status, y=value, fill=vacc_status)) +
  geom_boxplot(outlier.shape=NA) +
  geom_jitter(aes(color=status),width=.1, size=.5) +
  stat_compare_means(comparisons=list(c('Negative','Positive'),
                                      c('Positive','No data')),
                     method='wilcox.test', label = "p.signif", size=2, vjust=0) +
  facet_grid(feature ~ vacc, switch='y') + 
  my_theme() +
  scale_fill_manual(values=c('Negative'='red3','Positive'='gray')) +
  scale_color_manual(values=diagnosis_colors) +
  labs(x='',y='') +
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        strip.clip='off',
        axis.text.x=element_text(angle=45, hjust=1, vjust=1))
```

check aggregated scores

```{r FigS4F, fig.width=4.5 * fig_scale, fig.height=3.5 * fig_scale}
FB.X.aggregated <-  FB.data.pooled[FB.samples, selected.cols] %>%
  dplyr::mutate_if(is.numeric, scale) %>%
  tibble::rownames_to_column('sample') %>%
  gather(col, value, -sample) %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::group_by(sample, category) %>%
  dplyr::summarise(value=mean(abs(value))) %>%
  dplyr::left_join(FB.meta %>% tibble::rownames_to_column('sample'), by='sample') %>%
  dplyr::filter(!(Diagnosis %in% c('Control','Subclass','Atypical'))) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(Donor,'_',Timepoint)))) %>%
  dplyr::select(name,category,value) %>%
  spread(category,value) %>%
  tibble::column_to_rownames('name')

FB.meta.aggregated <- FB.meta[,c('Diagnosis','Sex','Age',
                                 'Donor','Timepoint'),drop=FALSE] %>% 
  dplyr::filter(!(Diagnosis %in% c('Control','Subclass','Atypical'))) %>%
  tibble::remove_rownames() %>%
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(Donor,'_',Timepoint)))) %>%
  dplyr::left_join(class_df,by='name') %>%
  dplyr::select(-c(Donor,Timepoint)) %>%
  tibble::column_to_rownames('name') %>%
  dplyr::mutate(Diagnosis=factor(Diagnosis,
                                 levels=c('CVID','ICOS','Kabuki',
                                          'AID','IgG4def','NFKB2','PI3K','TNFRSF13B')))

samples <- row.names(FB.meta.aggregated[!(FB.meta.aggregated$Diagnosis  %in% c('Control','Subclass','Atypical')),])
dend <- hclust(dist(FB.X.aggregated[samples,]))
dend <- reorder(as.dendrogram(dend),wts=order(match(FB.meta.aggregated %>%
                                                      tibble::rownames_to_column('sample') %>%
                                                      dplyr::arrange(Class) %>%
                                                      dplyr::pull(sample),
                                                    samples)),agglo.FUN=mean)

hm2 <- Heatmap(t(FB.X.aggregated[samples,]),
               cluster_columns=dend,
               col=colorRamp2(c(0, 2), c("lightyellow2","darkorchid3")),
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               top_annotation=HeatmapAnnotation(df=FB.meta.aggregated[samples,],
                                                show_annotation_name=TRUE,
                                                show_legend=TRUE,
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_legend_param=list(title_gp=gpar(fontsize=7),
                                                                             labels_gp=gpar(fontsize=7),
                                                                             legend_height = unit(5, "mm"),
                                                                             legend_width = unit(3, "mm"),
                                                                             title_position = "leftcenter-rot",
                                                                             grid_height = unit(3, "mm"),
                                                                             grid_width = unit(3, "mm")),
                                                annotation_name_gp=gpar(fontsize=7),
                                                col=list('Diagnosis'=diagnosis_colors,
                                                         'Class'=class_colors,
                                                         'Age'=colorRamp2(c(20,50,80),c("yellow2","orange2","coral3")),
                                                         'Sex'=c('male'='mistyrose2',
                                                                 'female'='lightseagreen'))),
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize = 7),
               row_dend_width = unit(7, "mm"),
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=7),
                                         at=c(0,1,2),
                                         title_position = "leftcenter-rot",
                                         grid_height=unit(4, 'mm'),
                                         grid_width=unit(3, 'mm'),
                                         legend_height=unit(5,'mm'),
                                         legend_width=unit(3,'mm')),
               row_title=NULL,
               name='Category score')  

hm2
```

# session info

```{r sessioninfo}
sessionInfo()
```
