---
title: "paper figures"
author: "Benedikt Obermayer"
date: "2024/12/19"
output: 
  html_document:
    toc: true
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, cache.lazy=FALSE, message=FALSE, warning=FALSE,dev=c('pdf'))
library(ggpubr)
library(stats)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(caret)
library(scales)
library(cowplot)
library(ComplexHeatmap)
library(circlize)
library(ggrepel)
library(car)
library(glmnet)
library(pROC)
library(readxl)
library(variancePartition)
library(RColorBrewer)
library(gtools)
library(grid)
#library(DescTools)
library(lme4)
```

setup

```{r definitions}
fig_scale <- 1.
columns <- list('diversity'=c('nclusters_final_downsampled',
                              'nclusters_eff_downsampled',
                              'mean_cluster_size_downsampled',
                              'std_cluster_size_downsampled',
                              'cluster_gini_downsampled',
                              'cluster_entropy_downsampled',
                              'cluster_inverse_simpson_downsampled',
                              'top_clone_occupancy_downsampled',
                              'big_clones_occupancy_downsampled'),
                'isotypes'=c('alpha_ratio_clusters',
                             'frac_clusters_SA1','frac_clusters_SA2','frac_clusters_SE',
                             'frac_clusters_SG1','frac_clusters_SG2','frac_clusters_SG3',
                             'frac_clusters_SG4','frac_clusters_SM',
                             'frac_clusters_Sa','frac_clusters_Sg3',
                             'frac_clusters_Sg2b','frac_clusters_Sg2c',
                             'frac_clusters_Sm','frac_clusters_Sg1'),
                'structural'=c('insert_frequency',
                               'frac_breaks_inversions','frac_breaks_duplications',
                               'median_duplication_size','median_inversion_size'),
                'context'=c('homology_fw','homology_rv',
                            'donor_score_S', 'donor_score_W', 
                            'donor_score_WGCW', 'donor_score_GAGCT',
                            'donor_score_GGGST',
                            'donor_complexity',
                            'acceptor_score_S', 'acceptor_score_W', 
                            'acceptor_score_WGCW', 'acceptor_score_GAGCT',
                            'acceptor_score_GGGST',
                            'acceptor_complexity',
                            'n_homology_switch', 
                            'n_untemplated_switch',
                            'frac_blunt'),
                'breaks'=c('mean_length',
                           'std_length',
                           'mean_GC',
                           'std_GC',
                           'mean_length_SA',
                           'mean_length_SG',
                           'std_length_SA',
                           'std_length_SG',
                           'frac_breaks_single',
                           'frac_breaks_sequential',
                           'frac_breaks_intra',
                           'mean_intra_break_size',
                           'std_intra_break_size',
                           'spread_SM','spread_SA','spread_SG',
                           'spread_SA1','spread_SA2',
                           'spread_SG1','spread_SG2',
                           'spread_SG2B','spread_SG2C',
                           'spread_SG3','spread_SG4',
                           'frac_breaks_SM_SM','frac_breaks_SM_SG',
                           'frac_breaks_SG_SG','frac_breaks_SM_SA',
                           'frac_breaks_SG_SA','frac_breaks_SA_SA'),
                'context_isotype'=c('homology_fw_SM_SA','homology_rv_SM_SA',
                                    'homology_fw_SM_SG','homology_rv_SM_SG',
                                    'donor_score_GAGCT_SM_SG','donor_score_GGGST_SM_SA',
                                    'donor_score_GGGST_SM_SG','donor_score_S_SM_SA',
                                    'donor_score_S_SM_SG','donor_score_W_SM_SA',
                                    'donor_score_W_SM_SG','donor_score_WGCW_SM_SA',
                                    'donor_score_WGCW_SM_SG','donor_score_GAGCT_SM_SA',
                                    'donor_complexity_SM_SG','donor_complexity_SM_SA',
                                    'acceptor_score_GAGCT_SM_SA','acceptor_score_GAGCT_SM_SG',
                                    'acceptor_score_GGGST_SM_SA','acceptor_score_GGGST_SM_SG',
                                    'acceptor_score_S_SM_SA','acceptor_score_WGCW_SM_SG',  
                                    'acceptor_score_S_SM_SG','acceptor_score_W_SM_SA',
                                    'acceptor_score_W_SM_SG','acceptor_score_WGCW_SM_SA',
                                    'acceptor_complexity_SM_SA',
                                    'acceptor_complexity_SM_SG',
                                    'n_homology_switch_SA', 
                                    'n_homology_switch_SG', 
                                    'n_untemplated_switch_SA',
                                    'n_untemplated_switch_SG',
                                    'frac_blunt_SA',
                                    'frac_blunt_SG'),
                'diversity_raw'=c('nclusters_final',
                                  'nclusters_eff',
                                  'mean_cluster_size',
                                  'std_cluster_size',
                                  'cluster_gini',
                                  'cluster_entropy',
                                  'cluster_inverse_simpson',
                                  'top_clone_occupancy',
                                  'big_clones_occupancy'),
                'variants'=c('frac_variants_germline',
                             'frac_variants_transitions',
                             'num_variants')) %>%
  unlist() %>%
  data.frame(col=.) %>%
  tibble::rownames_to_column('category') %>%
  dplyr::mutate(category=factor(gsub('[0-9]*$','',category), 
                                levels=c('library',
                                         'isotypes',
                                         'diversity',
                                         'diversity_raw',
                                         'structural',
                                         'context',
                                         'context_isotype',
                                         'breaks',
                                         'variants')))

cat_cols <- c('structural'='#FD9E40',
              'context'='#FED6AC',
              'context_isotype'='#FED6AC',
              'breaks'='#1E8AC4',
              'isotypes'='#AEF58E',
              'diversity_raw'='#3A7A2C',
              'diversity'='#3A7A2C',
              'variants'='lightpink3',
              'library'='gray40')

genotype_colors <- c('WT'='gray40','Xlf'='#e1c2f0','Shld1'='#b870dc','Shld1_Xlf'='#a447d2',
                     'Trp53bp1 KO'='darksalmon','Brca1 KO'='darkkhaki','Rif1 KO'='skyblue1','Lig4 KO'='mediumpurple2', 
                     'Trp53bp1'='darksalmon','Atm'='#A65628','Lig4'='mediumpurple2', 'Rif1'='skyblue1', 'Brca1'='darkkhaki')

tissue_colors <- c('splenocytes'='orange2','lymph node'='orangered3','CH12'='forestgreen')
var_colors <- c('batch'='palegreen2','diagnosis'='tan2','donor'='firebrick4','sequencing'='slategray1','unexplained'='gray80',
                'sex'='steelblue2','age'='darkseagreen4')

diagnosis_colors <- c('control'='gray20','HD'='gray20',
                      'subclass'='lightslategray','CVID'='coral1',
                      'ICOS'='#8DD3C7','Kabuki'='#BEBADA',
                      'IgG4def'='burlywood4','TNFRSF13B'='palegreen4',
                      'other'='#BC80BD','AID'='darkgoldenrod3',
                      'ATM'='#A65628','LIG4'='mediumpurple2',
                      'NIPBL'='#FCCDE5','BRCA1'='darkkhaki',
                      'NFKB2'='#FFED6F','PI3K'='#80B1D3')

batch_colors <- setNames(c(RColorBrewer::brewer.pal('Set1',n=9),RColorBrewer::brewer.pal('Set1',n=6)),
                         c('20230728','20230731','20230904','20230908','20231025',
                           '20231030','20240603','20240711','20240725','20220425',
                           '20221118','20221212','20230120','20230213','20240830'))

group_colors <- c('SmB+CD21n'='gray20',
                  'SmB-CD21lo'='orange3',
                  'SmB-CD21n'='lightseagreen')

col_cols <- columns %>% 
  dplyr::mutate(color=cat_cols[as.character(category)]) %>%
  dplyr::mutate(col=gsub('_downsampled|_switch','',col)) %>%
  dplyr::pull(color, col)

my_theme <- function() {
  theme_classic(base_size=7) +
  theme(strip.background=element_blank(),
        strip.placement='outside',
        strip.text = element_text(size=7),
        legend.key.size=unit(2.5, 'mm'),
        legend.spacing=unit(1,'mm'),
        legend.text = element_text(size=7),
        axis.text = element_text(size=7),
        axis.title = element_text(size=7))
}
```

# Figure 1 + supplements

plot # clusters vs. # clones

```{r Fig1D, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
# simulation data for Fig. 1
synthetic.data <- read.csv('data/synthetic_data.csv',header=1, row.names=1)
ggplot(synthetic.data %>% 
         dplyr::mutate(nreads=factor(nreads, levels=c('10k','20k','50k','100k'))), 
       aes(x=nclones, y=nclusters_final, color=nreads)) + 
  geom_abline(slope=1,intercept=0, size=.5, color='gray80') +
  geom_point(size=.5) + 
  geom_smooth(method='lm', alpha=.2, size=.5, se=FALSE) +
  scale_x_log10() + 
  scale_y_log10() +
  my_theme() +
  scale_color_brewer(palette='Set2') +
  stat_cor(method = "pearson", aes(label=..rr.label..), size=2, show.legend=FALSE) +
  theme(legend.position=c(.85,.35),
        legend.background = element_rect(fill=NA)) +
  labs(x='# clones', y='# filtered clusters', color='# reads')
```

cell switch: plot # clusters vs. # cells

```{r Fig1E, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
# cell number experiment for  Fig. 1
cellswitch.data <- read.csv('data/cellswitch_data.csv',header=1, row.names=1)

ggplot(cellswitch.data %>%
         dplyr::mutate(donor=plyr::revalue(donor,
                                           c('21084'='A',
                                             '21085'='B',
                                             '21086'='C'))),
       aes(x=ncells, y=nclusters_final)) + 
  geom_abline(slope=1,intercept=0, size=.5, color='gray80') +
  geom_point(aes(color=donor),size=.5) + 
  geom_smooth(size=.5, method='lm', alpha=.2, color='black') +
  stat_cor(method = "pearson", aes(label=..rr.label..), size=2, show.legend=FALSE) +
  scale_x_log10() + 
  scale_y_log10() +
  my_theme() +
  scale_color_brewer(palette='Set1') +
  theme(legend.position=c(.8,.25),
        legend.background = element_rect(fill=NA)) +
  labs(x='# input cells', y='# filtered clusters', color='donor')
```

show VDJ results: # clones depending on # of input cells

```{r Fig1F, fig.width=1.4 * fig_scale, fig.height=1.4 * fig_scale}
# bulk VDJ results for Fig. 1
VDJ_results <- read.csv('data/VDJ_results.csv',header=1, row.names=1)

vdj_stats <- VDJ_results %>%
  dplyr::mutate(clone=paste0(gsub('\\([0-9]*\\)$','',allVHitsWithScore,'_',str_length(aaSeqCDR3)))) %>%
  dplyr::group_by(donor,ncells,sample) %>%
  dplyr::summarise(nclusters_final=n_distinct(clone),
                   alpha_ratio_clusters=mean(grepl('IGHA',allCHitsWithScore))) %>%
  dplyr::mutate(method='VDJ')

swibrid_stats <- cellswitch.data %>%
  dplyr::select(donor,ncells,
                nclusters_final,
                alpha_ratio_clusters) %>%
  dplyr::filter(ncells %in% c(50000,100000)) %>%
  dplyr::mutate(method='swibrid')

vdj_cols <- intersect(colnames(swibrid_stats),colnames(vdj_stats))

vdj_df <- rbind(swibrid_stats[,vdj_cols],vdj_stats[,vdj_cols]) %>%
  gather(metric,value, c(nclusters_final,alpha_ratio_clusters)) %>%
  dplyr::mutate(ncells=factor(as.character(as.integer(ncells)), levels=c('50000','100000')))

vdj_df %>% 
  dplyr::filter(metric=='nclusters_final') %>%
  ggplot(aes(x=factor(gsub('000$','',ncells), levels=c('50','100')), y=value)) +
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_point(aes(color=gsub('donor ','',donor)),size=.5) +
  facet_wrap(~method, ncol=2, scales='free_y') + 
  my_theme() +
  scale_color_brewer(palette='Set1') +
  theme(strip.clip='off',
        legend.position='none',
        legend.background =element_rect(fill=NA)) +
  labs(y='# clones',x='# cells / 1000',color='')
```

isotype ratios in VDJ vs. SWIBRID

```{r Fig1G, fig.width=1.4 * fig_scale,fig.height=1.4 * fig_scale}
vdj_df %>% 
  dplyr::mutate(ncells=factor(gsub('000$','k',ncells), levels=c('50k','100k')),
                donor=gsub('donor ','',donor)) %>%
  dplyr::filter(metric=='alpha_ratio_clusters') %>%
  dplyr::group_by(ncells, donor, method) %>%
  dplyr::summarise(mean_alpha_ratio=mean(value),
                   se_alpha_ratio=sd(value)/sqrt(length(value))) %>%
  pivot_wider(id_cols=c(ncells,donor),names_from=method,values_from=c('mean_alpha_ratio','se_alpha_ratio')) %>%
  ggplot(aes(x=mean_alpha_ratio_VDJ, y=mean_alpha_ratio_swibrid,
             xmin=mean_alpha_ratio_VDJ-se_alpha_ratio_VDJ,
             xmax=mean_alpha_ratio_VDJ+se_alpha_ratio_VDJ,
             ymin=mean_alpha_ratio_swibrid-se_alpha_ratio_swibrid,
             ymax=mean_alpha_ratio_swibrid+se_alpha_ratio_swibrid)) +
  geom_point(aes(color=donor,shape=ncells),size=1) +
  geom_errorbar(linewidth=.25) +
  geom_errorbarh(linewidth=.25) +
  scale_x_continuous(breaks=c(.6,.65)) +
  my_theme() +
  scale_color_brewer(palette='Set1') +
  theme(strip.background=element_blank()) +
  labs(y='fraction IGHA swibrid',x='fraction IGHA VDJ',shape='# cells')
```

# Figure 2 + supplements 

fraction of explained variance for different feature groups in HD cohort

```{r Fig2B,fig.width=7 * fig_scale,fig.height=1.4 * fig_scale}
my_fitExtractVarPartModel <- function(formula, data) {
  # replace fitExtractVarPartModel from variancePartition package by simple function to run over single features one at a time
  # extract relevant parts of code from calcVarPart.R
  # https://github.com/GabrielHoffman/variancePartition/blob/b578069c0e5dab6234516b12e624e3ce995f850d/R/calcVarPart.R#L292-L363
  
  vars <- c(all.vars(formula),'Residuals')
  res <- setNames(rep(NA, length(vars)-1),sort(vars[2:length(vars)]))
  
  try({
    fit <- lmer(formula, data, REML=FALSE)
    w <- weights(fit)
    if (is.null(w)) {
      w <- rep(1, nrow(fit$model))
    }
    varComp <- lapply(lme4::VarCorr(fit), function(fit) attr(fit, "stddev")^2)
    varComp <- varComp[order(names(varComp))]
    idx <- which(colnames(fit@pp$X) != "(Intercept)")
    if (length(idx) > 0) {
      fxeff <- sapply(idx, function(i) {
        fit@pp$X[, i] * lme4::fixef(fit)[i]
      })
      colnames(fxeff) <- colnames(fit@pp$X)[idx]
      N <- nrow(fxeff)
      fixedVar <- apply(fxeff, 2, function(x) {
        weighted.var(x, w) * (N - 1) / N
      })
      varFixedTotal <- weighted.var(rowSums(fxeff), w) * (N - 1) / N
      for (key in names(fixedVar)) {
        varComp[[key]] <- as.array(fixedVar[[key]] / sum(fixedVar) *
                                     varFixedTotal)
      }
    }
    varComp$Residuals <- sigma(fit)^2
    varComp <- unlist(varComp)
    names(varComp) <- gsub("\\.\\(Intercept\\)", "", names(varComp))
    res <- varComp/sum(varComp)
  })
  res
}

# HD data for Fig. 2
HD.data <- read.csv('data/HD_data.csv', header=1, row.names=1)

HD.scaled <- HD.data %>%
  dplyr::filter(cohort=='C1') %>%
  dplyr::group_by(feature) %>%
  dplyr::mutate(value=as.vector(scale(value))) %>%
  dplyr::filter(!is.na(value)) %>%
  dplyr::ungroup()

HD.var <- sapply(unique(HD.scaled$feature), 
                 function(ftr)
                   tryCatch(my_fitExtractVarPartModel(value ~ (1 | donor) + (1 | sex) + (1 | age) + (1 | sequencing) + (1 | batch),
                                                      HD.scaled %>%
                                                        dplyr::filter(feature==ftr)),
                            error=setNames(rep(NA,5), c('donor','sex','age','sequencing','batch')))) %>%
                   t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  gather(variable,variance,-col)

cols.here <- c('cluster_entropy_downsampled',
               'cluster_gini_downsampled',
               'mean_cluster_size_downsampled',
               'std_cluster_size_downsampled',
               'top_clone_occupancy_downsampled',
               'alpha_ratio_clusters',
               'frac_clusters_SA1',
               'frac_clusters_SA2',
               'frac_clusters_SG1',
               'frac_clusters_SG2',
               'insert_frequency',
               'frac_breaks_inversions',
               'frac_breaks_duplications',
               'median_duplication_size',
               'median_inversion_size',
               'homology_fw',
               'donor_score_WGCW',
               'acceptor_complexity',
               'n_homology_switch',
               'n_untemplated_switch',
               'frac_breaks_single',
               'frac_breaks_sequential',
               'frac_breaks_intra',
               'spread_SM',
               'spread_SA')

HD.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(gsub('Residuals','unexplained',variable),
                                levels=rev(c('donor','sex','age','batch','sequencing','unexplained'))),
                category=factor(category, levels=c('structural','context','breaks','diversity','isotypes'))) %>%
  dplyr::filter(col %in% cols.here) %>%
  dplyr::mutate(col=factor(gsub('_downsampled|_switch','',col), levels=gsub('_downsampled|_switch','',cols.here))) %>%
  ggplot(aes(x=col, y=100*variance, fill=variable)) + 
  geom_col(width=.8) +
  my_theme() +
  facet_wrap(~category,scales='free_y', ncol=5) +
  theme(strip.clip='off',
        legend.position='bottom',
        legend.box='horizontal') +
  guides(fill=guide_legend(ncol=6)) +
  coord_flip() +
  scale_y_continuous(breaks=c(0,50,100)) +
  scale_fill_manual(values=var_colors,
                    breaks=c('donor','sex','age','batch','sequencing','unexplained')) +
  labs(x='', y='% explained variance', fill='')
```

```{r Fig2D, fig.width=1.75 * fig_scale,fig.height=1.7 * fig_scale}
donor_cutoff <- .25
batch_cutoff <- .35
sequencing_cutoff <- .2
selected.cols <- HD.var %>%
  dplyr::left_join(columns, by='col') %>%
  spread(variable, variance) %>%
  dplyr::filter(!(category %in% c('diversity_raw','context_isotype','library')), 
                !(col %in% c('log10_nreads','mean_GC')), 
                donor >= donor_cutoff, 
                sequencing < sequencing_cutoff, 
                batch < batch_cutoff) %>%
  dplyr::arrange(category) %>%
  dplyr::pull(col)

HD.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(gsub('Residuals','unexplained',variable),
                                levels=c('sequencing','batch','donor','sex','age'))) %>%
  dplyr::filter(variable!='unexplained', category %in% c('library','isotypes','diversity','structural','context','breaks','variants')) %>%
  ggplot(aes(x=variable, y=100*variance)) + 
  geom_boxplot(aes(fill=variable), outlier.shape=NA, lwd=.5, legend=FALSE) + 
  geom_jitter(aes(color=ifelse(col %in% selected.cols, 'selected','unused')), position=position_jitter(width=.25), size=.25) +
  annotate("rect", xmin = .5, xmax = 1.5, ymax = 100*sequencing_cutoff, ymin = 0, fill='orangered', linetype=1, size=.25, alpha=.25) +
  annotate("rect", xmin = 1.5, xmax = 2.5, ymax = 100*batch_cutoff, ymin = 0, fill='orangered', linetype=1, size=.25, alpha=.25) +
  annotate("rect", xmin = 2.5, xmax = 3.5, ymax = Inf, ymin = 100*donor_cutoff, fill='orangered', linetype=1, size=.25, alpha=.25) +
  annotate('text',.6,78,label='unused',color='black',size=2, hjust=0) +
  annotate('text',.6,73,label='selected',color='orangered',size=2, hjust=0) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='none',
        legend.box='vertical') +
  scale_fill_manual(values=var_colors, guide='none') +
  scale_color_manual(values=c('unused'='black','selected'='orangered')) +
  labs(x='', y='% explained variance', fill='', color='')
```

we now have `r length(selected.cols)` robust features

```{r FigS2F,fig.width=3.5 * fig_scale,fig.height=1.7 * fig_scale}
HD.var %>%  
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable, levels=c('donor','sex','age','batch','sequencing','Residuals'))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('library','isotypes','diversity','structural','context','breaks','variants')) %>%
  ggplot(aes(x=category, y=100*variance, fill=variable)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.25) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='top',
        legend.box='horizontal') +
  scale_fill_manual(values=var_colors) +
  labs(x='', y='% explained variance', fill='')
```

confirm that downsampled diversity measures are better

```{r FigS2J_1, fig.width=1.4 * fig_scale, fig.height=1.7 * fig_scale}
tmp <- HD.var %>%  
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(gsub('log10_nreads','# reads',variable),
                                levels=rev(c('donor','batch','sequencing','# reads','Residuals')))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('diversity','diversity_raw')) 

pvals <- tmp %>%
  dplyr::group_by(variable) %>%
  dplyr::do(wc=wilcox.test(variance ~ category, data=.)) %>%
  dplyr::summarise(variable, p=wc$p.value)

ggplot(tmp,aes(x=variable, y=100*variance, fill=category)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.5) +
  geom_text(data=pvals, aes(x=variable, label=stars.pval(p)), y=Inf, inherit.aes=FALSE, size=5) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='top',
        legend.box='horizontal') +
  guides(fill=guide_legend(title.position='top')) +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=c('diversity'='blue3','diversity_raw'='lightblue1'),
                    labels=c('downsampled','raw')) +
  labs(x='', y='% explained variance', fill='diversity measures')
```

confirm that using isotype-specific context features doesn't help

```{r FigS2J_2, fig.width=1.4 * fig_scale, fig.height=1.7 * fig_scale}
tmp <- HD.var %>%  
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(gsub('log10_nreads','# reads',variable),
                                levels=rev(c('donor','cohort','batch','sequencing','# reads','Residuals')))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('context','context_isotype'), variable!='cohort')

pvals <- tmp %>% 
  dplyr::group_by(variable) %>%
  dplyr::do(wc=wilcox.test(variance ~ category, data=., exact=FALSE)) %>%
  dplyr::summarise(variable, p=wc$p.value)

ggplot(tmp, aes(x=variable, y=100*variance, fill=category)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.5) +
  geom_text(data=pvals, aes(x=variable, label=stars.pval(p)), y=Inf, inherit.aes=FALSE, size=5) +
  my_theme() +
  guides(fill=guide_legend(title.position='top')) +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='top',
        legend.box='horizontal') +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=c('context'='darkorchid3','context_isotype'='orchid1'),
                    labels=c('global','per isotype')) +
  labs(x='', y='% explained variance', fill='context measures')
```

test the effects of averaging over clusters or reads

```{r FigS2J_3, fig.width=1.4 * fig_scale,fig.height=1.7 * fig_scale}
HD.data.br <- read.csv('data/HD_data_by_reads.csv', header=1, row.names=1) 

HD.scaled.br <- HD.data.br %>%
  dplyr::filter(cohort=='C1') %>%
  dplyr::group_by(feature) %>%
  dplyr::mutate(value=as.vector(scale(value))) %>%
  dplyr::filter(!is.na(value)) %>%
  dplyr::ungroup()

HD.var.br <- sapply(unique(HD.scaled.br$feature), 
                    function(ftr)
                      my_fitExtractVarPartModel(value ~ (1 | donor) + (1 | sex) + (1 | age) + (1 | sequencing) + (1 | batch),
                                                HD.scaled.br %>%
                                                  dplyr::filter(feature==ftr))) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  gather(variable,variance,-col)

fts <- intersect(HD.var$col, HD.var.br$col)

tmp <- rbind(HD.var %>% 
               dplyr::mutate(averaging='clusters'), 
             HD.var.br %>%
               dplyr::mutate(averaging='reads')) %>% 
  dplyr::filter(col %in% fts) %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable, levels=rev(c('donor','sex','age','batch','sequencing','Residuals')))) %>%
  dplyr::filter(variable!='Residuals', 
                category %in% c("context","breaks"),
                variable %in% c("donor", "batch", "sequencing")) %>% 
  dplyr::arrange(col,averaging,variable) 

pvals <- tmp %>%
  dplyr::group_by(variable) %>%
  dplyr::do(wc=wilcox.test(variance ~ averaging, paired=TRUE, data=., exact=FALSE)) %>%
  dplyr::summarise(variable, p=wc$p.value)

ggplot(tmp, aes(x=variable, y=100*variance, fill=averaging)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.1,dodge.width=.8), size=.5) +
  geom_text(data=pvals, aes(x=variable, label=stars.pval(p)), y=Inf, inherit.aes=FALSE, size=5) +
  my_theme() +
  coord_cartesian(clip='off') +
  guides(fill=guide_legend(title.position='top')) +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='top',
        legend.box='horizontal') +
  scale_fill_manual(values=c('clusters'='aquamarine4','reads'='darkseagreen1')) +
  labs(x='', y='% explained variance', fill='averages over')
```

show top examples for features related to age or sex

```{r Fig2E, fig.width=.9 * fig_scale, fig.height=1.7 * fig_scale}
age.cols <- HD.var %>%
  dplyr::left_join(columns, by='col') %>%
  spread(variable, variance) %>%
  dplyr::filter(!(category %in% c('library','diversity_raw','context_isotype'))) %>%
  dplyr::slice_max(age, n=1) %>%
  dplyr::pull(col)
age.cols <- 'top_clone_occupancy_downsampled'

sex.cols <- HD.var %>%
  dplyr::left_join(columns, by='col') %>%
  spread(variable, variance) %>%
  dplyr::filter(!(category %in% c('library','diversity_raw','context_isotype'))) %>%
  dplyr::slice_max(sex, n=1) %>%
  dplyr::pull(col)
sex.cols <- 'spread_SG1'

HD.data %>%
  dplyr::filter(feature %in% sex.cols, cohort=='C1') %>%
  dplyr::mutate(feature=factor(gsub('_downsampled|_switch','',feature),
                               levels=gsub('_downsampled|_switch','',sex.cols))) %>% 
  ggplot(aes(x=sex, y=value)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_jitter(width=.2, size=.1) + 
  stat_compare_means(comparisons=list(c('male','female')),
                     method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2.5) +
  facet_wrap(~feature, scales='free', ncol=8, switch='y') + 
  my_theme() + 
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  labs(x='',y='')

HD.data %>%
  dplyr::filter(feature %in% age.cols, cohort=='C1') %>%
  dplyr::mutate(feature=factor(gsub('_downsampled|_switch','',feature),
                               levels=gsub('_downsampled|_switch','',age.cols))) %>% 
  ggplot(aes(x=age, y=value)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_jitter(width=.2, size=.1) + 
  stat_compare_means(comparisons=list(c('<50','>=50')),
                     method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2.5) +
  facet_wrap(~feature, scales='free', ncol=8, switch='y') + 
  my_theme() +
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  labs(x='',y='')
```

HD feature correlation (on FB data)

```{r FigS2H, fig.width=4.25 * fig_scale, fig.height=4.25 * fig_scale}
library(dendextend)
FB.X <- HD.data %>%
  dplyr::filter(cohort=='C2') %>%
  dplyr::select(sample, feature, value) %>%
  spread(feature, value) %>%
  dplyr::mutate_if(is.numeric, scale) %>%
  tibble::column_to_rownames('sample')

R <- cor(FB.X[,selected.cols], use='na.or.complete')
io <- gsub('_downsampled|_switch','',rownames(R)[order.hclust(hclust(dist(R)))])

tmp <- as.data.frame(R) %>% 
  tibble::rownames_to_column('var2') %>%
  gather(var1, R, -var2) %>%
  dplyr::left_join(columns, by=c('var1'='col')) %>%
  dplyr::mutate(cat1=category) %>%
  dplyr::select(-category) %>%
  dplyr::left_join(columns, by=c('var2'='col')) %>%
  dplyr::mutate(cat2=category) %>%
  dplyr::select(-category)  %>%
  dplyr::mutate(var1=factor(gsub('_downsampled|_switch','',var1), levels=io),
                var2=factor(gsub('_downsampled|_switch','',var2), levels=io),
                i1=as.vector(sapply(var1, function(x) which(io==x))),
                i2=as.vector(sapply(var2, function(x) which(io==x)))) 

anno_grob_y <- grid::rectGrob(x=1:length(io), y=0, gp=gpar(col=NA, fill=col_cols[io], alpha=.5))
anno_grob_x <- grid::rectGrob(y=1:length(io), x=0, gp=gpar(col=NA, fill=col_cols[io], alpha=.5))

ggplot(tmp %>%
  dplyr::filter(i1 <= i2),
  aes(x=var1,y=var2,fill=R)) + 
  geom_tile() +
  scale_fill_gradient2(low='blue3',high='red3',mid='gray',limits=c(-1,1)) + 
  my_theme() +
  scale_x_discrete(position = "top") +
  coord_cartesian(clip='off') +
  theme(axis.text.x.top=element_text(angle=90, hjust=0., vjust=.5), #, colour=col_cols[io]),
        legend.position=c(-.1,1.2)) +
  labs(x='',y='') +
  annotation_custom(grob=anno_grob_y, xmin = 0, xmax = 1, ymin = length(io) + 10, ymax = length(io) + 28) +
  annotation_custom(grob=anno_grob_x, xmin = -9, xmax = 8, ymin = 0, ymax = 1)

```

distributions of features in the two HD cohorts

```{r Fig2C, fig.width=4.25 * fig_scale, fig.height=1.5 * fig_scale}
library(gtools)
cols.here.human <- c('insert_frequency',
                     'n_untemplated_switch',
                     'frac_breaks_sequential',
                     'frac_breaks_intra',
                     'cluster_gini_downsampled',
                     'frac_clusters_SG1')

cohort.data <- HD.data %>%
  dplyr::filter(sequencing=='minION') %>% 
  dplyr::filter(feature %in% cols.here.human) %>%
  dplyr::left_join(columns, by=c('feature'='col')) %>%
  dplyr::mutate(feature=factor(gsub('_downsampled|_switch','',feature), 
                               levels=gsub('_downsampled|_switch','',cols.here.human))) %>%
  dplyr::filter(category %in% c('diversity','isotypes','breaks','context','structural')) 
  
ggplot(cohort.data, aes(x=cohort, y=value)) + 
  geom_boxplot(aes(group=cohort), position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25, fill='gray80') + 
  geom_pointrange(data=cohort.data %>%
                    dplyr::group_by(donor,feature,category,cohort) %>%
                    dplyr::summarise(se=sd(value)/sqrt(n()),
                                     value=mean(value),
                                     n=n()) %>% dplyr::filter(n > 1),
                  aes(fill=cohort, ymin=value-se,ymax=value+se, group=cohort),
                  position=position_jitterdodge(jitter.width=.5, dodge.width=.75, jitter.height = 0), 
                  color='black',size=1,stroke=0.1,
                  linewidth=.1,fatten=1) + 
  my_theme() +
  coord_cartesian(clip='off') +
  facet_wrap(~feature, switch='y', scales='free_y', ncol=7) +
  theme(legend.position='none',
        axis.text.x=element_text(angle=90, hjust=1, vjust=.5)) +
  labs(x='',y='')
```

same thing for mouse tissue and Vincendeau

```{r Fig2F, fig.width=4. * fig_scale, fig.height=1.75 * fig_scale}
library(gtools)
mouse.data <- read.csv('data/mouse_data.csv',header=1, row.names=1)  %>%
  tibble::column_to_rownames('sample')
vincendeau.data <- read.csv('data/vincendeau_data.csv', header=1, row.names=1)  %>%
  tibble::column_to_rownames('sample')

cols.here.mouse <- c('frac_breaks_inversions',
                     'n_homology_switch',
                     'mean_cluster_size_downsampled')

cohort.data <- rbind(mouse.data %>%
                       dplyr::mutate(species='mouse tissue',
                                     cohort=material) %>%
                       dplyr::select(species, cohort, cols.here.mouse),
                     vincendeau.data %>%
                       dplyr::mutate(species='Vincendeau',
                                     cohort=genotype) %>%
                       dplyr::select(species, cohort, cols.here.mouse)) %>%
  dplyr::mutate(cohort=factor(cohort, levels=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'))) %>%
  gather(variable, value, cols.here.mouse) %>%
  dplyr::select(variable, value, species, cohort) %>%
  dplyr::left_join(columns, by=c('variable'='col')) %>%
  dplyr::mutate(variable=factor(gsub('_downsampled|_switch','',variable), 
                                levels=gsub('_downsampled|_switch','',cols.here.mouse))) %>%
  dplyr::filter(category %in% c('diversity','isotypes','breaks','context','structural'))

cohort.data %>%
  ggplot(aes(x=cohort, y=value, fill=cohort)) + 
  geom_boxplot(position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25, fill='gray80') + 
  geom_point(position=position_jitterdodge(jitter.width=.2, dodge.width=.75, jitter.height = 0), size=.5) + 
  my_theme() +
  coord_cartesian(clip='off') +
  #annotate('text',x=5.5,y=-2,label='Vincendeau et al.', size=2) +
  facet_wrap(~variable, ncol=3, switch='y', scales='free_y') +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='none') +
  scale_x_discrete(breaks=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
                   limits=c('lymph node','splenocytes','','WT','Xlf','Shld1','Shld1_Xlf'), drop=FALSE) +
  labs(x='',y='', fill='') 
```

check length dependence of Vincendeau data

```{r FigS2K, fig.width=1.5 * fig_scale, fig.height=1.75 * fig_scale}
cols.here.mouse <- c('mean_length_SG')
cohort.data <- rbind(mouse.data %>%
                       dplyr::mutate(species='mouse tissue',
                                     cohort=material) %>%
                       dplyr::select(species, cohort, cols.here.mouse),
                     vincendeau.data %>%
                       dplyr::mutate(species='Vincendeau',
                                     cohort=genotype) %>%
                       dplyr::select(species, cohort, cols.here.mouse)) %>%
  dplyr::mutate(cohort=factor(cohort, levels=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'))) %>%
  gather(variable, value, cols.here.mouse) %>%
  dplyr::select(variable, value, species, cohort) %>%
  dplyr::left_join(columns, by=c('variable'='col')) %>%
  dplyr::mutate(variable=factor(gsub('_downsampled|_switch','',variable), 
                                levels=gsub('_downsampled|_switch','',cols.here.mouse))) %>%
  dplyr::filter(category %in% c('diversity','isotypes','breaks','context','structural'))

cohort.data %>%
  ggplot(aes(x=cohort, y=value, fill=cohort)) + 
  geom_boxplot(position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25, fill='gray80') + 
  geom_point(position=position_jitterdodge(jitter.width=.2, dodge.width=.75, jitter.height = 0), size=.5) + 
  my_theme() +
  coord_cartesian(clip='off') +
  #annotate('text',x=5.5,y=-2,label='Vincendeau et al.', size=2) +
  facet_wrap(~variable, ncol=3, switch='y', scales='free_y') +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='none') +
  scale_x_discrete(breaks=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
                   limits=c('lymph node','splenocytes','','WT','Xlf','Shld1','Shld1_Xlf'), drop=FALSE) +
  labs(x='',y='', fill='')
```

compare minION against pacBio in HD cohort for selected features

```{r FigS2G, fig.width=2.75 * fig_scale, fig.height=5 * fig_scale}
tmp <- HD.scaled %>%
  dplyr::filter(feature %in% selected.cols, batch=='20230313') %>%
  dplyr::select(feature,value,sequencing,donor) %>%
  spread(sequencing,value) %>% 
  dplyr::left_join(columns,by=c('feature'='col')) %>% 
  dplyr::group_by(feature) %>%
  dplyr::mutate(R=cor(minION,pacBio)) %>%
  dplyr::mutate(feature=factor(feature, levels=columns %>% 
                                 dplyr::filter(col %in% selected.cols) %>% 
                                 dplyr::pull(col)),
                feature_name=gsub('_downsampled|_switch','',feature)) %>% #,' (R=',signif(R,2),')')) %>%
  dplyr::arrange(feature)

cols_shapes <- tmp %>% 
  dplyr::distinct(category, feature, feature_name) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(feature=factor(feature, levels=levels(tmp$feature))) %>% 
  dplyr::arrange(feature)

ggplot(tmp,aes(x=minION,y=pacBio)) +
  geom_abline(slope=1,intercept=0,color='gray',linewidth=.5) +
  geom_point(aes(color=feature,shape=feature),size=1) + 
  #geom_text_repel(aes(label=feature), size=1) +
  my_theme() +
  stat_cor(aes(label = ..r.label..), size=2.5) +
  labs(color='',shape='') +
  scale_color_manual(values=cols_shapes %>%
                       dplyr::pull(var_col,feature),
                     labels=cols_shapes %>%
                       dplyr::pull(feature_name)) +
  scale_shape_manual(values=cols_shapes %>%
                       dplyr::pull(var_ind,feature),
                     labels=cols_shapes %>%
                       dplyr::pull(feature_name)) +
  guides(color=guide_legend(ncol=2),
         shape=guide_legend(ncol=2))  +
  theme(legend.position='bottom',
        legend.key.size=unit(1.5,'mm'),
        legend.key.spacing.y=unit(0, 'mm'))
```

show a PCA of the Vincendeau data

```{r Fig2G, fig.width=2.5 * fig_scale, fig.height=1.7 * fig_scale}
selected.cols.mouse <- gsub('clusters_SG1','clusters_Sg2b',
                            gsub('clusters_SG2','clusters_Sg2c',
                                 gsub('clusters_SG3','clusters_Sg3',
                                      gsub('clusters_SA1','clusters_Sa',selected.cols))))
vincendeau.cols <- intersect(colnames(vincendeau.data),selected.cols.mouse)
vincendeau.X <- vincendeau.data[,vincendeau.cols] %>%
  dplyr::mutate_if(is.numeric, scale) 
vincendeau.X <- vincendeau.X[,colSums(is.na(vincendeau.X))==0]
vincendeau.meta <- vincendeau.data %>%
  dplyr::mutate(genotype=factor(genotype,levels=c('WT','Xlf','Shld1','Shld1_Xlf'))) %>%
  dplyr::select(genotype)

pca <- prcomp(vincendeau.X)
pcadat <- cbind(pca$x,vincendeau.meta)
pcarot <- data.frame(coef=pca$rot[,'PC1'],
                     col=gsub('_downsampled|_switch','',row.names(pca$rot)))  %>%
  dplyr::arrange(coef) %>%
  dplyr::mutate(col=factor(col, levels=col)) 

ggplot(pcadat,aes(x=PC1,y=PC2, color=genotype)) +
  geom_point(size=1) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=genotype)) +
  my_theme() +
  theme(aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position=c(.15,.9),
        legend.background = element_blank()) +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='')  +
  scale_color_manual(values=genotype_colors) +
  scale_fill_manual(values=genotype_colors)
```

(this PCA uses `r dim(vincendeau.X)[2]` features)

show mean scaled values of Vincendeau in a scatter plot (big dots are significant, small ones are not)

```{r Fig2H, fig.width=7 * fig_scale, fig.height=2.25 * fig_scale}
tmp <- vincendeau.data %>%
  gather(variable, value, vincendeau.cols) %>%
  dplyr::group_by(variable) %>%
  dplyr::mutate(value=scale(value)) %>%
  dplyr::group_by(genotype, variable) %>%
  dplyr::summarise(mean=mean(value),
                   se=sd(value)/sqrt(n())) %>%
  tidyr::pivot_wider(id_cols=c('variable'),names_from='genotype',values_from=c('mean','se')) %>%
  dplyr::left_join(columns,by=c('variable'='col')) %>%
  dplyr::group_by(variable) %>%
  dplyr::mutate(variable=factor(variable, levels=columns %>% 
                                  dplyr::filter(col %in% vincendeau.cols) %>% 
                                  dplyr::pull(col)),
                var_col=cat_cols[category]) %>%
  dplyr::arrange(variable)

cols_shapes <- tmp %>% 
  dplyr::distinct(category, variable) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)],
                variable_name=gsub('_downsampled','',variable)) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(variable=factor(variable, levels=levels(tmp$variable))) %>% 
  dplyr::arrange(variable)

comparisons <- list(c('WT','Xlf'),
                    c('WT','Shld1'),
                    c('WT','Shld1_Xlf'),
                    c('Shld1','Xlf'))

plots <- lapply(comparisons, function(x) 
  tmp %>%
    dplyr::mutate_(mean1=paste0('mean_',x[[1]]),
                   mean2=paste0('mean_',x[[2]]),
                   se1=paste0('se_',x[[1]]),
                   se2=paste0('se_',x[[2]])) %>%
    dplyr::mutate(tstat=abs(mean2-mean1)/(sqrt(se2**2+se1**2)),
                  pval=dt(tstat,4)) %>%
    ggplot(aes(x=mean1,y=mean2)) +
    geom_abline(slope=1,intercept=0,color='gray',linewidth=.5) +
    geom_point(aes(color=variable,shape=variable,size=ifelse(pval < .05,'sig','ns'))) + 
    stat_cor(method = "pearson", aes(label=..r.label..), size=2.5, show.legend=FALSE) +
    geom_errorbar(aes(ymin=mean2-se2,
                      ymax=mean2+se2,
                      color=variable),
                  size=.25) +
    geom_errorbarh(aes(xmin=mean1-se1,
                       xmax=mean1+se1,
                       color=variable),
                   size=.25) +
    my_theme() +
    labs(color='',shape='') +
    scale_color_manual(values=cols_shapes %>% 
                         dplyr::pull(var_col,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable_name)) +
    scale_shape_manual(values=cols_shapes %>%
                         dplyr::pull(var_ind,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable_name)) +
    scale_size_manual(values=c('sig'=2,'ns'=.5),guide='none') +
    guides(color=guide_legend(ncol=5),
           shape=guide_legend(ncol=5)) +
    labs(x=x[[1]],y=x[[2]],fill='',color='',size=''))

plot_grid(plot_grid(plotlist=lapply(plots, function(x) x + theme(legend.position='none')), 
                    ncol=4, align='vh'),
          get_legend(plots[[1]] + theme(legend.margin=margin(0,0,0,0,'mm'))), ncol=1, rel_heights=c(.65,.35))
```

# Figure 3 + supplement (CH12)

```{r FigS3A, fig.width=5.5 * fig_scale, fig.height=4.5 * fig_scale}
HTGTS.data <- read.csv("data/HTGTS_data.csv", header=1, row.names=1) %>%
  tibble::column_to_rownames('sample')

HTGTS.cols <- intersect(colnames(HTGTS.data),selected.cols.mouse)

tmp <- HTGTS.data %>%
  gather(variable, value, HTGTS.cols) %>%
  dplyr::group_by(variable, celltype) %>%
  dplyr::mutate(value=scale(value)) %>%
  dplyr::group_by(celltype, genotype, variable) %>%
  dplyr::summarise(mean=mean(value),
                   se=sd(value)/sqrt(n())) %>%
  tidyr::pivot_wider(id_cols=c('variable'),names_from=c('celltype','genotype'),values_from=c('mean','se')) %>%
  dplyr::left_join(columns,by=c('variable'='col')) %>%
  dplyr::group_by(variable) %>%
  dplyr::mutate(variable=factor(variable, levels=columns %>% 
                                  dplyr::filter(col %in% HTGTS.cols) %>% 
                                  dplyr::pull(col)),
                var_col=cat_cols[category]) %>%
  dplyr::arrange(variable)

cols_shapes <- tmp %>% 
  dplyr::distinct(category, variable) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)],
                variable_name=gsub('_downsampled','',variable)) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(variable=factor(variable, levels=levels(tmp$variable))) %>% 
  dplyr::arrange(variable)

comparisons <- list(c('CH12','WT','Atm'),
                    c('CH12','WT','Trp53bp1'),
                    c('CH12','WT','Lig4'),
                    c('spleen','WT','Atm'),
                    c('spleen','WT','Trp53bp1'))

plots <- lapply(comparisons, function(x) 
  tmp %>%
    dplyr::mutate_(mean1=paste0('mean_',x[[1]],'_',x[[2]]),
                   mean2=paste0('mean_',x[[1]],'_',x[[3]]),
                   se1=paste0('se_',x[[1]],'_',x[[2]]),
                   se2=paste0('se_',x[[1]],'_',x[[3]])) %>%
    dplyr::mutate(tstat=abs(mean2-mean1)/(sqrt(se2**2+se1**2)),
                  pval=dt(tstat,4)) %>%
    ggplot(aes(x=mean1,y=mean2)) +
    geom_abline(slope=1,intercept=0,color='gray',linewidth=.5) +
    geom_point(aes(color=variable,shape=variable,size=ifelse(pval < .05,'sig','ns'))) + 
    stat_cor(method = "pearson", aes(label=..r.label..), size=2.5, show.legend=FALSE, label.y.npc = 'bottom') +
    geom_errorbar(aes(ymin=mean2-se2,
                      ymax=mean2+se2,
                      color=variable),
                  size=.25) +
    geom_errorbarh(aes(xmin=mean1-se1,
                       xmax=mean1+se1,
                       color=variable),
                   size=.25) +
    my_theme() +
    labs(color='',shape='') +
    scale_color_manual(values=cols_shapes %>% 
                         dplyr::pull(var_col,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable_name)) +
    scale_shape_manual(values=cols_shapes %>%
                         dplyr::pull(var_ind,variable),
                       labels=cols_shapes %>%
                         dplyr::pull(variable_name)) +
    scale_size_manual(values=c('sig'=2,'ns'=.5),guide='none') +
    guides(color=guide_legend(ncol=4),
           shape=guide_legend(ncol=4)) +
    labs(x=x[[2]],y=x[[3]],title=x[[1]],fill='',color='',size=''))

plot_grid(plot_grid(plotlist=lapply(plots, function(x) x + theme(legend.position='none')), 
                    ncol=3, align='vh'),
          get_legend(plots[[1]] + theme(legend.margin=margin(0,0,0,0,'mm'))), ncol=1, rel_heights=c(.65,.35))
```

```{r FigS3C, fig.width=3.75 * fig_scale, fig.height=2.5 * fig_scale}
library(gtools)

cols.here.HTGTS <- c('n_homology_switch',
                     'frac_breaks_inversions',
                     'cluster_entropy',
                     'mean_cluster_size',
                     'acceptor_score_W',
                     'donor_score_W')

HTGTS.data %>%
  dplyr::mutate(celltype=factor(celltype, levels=c('CH12','spleen')),
                genotype=factor(genotype, levels=c('WT','Atm','Lig4','Trp53bp1'))) %>%
  gather(variable, value, cols.here.HTGTS) %>%
  dplyr::select(variable, value, genotype, celltype) %>%
  dplyr::left_join(columns, by=c('variable'='col')) %>%
  dplyr::mutate(variable=factor(gsub('_downsampled|_switch','',variable), 
                                levels=gsub('_downsampled|_switch','',cols.here.HTGTS))) %>%
  dplyr::filter(category %in% c('diversity','isotypes','breaks','context','structural')) %>%
  ggplot(aes(x=genotype, y=value, fill=genotype)) + 
  geom_boxplot(position=position_dodge(width=.75), width=.7, outlier.shape=NA, lwd=.25) + 
  geom_point(position=position_jitterdodge(jitter.width=.2, dodge.width=.75, jitter.height = 0), size=.5) + 
  my_theme() +
  coord_cartesian(clip='off') +
  stat_compare_means(ref.group='WT',method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2) +
  ggh4x::facet_grid2(celltype~variable, switch='y', scales='free_y', independent='y', strip=ggh4x::strip_vanilla(clip='off')) +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='none') +
  scale_fill_manual(values=genotype_colors) +
  #scale_x_discrete(breaks=c('lymph node','splenocytes','WT','Xlf','Shld1','Shld1_Xlf'),
  #                 limits=c('lymph node','splenocytes','','WT','Xlf','Shld1','Shld1_Xlf'), drop=FALSE) +
  labs(x='',y='', fill='') 
```

show homology values for HTGTS data

```{r FigS3D, fig.width=3.25 * fig_scale, fig.height=1.25 * fig_scale}
read.csv('data/HTGTS_homology.csv',header=1, row.names=1) %>%
  dplyr::mutate(genotype=factor(genotype, levels=c('WT','Atm','Lig4','Trp53bp1'))) %>%
  dplyr::filter(xor(isotype=='Sa', celltype=='spleen')) %>%
  dplyr::group_by( celltype, genotype, isotype, n_homology) %>%
  dplyr::summarise(mean_frac=mean(frac),
                   se_frac=sd(frac)/sqrt(n())) %>%
  ggplot(aes(x=n_homology, y=mean_frac, color=genotype)) +
  geom_line(aes(group=genotype)) +
  geom_errorbar(aes(ymin=mean_frac-se_frac,ymax=mean_frac+se_frac),width=0) +
  ggh4x::facet_nested(. ~ celltype  + isotype , scales='free', space='free_x', nest_line = TRUE, switch='y') +
  my_theme() +
  scale_color_manual(values=genotype_colors) +
  labs(color='', y='fraction junctions') +
  coord_cartesian(clip='off') 
```

```{r FigS3B, fig.width=2.5 * fig_scale, fig.height=3. * fig_scale}
HTGTS.CH12.X <- HTGTS.data[HTGTS.data$celltype=='CH12',HTGTS.cols] %>%
  dplyr::mutate_if(is.numeric, scale) 
HTGTS.spleen.X <- HTGTS.data[HTGTS.data$celltype=='spleen',HTGTS.cols] %>%
  dplyr::mutate_if(is.numeric, scale) 

HTGTS.meta <- HTGTS.data %>%
  dplyr::mutate(genotype=factor(genotype,levels=c('WT','Atm','Lig4','Trp53bp1'))) %>%
  dplyr::select(celltype,genotype)

pca.CH12 <- prcomp(HTGTS.CH12.X)
pcadat.CH12 <- cbind(pca.CH12$x,HTGTS.meta[row.names(HTGTS.CH12.X),])
pca.spleen <- prcomp(HTGTS.spleen.X)
pcadat.spleen <- cbind(pca.spleen$x,HTGTS.meta[row.names(HTGTS.spleen.X),])

p1 <- ggplot(pcadat.CH12,aes(x=PC1,y=PC2, color=genotype)) +
  geom_point(size=1) +
  geom_polygon(data = pcadat.CH12 %>%
                 dplyr::group_by(genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=genotype)) +
  my_theme() +
  theme(aspect.ratio=pca.CH12$sdev[2]**2/pca.CH12$sdev[1]**2,
        legend.position=c(.92,.3),
        legend.background = element_blank()) +
  labs(x=paste0('PC 1 (',100*signif(pca.CH12$sdev[1]**2/sum(pca.CH12$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.CH12$sdev[2]**2/sum(pca.CH12$sdev**2),3),'%)'),
       color='',shape='',fill='', title='CH12')  +
  scale_color_manual(values=genotype_colors) +
  scale_fill_manual(values=genotype_colors)

p2 <- ggplot(pcadat.spleen,aes(x=PC1,y=PC2, color=genotype)) +
  geom_point(size=1) +
  geom_polygon(data = pcadat.spleen %>%
                 dplyr::group_by(genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=genotype)) +
  my_theme() +
  theme(aspect.ratio=pca.spleen$sdev[2]**2/pca.spleen$sdev[1]**2,
        legend.position='none',
        legend.background = element_blank()) +
  labs(x=paste0('PC 1 (',100*signif(pca.spleen$sdev[1]**2/sum(pca.spleen$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.spleen$sdev[2]**2/sum(pca.spleen$sdev**2),3),'%)'),
       color='',shape='',fill='', title='spleen')  +
  scale_color_manual(values=genotype_colors) +
  scale_fill_manual(values=genotype_colors)

plot_grid(p1,p2, ncol=1, align='vh')
```

these PCAs use `r dim(HTGTS.spleen.X)[2]` (spleen) and `r dim(HTGTS.CH12.X)[2]` (CH12) features

show a PCA of spleen and lymph node data together with PP and CH12 using all switch regions or only Sm+Sa

```{r Fig3B, fig.width=3.5 * fig_scale, fig.height=1.75 * fig_scale}
CH12.data <- read.csv('data/CH12_data.csv',header=1, row.names=1) %>%
  tibble::column_to_rownames('sample')
CH12.data.noSg <- read.csv('data/CH12_data_noSg.csv', header=1, row.names=1) %>%
  tibble::column_to_rownames('sample')
mouse.data.noSg <- read.csv('data/mouse_data_noSg.csv', header=1, row.names=1)

mouse.cols <-  intersect(selected.cols.mouse, intersect(colnames(CH12.data),
                                                        colnames(mouse.data))) 

CH12.take <- CH12.data$batch != '20240830'
mouse.take <- rep(TRUE, dim(mouse.data)[1]) 
mouse.X <- rbind(CH12.data[CH12.take,mouse.cols],
                 mouse.data[mouse.take,mouse.cols]) %>%
  dplyr::mutate_if(is.numeric, scale)
mouse.cols <- mouse.cols[colSums(is.na(mouse.X))==0]

pca <- prcomp(mouse.X[,mouse.cols])
pcadat <- cbind(pca$x,data.frame(material=c(rep('CH12',sum(CH12.take)),
                                            mouse.data$material[mouse.take]),
                                 batch=c(CH12.data$batch[CH12.take],
                                         mouse.data$batch[mouse.take]),
                                 genotype=c(CH12.data$genotype[CH12.take],
                                            rep('WT',sum(mouse.take)))))

p1 <- ggplot(pcadat,aes(x=PC1,y=PC2, color=material)) +
  geom_point(aes(shape=genotype),size=1) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(material) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=material)) +
  my_theme() +
  scale_fill_manual(values=tissue_colors) +
  scale_color_manual(values=tissue_colors) +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='', title='all switch regions') 

CH12.take.noSg <- rep(TRUE, dim(CH12.data.noSg)[1]) 
mouse.take.noSg <- rep(TRUE, dim(mouse.data.noSg)[1]) 
mouse.cols.noSg <-  grep('SG|Sg|sequential',mouse.cols,
                         value=TRUE,invert=TRUE)
mouse.X.noSg <- rbind(CH12.data.noSg[CH12.take.noSg,mouse.cols.noSg],
                     mouse.data.noSg[mouse.take.noSg,mouse.cols.noSg]) %>%
  dplyr::mutate_if(is.numeric, scale)

mouse.cols.noSg <- mouse.cols.noSg[colSums(is.na(mouse.X.noSg))==0]

pca.noSg <- prcomp(mouse.X.noSg[,mouse.cols.noSg])
pcadat.noSg <- cbind(pca.noSg$x,data.frame(material=c(rep('CH12',sum(CH12.take.noSg)),
                                                      mouse.data.noSg$material[mouse.take.noSg]),
                                           batch=c(CH12.data.noSg$batch[CH12.take.noSg],
                                                   mouse.data.noSg$batch[mouse.take.noSg]),
                                           genotype=c(CH12.data.noSg$genotype[CH12.take.noSg],
                                                      rep('WT',sum(mouse.take.noSg)))))

p2 <- ggplot(pcadat.noSg,aes(x=PC1,y=PC2, color=material)) +
  geom_point(aes(shape=genotype), size=1) +
  geom_polygon(data = pcadat.noSg %>%
                 dplyr::group_by(material) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=material)) +
  my_theme() +
  scale_fill_manual(values=tissue_colors) +
  scale_color_manual(values=tissue_colors) +
  labs(x=paste0('PC 1 (',100*signif(pca.noSg$sdev[1]**2/sum(pca.noSg$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.noSg$sdev[2]**2/sum(pca.noSg$sdev**2),3),'%)'),
       color='',shape='',fill='', title='only Sm + Sa') 

plot_grid(p1 + theme(legend.position='none'),
          p2 + theme(legend.position='none'),
          get_legend(p1 + theme(legend.box='vertical')),
          rel_widths=c(.4,.4,.2),
          ncol=3)

# p3 <- ggplot(pcadat,aes(x=PC3,y=PC4, color=material)) +
#   geom_point(aes(shape=genotype),size=1) +
#   geom_polygon(data = pcadat %>%
#                  dplyr::group_by(material) %>%
#                  dplyr::slice(chull(PC3, PC4)), alpha = 0.25, aes(fill=material)) +
#   scale_fill_manual(values=tissue_colors) +
#   scale_color_manual(values=tissue_colors) +
#   my_theme() +
#   labs(x=paste0('PC 3 (',100*signif(pca$sdev[3]**2/sum(pca$sdev**2),3),'%)'),
#        y=paste0('PC 4 (',100*signif(pca$sdev[4]**2/sum(pca$sdev**2),3),'%)'),
#        color='',shape='',fill='', title='all switch regions') 
# 
# p4 <- ggplot(pcadat.noSg,aes(x=PC3,y=PC4, color=material)) +
#   geom_point(aes(shape=genotype),size=1) +
#   geom_polygon(data = pcadat.noSg %>%
#                  dplyr::group_by(material) %>%
#                  dplyr::slice(chull(PC3, PC4)), alpha = 0.25, aes(fill=material)) +
#   my_theme() +
#   scale_fill_manual(values=tissue_colors) +
#   scale_color_manual(values=tissue_colors) +
#   labs(x=paste0('PC 3 (',100*signif(pca.noSg$sdev[3]**2/sum(pca.noSg$sdev**2),3),'%)'),
#        y=paste0('PC 4 (',100*signif(pca.noSg$sdev[4]**2/sum(pca.noSg$sdev**2),3),'%)'),
#        color='',shape='',fill='', title='only Sm + Sa') 
# 
# plot_grid(p3 + theme(legend.position='none'),
#           p4 + theme(legend.position='none'),
#           get_legend(p3 + theme(legend.box='vertical')),
#           rel_widths=c(.4,.4,.2),
#           ncol=3)
```

(this PCA uses `r dim(mouse.X)[2]` and `r dim(mouse.X.noSg)[2]` features)

```{r FigS3X1, fig.width=4 * fig_scale, fig.height=1.75 * fig_scale, include=FALSE, eval=FALSE}
dualPCA <- function(X, Y) {
  library(irlba)
  # use CCA-like extension of dual PCA as described here: https://xinmingtu.cn/blog/2022/CCA_dual_PCA/
  #sv <- irlba(as.matrix(X) %*% as.matrix(t(Y)), nv=n)
  sv <- svd(as.matrix(X) %*% as.matrix(t(Y)))
  x <- sv$u %*% diag(sqrt(sv$d))
  row.names(x) <- row.names(X)
  colnames(x) <- paste0('dPC',seq(1,dim(x)[2]))
  y <- sv$v %*% diag(sqrt(sv$d))
  row.names(y) <- row.names(Y)
  colnames(y) <- paste0('dPC',seq(1,dim(y)[2]))
  list(x=x, y=y, sdev=sqrt(sv$d/(dim(X)[2]-1)))
}

CH12.X <- CH12.data[CH12.take, mouse.cols] %>% 
  dplyr::mutate_if(is.numeric, scale)
mouse.X <- mouse.data[mouse.take, mouse.cols] %>% 
  dplyr::mutate_if(is.numeric, scale)

dpca <- dualPCA(CH12.X, mouse.X)
dpcadat <- cbind(rbind(dpca$x, dpca$y),
                 data.frame(group=c(CH12.data[row.names(dpca$x),'genotype'],
                                    mouse.data$material)))

ggplot(dpcadat,aes(x=dPC1,y=dPC2, color=group)) +
  geom_point(size=1) +
  geom_polygon(data = dpcadat %>%
                 dplyr::group_by(group) %>%
                 dplyr::slice(chull(dPC1, dPC2)), alpha = 0.1, aes(fill=group), linewidth=.1) +
  my_theme() +
  scale_fill_manual(values=c(genotype_colors,tissue_colors)) +
  scale_color_manual(values=c(genotype_colors,tissue_colors)) +
  theme(aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2,
        legend.spacing=unit(0,'mm')) +
  labs(x=paste0('dPC 1 (',100*signif(dpca$sdev[1]**2/sum(dpca$sdev**2),3),'%)'),
       y=paste0('dPC 2 (',100*signif(dpca$sdev[2]**2/sum(dpca$sdev**2),3),'%)'),
       color='',shape='',fill='')
```


```{r FigS3X2, fig.width=3 * fig_scale, fig.height=4 * fig_scale, include=FALSE, eval=FALSE}
tmp <- t(t(t(as.matrix(rbind(CH12.X, mouse.X))) %*% as.matrix(dpcadat[,c('dPC1','dPC2')])) / colSums(dpcadat[,c('dPC1','dPC2')]**2)) %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(col=factor(col, levels=columns %>% 
                             dplyr::filter(col %in% mouse.cols) %>%
                             dplyr::pull(col)),
                col_name=gsub("_downsampled|_switch",'',col))

cols_shapes <- tmp %>% 
  dplyr::distinct(category, col,col_name) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(col=factor(col, levels=levels(tmp$col))) %>% 
  dplyr::arrange(col)

ggplot(tmp,aes(x=dPC1,y=dPC2, color=col, shape=col)) +
  geom_point(size=1) +
  #geom_text_repel(aes(label=col), size=1) +
  my_theme() +
  scale_color_manual(values=cols_shapes %>%
                       dplyr::pull(var_col,col),
                     labels=cols_shapes %>%
                       dplyr::pull(col_name)) +
  scale_shape_manual(values=cols_shapes %>%
                       dplyr::pull(var_ind,col),
                     labels=cols_shapes %>%
                       dplyr::pull(col_name)) +
  theme(legend.key.size=unit(1.5,'mm'),
        aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2,
        legend.spacing=unit(0,'mm'),
        legend.key.spacing.y=unit(0,'mm'),
        legend.position='bottom',
        legend.box='horizontal') +
  guides(color=guide_legend(ncol=2),
         shape=guide_legend(ncol=2)) +
  labs(x='dPC 1', y='dPC 2',
       color='',shape='',fill='',alpha='')
```

selected features in the CH12 data (2023 batches)

```{r Fig3D,fig.width=3.5 * fig_scale,fig.height=2.5 * fig_scale}
CH12.features <- c('frac_breaks_single',
                   'mean_length',
                   'frac_breaks_duplications',
                   'frac_breaks_inversions',
                   'insert_frequency',
                   'mean_cluster_size_downsampled')

CH12.genotypes <- c('WT','Trp53bp1 KO','Rif1 KO','Lig4 KO','Brca1 KO')

CH12.meta <- CH12.data %>%
  dplyr::mutate(batch=as.character(batch),
                genotype=factor(genotype,levels=CH12.genotypes)) %>%
  dplyr::select(batch,genotype)

tmp <- CH12.data %>% 
  dplyr::filter(batch != '20240830') %>%
  dplyr::select(genotype,CH12.features) %>%
  dplyr::mutate(genotype=factor(genotype,levels=CH12.genotypes)) %>%
  dplyr::group_by(genotype) %>%
  dplyr::mutate(ng=n(),
                genotype_new=paste0(genotype,' (n=',ng,')')) %>%
  gather(col,value,-c(genotype,ng,genotype_new)) %>% 
  dplyr::mutate(col=factor(gsub("_downsampled|_switch",'',col),
                           levels=gsub("_downsampled|_switch","",CH12.features)))

ggplot(tmp, aes(x=genotype,y=value,fill=genotype)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1,height=0,size=.5) + 
  stat_compare_means(ref.group='WT',method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2) +
  facet_wrap(~col,ncol=3,scales='free_y') +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        strip.clip='off',
        axis.text.x=element_text(angle=90,hjust=1,vjust=.5)) +
  scale_fill_manual(values=genotype_colors) +
  scale_x_discrete(breaks=tmp$genotype,labels=tmp$genotype_new) +
  labs(x='',y='', color='# reads')
```

a PCA with all selected features
  
```{r Fig3E, fig.width=2. * fig_scale, fig.height=1.7 * fig_scale}
CH12.samples <- row.names(CH12.data)[CH12.data$batch != '20240830']
CH12.cols <- setdiff(grep('SG|Sg',intersect(selected.cols.mouse,colnames(CH12.data)),value=TRUE,invert=TRUE),
                     c('frac_breaks_sequential','frac_clusters_Sm'))
CH12.X <- CH12.data[CH12.samples,CH12.cols] %>%
  dplyr::mutate_if(is.numeric, scale)
CH12.cols <- CH12.cols[colSums(is.na(CH12.X))==0]

pca <- prcomp(CH12.X[CH12.samples,CH12.cols])
pcadat <- cbind(pca$x,CH12.meta[CH12.samples,]) 

ggplot(pcadat,aes(x=PC1,y=PC2, color=genotype)) +
  geom_point(size=2) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(genotype) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=genotype)) +
  my_theme() +
  theme(aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position='none') +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='') +
  scale_fill_manual(values=genotype_colors) +
  scale_color_manual(values=genotype_colors) 
```

(this PCA uses `r dim(CH12.X)[2]` features)

heatmap of all selected features

```{r Fig3C, fig.width=3.5 * fig_scale, fig.height=3.75 * fig_scale}
breaks <- seq(-2,2,.2)
samples <- CH12.meta %>%
  tibble::rownames_to_column('sample') %>%
  dplyr::filter(batch != '20240830') %>%
  dplyr::arrange(genotype) %>%
  dplyr::pull(sample)
colnames(CH12.X) <- gsub('_switch|_downsampled','',colnames(CH12.X))

dend <- hclust(dist(CH12.X[samples,]))
dend <- reorder(as.dendrogram(dend),wts=order(match(CH12.meta %>%
                                                      tibble::rownames_to_column('sample') %>%
                                                      dplyr::mutate(genotype=factor(genotype, levels=unique(genotype))) %>%
                                                      dplyr::arrange(genotype) %>%
                                                      dplyr::pull(sample),
                                                    samples)),agglo.FUN=mean)

hm1 <- Heatmap(t(CH12.X[samples,gsub('_switch|_downsampled','',CH12.cols)]),
               col=colorRamp2(c(-2,0, 2), c("blue",'gray90', "red")),
               cluster_columns=dend,
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=FALSE,
               show_row_dend=TRUE,
               column_dend_height = unit(5, "mm"),
               top_annotation=HeatmapAnnotation(df=CH12.meta[samples,c('genotype','batch')],
                                                show_annotation_name=TRUE,
                                                show_legend=c(TRUE,FALSE),
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_name_gp=gpar(fontsize=8),
                                                annotation_legend_param=list(labels_gp=gpar(fontsize=7),
                                                                             title_gp=gpar(fontsize=8),
                                                                             legend_width=unit(2,'mm'),
                                                                             legend_height=unit(2,'mm'),
                                                                             legend_direction="horizontal",
                                                                             nrow=1),
                                                col=list('genotype'=genotype_colors[levels(CH12.meta$genotype)],
                                                         'batch'=batch_colors[unique(CH12.meta[samples,'batch'])])),
               column_names_gp=gpar(fontsize=7),
               row_names_gp =gpar(fontsize = 7),
               row_split = columns %>% tibble::column_to_rownames('col') %>% .[CH12.cols, 'category'], 
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=8),
                                         legend_direction = "horizontal", 
                                         legend_width=unit(20,'mm'),
                                         legend_height=unit(2,'mm')),
               row_title=NULL,
               name='z-score')

draw(hm1, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")#, legend_grouping="original")

```

do ridge regression

```{r Fig3G, fig.width=1.25 * fig_scale, fig.height=3.5 * fig_scale}
train_batches <- c('20220425','20221118','20221212','20230120','20230213')
test_batches <- '20240830'
train <- CH12.meta$batch %in% train_batches
test <- CH12.meta$batch %in% test_batches

CH12.X <- CH12.data[, CH12.cols] %>%
  # dplyr::mutate_if(is.numeric, function(x) (x-mean(x[train]))/sd(x[train])) 
  dplyr::mutate_if(is.numeric, function(x) { 
    s <- numeric(sum(train | test)); 
    s[train] <- (x[train]-mean(x[train]))/sd(x[train]);
    s[test] <- (x[test]-mean(x[test]))/sd(x[test]);
    s
  }) 
#CH12.X.test <- CH12.data[test, CH12.cols] %>%
#  dplyr::mutate_if(is.numeric, scale) #function(x) (x-mean(x[train]))/sd(x[train]))

set.seed(3)
cv_model <- cv.glmnet(as.matrix(CH12.X[train,]), CH12.meta$genotype[train], alpha = 0, family='multinomial')

mod <- glmnet(as.matrix(CH12.X[train,]), CH12.meta$genotype[train], alpha = 0, family = "multinomial",
              lambda = cv_model$lambda.min)

tmp <- as.data.frame(predict(mod, type = "response", 
                              newx=as.matrix(CH12.X[test,]))) %>%
  tibble::rownames_to_column('sample') %>% 
  gather(gt,p,-sample) %>% 
  dplyr::mutate(prediction=factor(gsub('.s0','',gt),levels=CH12.genotypes),
                p=as.numeric(p)) %>%
  dplyr::left_join(CH12.meta[test,] %>% 
                     tibble::rownames_to_column('sample') %>% 
                     dplyr::select(sample,genotype),by='sample') %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(c=ifelse(p == max(p),ifelse(genotype==prediction, 'green','red'),'black'),
                genotype=factor(genotype,levels=CH12.genotypes),
                sample=gsub('[0-9]*_(BC[0-9]*)_switch','\\1',sample)) %>%
  dplyr::ungroup() %>%
  dplyr::arrange(genotype) %>%
  dplyr::mutate(sample=factor(gsub('20240830_','',sample),
                              levels=gsub('20240830_','',unique(sample)))) 
ggplot(tmp, aes(x=prediction,y=sample,
                label=round(p,2),fill=p,color=c)) + 
  geom_tile() + 
  geom_text(data=. %>% dplyr::filter(p > .25),
            size=2) +
  scale_fill_gradient(low='white',high='steelblue3') +
  scale_color_manual(values=c(black='black',green='darkgreen',red='red3')) +
  my_theme() +
  scale_y_discrete(labels=ifelse(grepl('\\.20|\\.27|\\.14|\\.01|\\.04',tmp$sample),"*","")) +
  theme(legend.position='none',
        axis.ticks.y = element_blank(),
        #axis.text.y = element_blank(),
        axis.text.x=element_text(angle=90, hjust=1, vjust=.5)) +
  labs(x='',y='36 new samples (* 25ng IL4)')
```

show the ridge coefficients 

```{r Fig3F, fig.width=2. * fig_scale, fig.height=2.5 * fig_scale}
tmp <- do.call(cbind,lapply(coef(mod), function(x) x[,'s0'])) %>%
  as.data.frame() %>%
  dplyr::filter(rowSums(abs(.)) > 0) %>%
  dplyr::slice_max(rowSums(abs(.)),n=20) %>%
  dplyr::arrange(rowSums(abs(.))) %>%
  tibble::rownames_to_column('col') %>%
  dplyr::filter(col %in% columns$col) 

col.order <- gsub('_downsampled|_switch','',tmp$col[order.hclust(hclust(dist(tmp[,2:6])))])

tmp <- tmp %>%
  dplyr::mutate(col=gsub('_downsampled|_switch','',col),
                col=factor(col, levels=col.order)) 

anno_grob_y <- grid::rectGrob(y=1:length(tmp$col), x=0, gp=gpar(col=NA, fill=col_cols[rev(levels(tmp$col))], alpha=.5))

tmp %>%
  gather(genotype,coef,-col) %>%
  dplyr::mutate(genotype=factor(genotype, levels=CH12.genotypes),
                col=factor(col, levels=rev(levels(tmp$col)))) %>%
  ggplot(aes(x=genotype, y=col, fill=coef, label=round(coef,2))) + 
  geom_tile() +
  scale_fill_gradient2(low='pink2',high='steelblue3', mid='white', breaks=c(-.3, 0, .3)) +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position=c(-1,-.15),
        legend.text=element_text(angle=90, hjust=.5, vjust=1),
        legend.direction='horizontal') +
  annotation_custom(grob=anno_grob_y, ymin = 0, ymax = 1, xmin = -4, xmax = 4.5) +
  labs(x='',y='') 
```

evaluate ridge regression by cross-validation

```{r FigS3X5, fig.width=1. * fig_scale, fig.height=1.75 * fig_scale, include=FALSE}
batch.CV <- c()
for (batch in unique(CH12.meta$batch)) {
  train <- CH12.meta$batch != batch
  test <- CH12.meta$batch == batch
  X <- CH12.data[, CH12.cols] %>%
    dplyr::mutate_if(is.numeric, function(x) { 
      s <- numeric(sum(train | test)); 
      s[train] <- (x[train]-mean(x[train]))/sd(x[train]);
      s[test] <- (x[test]-mean(x[test]))/sd(x[test]);
      s
      }) %>%
    as.matrix()
  
  cv_model <- cv.glmnet(X[train,], CH12.meta$genotype[train], alpha = 0, family='multinomial')
  mod <- glmnet(X[train,], CH12.meta$genotype[train], alpha = 0, family = "multinomial",
                lambda = cv_model$lambda.min)
  batch.CV[[batch]] <- mean(predict(mod, type = "class", newx=X[test,])[,'s0'] == CH12.meta[test,'genotype'])
}  

set.seed(4)
sample.CV <- c()
for (k in seq(1,10)) {
  train <- runif(dim(CH12.meta)[1]) < .75
  test <- !train
  X <- CH12.data[, CH12.cols] %>%
    dplyr::mutate_if(is.numeric, function(x) { 
      s <- numeric(sum(train | test)); 
      s[train] <- (x[train]-mean(x[train]))/sd(x[train]);
      s[test] <- (x[test]-mean(x[test]))/sd(x[test]);
      s
      }) %>%
    as.matrix()

  cv_model <- cv.glmnet(X[train,], CH12.meta$genotype[train], alpha = 0, family='multinomial')
  mod <- glmnet(X[train,], CH12.meta$genotype[train], alpha = 0, family = "multinomial",
                lambda = cv_model$lambda.min)
  sample.CV[[as.character(k)]] <- mean(predict(mod, type = "class", newx=X[test,])[,'s0'] == CH12.meta[test,'genotype'])
}

data.frame(acc=c(unlist(batch.CV),unlist(sample.CV)),
           fold=c(names(batch.CV),names(sample.CV))) %>%
  dplyr::filter(!is.na(acc)) %>%
  dplyr::mutate(method=ifelse(fold %in% CH12.meta$batch,'batch','sample')) %>%
  dplyr::group_by(method) %>%
  dplyr::mutate(mean_acc=mean(acc)) %>%
  ggplot(aes(x=method,y=acc)) +
  geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1, height=0) + 
  scale_y_continuous(limits=c(0,1)) +
  my_theme() +
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=.5)) +
  labs(x='',y='accuracy')
```

(average accuracy is `r mean(unlist(batch.CV))` for cross-validation across batches and `r mean(unlist(sample.CV))` for cross-validation across samples)


# Figure 4 + supplement (FB data)

explained variance for FB cohort (make FB_13/21 control)

```{r Fig4B,fig.width=1 * fig_scale, fig.height=1.7 * fig_scale}
FB.data <- read.csv('data/FB_data.csv', header=1, row.names=1) %>%
  dplyr::mutate(batch=as.character(batch),
                diagnosis = ifelse(donor %in% c('FB_13','FB_21'), 'control', diagnosis)) %>%
  tibble::column_to_rownames('sample')
FB.X <- FB.data[grepl('2023',row.names(FB.data)),intersect(colnames(FB.data),columns$col)] %>%
  dplyr::mutate_if(is.numeric, scale)
FB.meta <- FB.data %>%
  dplyr::filter(grepl('2023',batch)) %>%
  dplyr::select(donor,batch,sex,age,diagnosis,timepoint)

FB.var <- fitExtractVarPartModel(t(FB.X[,colSums(is.na(FB.X))==0]), 
                                 ~ (1 | donor) + (1 | diagnosis) + (1 | sex) + age + (1 | batch), FB.meta) %>%
  as.data.frame() %>%
  tibble::rownames_to_column('col') %>%
  gather(variable,variance,-col)

FB.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable,levels=c('donor','diagnosis','sex','age','batch','Residuals'))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('library','isotypes','diversity','structural','context','breaks','variants')) %>%
  ggplot(aes(x=variable, y=100*variance)) + 
  geom_boxplot(aes(fill=variable), outlier.shape=NA, lwd=.5, legend=FALSE) + 
  geom_jitter(position=position_jitter(width=.25), size=.25) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        strip.clip='off',
        legend.position='none',
        legend.box='vertical') +
  scale_fill_manual(values=var_colors, guide='none') +
  labs(x='', y='% explained variance', fill='', color='')
```

```{r FigS4A,fig.width=3.5 * fig_scale,fig.height=1.7 * fig_scale}
FB.var %>%  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(variable=factor(variable, levels=c('donor','diagnosis','sex','age','batch','Residuals'))) %>%
  dplyr::filter(variable!='Residuals', category %in% c('library','isotypes','diversity','structural','context','breaks','variants')) %>%
  ggplot(aes(x=category, y=100*variance, fill=variable)) + 
  geom_boxplot(position=position_dodge(width=.8), outlier.shape=NA, lwd=.5) + 
  geom_jitter(position=position_jitterdodge(jitter.width=.3,dodge.width=.8), size=.5) +
  my_theme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
        legend.position='top',
        legend.box='horizontal') +
  guides(fill=guide_legend(override.aes = list(shape=22))) +
  scale_fill_manual(values=var_colors) +
  labs(x='', y='% explained variance', fill='')
```

a PCA with all selected features on pooled FB data from 2023
 
```{r Fig4D, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale} 
FB.data.pooled <- read.csv('data/FB_data_pooled.csv', header=1, row.names=1) %>%
  dplyr::mutate(num_variants_somatic = num_variants * (1 - frac_variants_germline),
                diagnosis = ifelse(donor %in% c('FB_13','FB_21'), 'control', diagnosis),
                status=ifelse(diagnosis %in% c('control','subclass'), 'control', 'CVID')) %>%
  tibble::column_to_rownames('sample')
QP.data <- read.csv('data/QP_data.csv', header=1, row.names=1) %>%
  tibble::column_to_rownames('sample')

FB_train_donors <- paste0('FB_',sprintf('%0.2d', 1:35))
human.meta <- rbind(FB.data.pooled %>%
                      dplyr::mutate(dataset=ifelse(grepl('FB',donor),
                                                   ifelse(sample_donor %in% FB_train_donors,'training','testing'),
                                                   'validation')) %>%
                      dplyr::select(donor,diagnosis,status,timepoint, dataset,sex,age,complication, EUROclass, pct_B, pct_IgA, pct_IgG),
                    QP.data %>%
                      dplyr::mutate(dataset='validation',
                                    timepoint='t1',
                                    complication=NA,
                                    EUROclass=NA,
                                    pct_B=NA,
                                    pct_IgA=NA,
                                    pct_IgG=NA,
                                    status=ifelse(diagnosis=='control','control','CVID')) %>%
                      dplyr::select(donor,diagnosis,status,timepoint,dataset,sex,age,complication, EUROclass, pct_B, pct_IgA, pct_IgG)) %>%
  dplyr::mutate(diagnosis=factor(diagnosis,levels=c('control','subclass','CVID','ICOS','Kabuki','IgG4def',
                                                    'AID','TNFRSF13B',
                                                    'NFKB2','PI3K','ATM','BRCA1','LIG4','NIPBL')),
                status=factor(status, levels=c('control','CVID')))

human.X <- cbind(rbind(FB.data.pooled[,selected.cols],
                       QP.data[,selected.cols]),
                 human.meta[,'dataset',drop=FALSE]) %>%
  tibble::rownames_to_column('sample') %>%
  dplyr::mutate(sample=gsub('[0-9]*_|pooled_','',sample)) %>%
  dplyr::group_by(dataset) %>%
  dplyr::mutate_if(is.numeric,scale) %>%
  dplyr::ungroup() %>%
  dplyr::select(c('sample',selected.cols)) %>%
  tibble::column_to_rownames('sample')

train <- human.meta$dataset=='training'
test <- human.meta$dataset=='testing'
validation <- human.meta$dataset=='validation'
pca <- prcomp(human.X[train,])
pcadat <- cbind(pca$x,human.meta[train,])

ggplot(pcadat,aes(x=PC1,y=PC2)) +
  geom_point(aes(color=diagnosis),size=1) +
  geom_line(aes(group=donor, color=diagnosis)) +
  geom_polygon(data = pcadat %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.1, aes(fill=status), legend=FALSE) +
  my_theme() +
  #geom_label_repel(aes(label=gsub('FB_','',donor)),size=2, segment.size=.25, label.padding = unit(.5,'mm'), force=20, show.legend=FALSE) +
  scale_fill_manual(values=diagnosis_colors, guide="none") +
  scale_color_manual(values=diagnosis_colors) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=3)) +
  coord_cartesian(xlim=c(-8,10),
                  ylim=c(-5,8),
                  clip='off') +
  theme(aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.key.size=unit(1.5,'mm'),
        legend.key.spacing.y=unit(0,'mm'),
        legend.position='bottom',
        legend.box='horizontal') +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='')
```

(this PCA uses `r dim(human.X)[2]` features)

another PCA with only subclass features

```{r FigS4C, fig.width=2.75 * fig_scale, fig.height=1.75 * fig_scale}
FB.cols.subclass <- grep('frac_clusters_S[MGA]+',colnames(FB.data.pooled),value=TRUE)
FB.X.subclass <- FB.data.pooled[FB.data.pooled$sample_donor %in% FB_train_donors,FB.cols.subclass] %>%
  dplyr::mutate_if(is.numeric,scale) 

pca.subclass <- prcomp(FB.X.subclass)
pcadat.subclass <- cbind(pca.subclass$x,human.meta[human.meta$dataset=='training',])

ggplot(pcadat.subclass,aes(x=PC1,y=PC2)) +
  geom_point(aes(color=diagnosis),size=1) +
  geom_line(aes(group=donor, color=diagnosis)) +
  geom_polygon(data = pcadat.subclass %>%
                 dplyr::mutate(status=ifelse(diagnosis=='subclass', 'subclass', as.character(status))) %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.25, aes(fill=status)) +
  my_theme() +
  #geom_label_repel(data = pcadat.subclass %>% 
  #                   dplyr::filter(diagnosis!='control'),
  #                 aes(label=gsub('FB_','',donor)),size=2, segment.size=.25, label.padding = unit(.5,'mm'), force=20, show.legend=FALSE) +
  scale_fill_manual(values=diagnosis_colors, guide='none') +
  scale_color_manual(values=diagnosis_colors) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=1)) +
  theme(aspect.ratio=pca.subclass$sdev[2]**2/pca.subclass$sdev[1]**2,
        legend.position='right',
        legend.key.size=unit(1.5,'mm'),
        legend.key.spacing.y=unit(0,'mm'),
        legend.box='horizontal') +
  labs(x=paste0('PC 1 (',100*signif(pca.subclass$sdev[1]**2/sum(pca.subclass$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca.subclass$sdev[2]**2/sum(pca.subclass$sdev**2),3),'%)'),
       color='',shape='',fill='')
```

(this PCA uses `r dim(FB.X.subclass)[2]` features)

show subclass features for HD and subclass-deficient people

```{r FigS4B, fig.width=2 * fig_scale, fig.height=1.7 * fig_scale}
subclass.cols <- c("frac_clusters_SA1","frac_clusters_SA2",
                   "frac_clusters_SG1","frac_clusters_SG2","frac_clusters_SG3","frac_clusters_SG4")

tmp <- FB.data.pooled %>%
  dplyr::filter(sample_donor %in% FB_train_donors, diagnosis %in% c('control','subclass')) %>%
  gather(feature, frac_clusters, subclass.cols) %>%
  dplyr::mutate(feature=factor(gsub('frac_clusters_','',feature),
                               levels=gsub('frac_clusters_','',subclass.cols)))

pvals <- tmp %>%
  dplyr::group_by(feature) %>%
  dplyr::do(wc=wilcox.test(frac_clusters ~ diagnosis, data=.),
            y=max(.$frac_clusters)) %>%
  dplyr::summarise(feature, p.value=wc$p.value, y)

ggplot(tmp, aes(x=feature, y=frac_clusters)) +
  geom_boxplot(aes(color=diagnosis), position=position_dodge(width=.8),width=.7,outlier.shape=NA) +
  geom_point(aes(color=diagnosis), position=position_dodge(width=.8), size=.5) +
  geom_line(data=tmp %>% dplyr::filter(diagnosis=='control'),
            aes(group=donor), position=position_nudge(x=-.2), alpha=.25, size=.25) +
  geom_line(data=tmp %>% dplyr::filter(diagnosis=='subclass'),
            aes(group=donor), position=position_nudge(x=+.2), alpha=.25, size=.25) +
  geom_text(data=pvals, aes(label=stars.pval(p.value), y=1.1*y)) +
  scale_color_manual(values=diagnosis_colors) +
  my_theme() +
  coord_cartesian(clip='off') +
  theme(legend.position=c(.8,.8)) +
  labs(color='',x='', y='fraction clusters')
```

correlate FACS against SWIBRID

```{r FigS4H, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale}
tmp <- FB.data.pooled %>%
  dplyr::mutate(alpha_ratio_FACS=pct_IgA/(pct_IgA + pct_IgG)) %>%
  dplyr::group_by(status) %>%
  dplyr::mutate(status_R=paste0(status," (R=",signif(cor(alpha_ratio_FACS,alpha_ratio_clusters,use='na.or.complete'),2),")"))
ggplot(tmp, aes(x=alpha_ratio_FACS, y=alpha_ratio_clusters, color=status)) + 
  geom_abline(slope=1, intercept=0, size=.5) +
  #stat_cor(method = "pearson", aes(label=..r.label..), size=2.5, show.legend=FALSE, label.x=.75, label.y=c(.4, .33)) +
  geom_point() + 
  my_theme() +
  scale_color_manual(values=diagnosis_colors, breaks=unique(tmp$status), labels=unique(tmp$status_R)) +
  labs(x='FACS', y='SWIBRID', title='IgA / (IgA + IgG)') +
  theme(legend.position=c(.8,.15),
        legend.box=element_blank())
```


box plots for selected features

```{r Fig4C, fig.width=4 * fig_scale, fig.height=2.7 * fig_scale}
FB.features <- c('frac_clusters_SG2',
                 'frac_clusters_SG3',
                 'mean_length',
                 'frac_breaks_single',
                 'n_homology_switch',
                 'frac_breaks_inversions',
                 'mean_cluster_size_downsampled',
                 'num_variants_somatic')

tmp <- FB.data.pooled[FB.data.pooled$sample_donor %in% FB_train_donors,] %>% 
  dplyr::select(donor,diagnosis,FB.features) %>%
  dplyr::mutate(diagnosis_simple=factor(ifelse(diagnosis %in% c('CVID','control','subclass'), diagnosis, 'other'), 
                                        levels=c('control','subclass','CVID','other'))) %>%
  dplyr::group_by(diagnosis_simple) %>%
  dplyr::mutate(ng=n_distinct(donor),
                diagnosis_new=paste0(diagnosis_simple,' (n=',ng,')')) %>%
  gather(col,value,-c(diagnosis_simple,ng,diagnosis_new,donor,diagnosis)) %>% 
  dplyr::mutate(col=factor(gsub("_downsampled|_switch",'',col),
                           levels=gsub("_downsampled|_switch","",FB.features))) 

ggplot(tmp, aes(x=diagnosis_simple,y=value,fill=diagnosis_simple)) + 
  geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1,height=0,size=.5) + 
  stat_compare_means(ref.group='control',method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2) +
  facet_wrap(~col,ncol=4,scales='free_y') +
  my_theme() +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=diagnosis_colors) +
  theme(legend.position='none',
        axis.text.x=element_text(angle=90,hjust=1,vjust=.5)) +
  scale_x_discrete(breaks=tmp$diagnosis_simple,labels=tmp$diagnosis_new) +
  labs(x='',y='', color='# reads')
```

project test data into the training data PCA

```{r Fig4E, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale}
pcadat.combined <- rbind(cbind(as.matrix(human.X[train,]) %*% pca$rotation, human.meta[train,]),
                         cbind(as.matrix(human.X[test,]) %*% pca$rotation, human.meta[test,]),
                         cbind(as.matrix(human.X[validation,]) %*% pca$rotation, human.meta[validation,])) %>%
  dplyr::mutate(dataset=factor(dataset, levels=c('training','testing','validation')))

ggplot(pcadat.combined %>% dplyr::filter(dataset!='validation'),
       aes(x=PC1,y=PC2)) +
  geom_polygon(data = pcadat.combined %>% dplyr::filter(dataset=='training') %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) +
  geom_point(aes(color=diagnosis,alpha=dataset, shape=dataset), size=1) +
  geom_line(aes(group=donor, color=diagnosis,alpha=dataset, shape=dataset)) +
  #geom_label_repel(data=pcadat.combined %>% dplyr::filter(dataset=='testing') %>% dplyr::group_by(donor) %>% dplyr::slice_head(),
  #                 aes(label=gsub("FB_","",donor)),size=2, segment.size=.25, label.padding = unit(.5,'mm'), show.legend=FALSE) +
  my_theme() +
  scale_fill_manual(values=diagnosis_colors, guide='none') +
  scale_color_manual(values=diagnosis_colors) +
  scale_alpha_manual(values=c('testing'=1,'training'=.25)) +
  scale_shape_manual(values=c('testing'=15,'training'=16)) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=3),
         shape=guide_legend(ncol=1),
         alpha=guide_legend(ncol=1)) +
  theme(legend.key.height=unit(1.5,'mm'),
        legend.key.spacing=unit(0,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position='bottom',
        legend.box='horizontal') +
  coord_cartesian(xlim=c(-8,10),
                  ylim=c(-5,8),
                  clip='off') +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='',alpha='')
```

project validation data into training PCA

```{r Fig4F, fig.width=2.25 * fig_scale, fig.height=2 * fig_scale}
ggplot(pcadat.combined %>% dplyr::filter(dataset!='testing'),
       aes(x=PC1,y=PC2)) +
  geom_polygon(data = pcadat.combined %>% dplyr::filter(dataset=='training') %>%
                 dplyr::group_by(status) %>%
                 dplyr::slice(chull(PC1, PC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) +
  geom_point(aes(color=diagnosis,alpha=dataset, shape=dataset), size=1) +
  geom_line(aes(group=donor, color=diagnosis,alpha=dataset, shape=dataset)) +
  # geom_label_repel(data=pcadat.combined %>% dplyr::filter(dataset=='validation') %>% dplyr::group_by(donor) %>% dplyr::slice_head(),
  #                 aes(label=donor),size=2, segment.size=.25, label.padding = unit(.5,'mm'), show.legend=FALSE) +
  my_theme() +
  scale_fill_manual(values=diagnosis_colors, guide='none') +
  scale_color_manual(values=diagnosis_colors) +
  scale_alpha_manual(values=c('training'=.25,'validation'=1)) +
  scale_shape_manual(values=c('training'=16,'validation'=15)) +
  guides(color=guide_legend(override.aes = list(shape=15,size=2), ncol=3),
         shape=guide_legend(ncol=1),
         alpha=guide_legend(ncol=1)) +
  theme(legend.key.height=unit(1.5,'mm'),
        legend.key.spacing=unit(0,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.position='bottom',
        legend.box='horizontal') +
  coord_cartesian(xlim=c(-8,10),
                  ylim=c(-5,8),
                  clip='off') +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='',shape='',fill='',alpha='')
```

do ridge regression from training to testing or validation data (take out subclass def.)

```{r Fig4H, fig.width=2. * fig_scale, fig.height=1.5 * fig_scale}
set.seed(1)
train.here <- train & human.meta[train,'diagnosis'] != 'subclass'
cv_model <- cv.glmnet(as.matrix(human.X[train.here,]), human.meta[train.here,'status']!='control', alpha = 0, family='binomial')
mod <- glmnet(as.matrix(human.X[train.here,]), human.meta[train.here,'status']!='control', alpha = 0, family = "binomial",
                lambda = cv_model$lambda.min)
pred_1 <- predict(mod, type = "response", newx=as.matrix(human.X[test,]))
ROC_1 <- roc(human.meta[test,'status']!='control', pred_1)

pred_2 <- predict(mod, type = "response", newx=as.matrix(human.X[validation,]))
ROC_2 <- roc(human.meta[validation,'status']!='control', pred_2)

FB_res <- data.frame(specificity=c(ROC_1$specificities,ROC_2$specificities),
                     sensitivity=c(ROC_1$sensitivities,ROC_2$sensitivities),
                     thresholds=c(ROC_1$thresholds,ROC_2$thresholds),
                     model=c(rep(paste0('training -> testing; AUC=',signif(ROC_1$auc,3)),length(ROC_1$specificities)),
                             rep(paste0('training -> validation; AUC=',signif(ROC_2$auc,3)),length(ROC_2$specificities)))) %>%
  dplyr::mutate(model=factor(model, levels=unique(model))) %>%
  dplyr::arrange(model,sensitivity) 

ggplot(FB_res, aes(x=specificity,y=sensitivity,color=model)) + 
  geom_step() +
  scale_x_reverse() +  
  geom_abline(slope=1,intercept=1,linewidth=.25,linetype='dashed') +
  my_theme() +
  guides(color=guide_legend(ncol=1)) +
  scale_color_manual(values=c('black','blue3')) +
  theme(legend.position=c(.5,.2)) +
  labs(color='',linetype='')
```

evaluate ridge regression by cross-validation

```{r FigS4X1, fig.width=.8 * fig_scale, fig.height=1 * fig_scale, include=FALSE}
set.seed(4)
sample.CV <- c()
FB.samples <- row.names(FB.data.pooled)[FB.data.pooled$diagnosis != 'subclass']
for (k in seq(1,20)) {
  tr <- runif(length(FB.samples)) < .75
  te <- !tr
  X <- FB.data.pooled[FB.samples, selected.cols] %>%
    dplyr::mutate_if(is.numeric, function(x) {
      s <- numeric(sum(tr | te));
      s[tr] <- (x[tr]-mean(x[tr]))/sd(x[tr]);
      s[te] <- (x[te]-mean(x[te]))/sd(x[te]);
      s
      }) %>%
    #dplyr::mutate_if(is.numeric, scale) %>%
    as.matrix()
  
  y <- FB.data.pooled[FB.samples, 'status'] != 'control'
  cv_model <- cv.glmnet(X[tr,], y[tr], alpha = 0, family='binomial')
  mod <- glmnet(X[tr,], y[tr], alpha = 0, family = "binomial",lambda = cv_model$lambda.min)
  sample.CV[[as.character(k)]] <- mean(predict(mod, type = "class", newx=X[te,])[,'s0'] == y[te])
}

data.frame(acc=unlist(sample.CV),
           fold=names(sample.CV)) %>%
  dplyr::filter(!is.na(acc)) %>%
  dplyr::mutate(mean_acc=mean(acc)) %>%
  ggplot(aes(x=1,y=acc)) +
  #geom_boxplot(outlier.shape=NA, lwd=.5) + 
  geom_jitter(width=.1, height=0, size=.25) + 
  scale_y_continuous(limits=c(0,1)) +
  scale_x_continuous(limits=c(.85,1.15)) +
  my_theme() +
  theme(axis.text.x=element_blank(),
        axis.ticks = element_blank()) +
  labs(x='',y='accuracy')
```

(average accuracy is `r mean(unlist(sample.CV))`)

heatmap

```{r Fig4G, fig.width=5.5 * fig_scale, fig.height=5.5 * fig_scale}
FB.samples <- row.names(FB.data.pooled)[grepl('FB',FB.data.pooled$donor)]
FB.X <- FB.data.pooled[FB.samples, selected.cols] %>%
  dplyr::mutate_if(is.numeric, scale)

row.names(FB.X) <- FB.data.pooled[FB.samples, c('donor','timepoint')] %>% 
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(donor,'_',timepoint)))) %>% 
  dplyr::pull(name)
colnames(FB.X) <- gsub('_downsampled|_switch','',colnames(FB.X))
pca_hm <- pcadat.combined[gsub('pooled_(FB)_([0-9]*)','\\1\\2',FB.samples), c('PC1','PC2','dataset','donor')]
row.names(pca_hm) <- row.names(FB.X)

FB.meta <- FB.data.pooled[FB.samples,] %>%
  dplyr::mutate(sample_donor=gsub("FB_","FB",sample_donor),
                dataset=ifelse(grepl('FB',donor),
                               ifelse(sample_donor %in% FB_train_donors,'training','testing'),
                               'validation')) %>%
  dplyr::select(donor,sample_donor,diagnosis,dataset,sex,age,timepoint,
                splenomegaly,lymphoproliferation,pct_B,pct_IgA,pct_IgG,complication)

dend <- hclust(dist(FB.X))
# dend <- reorder(as.dendrogram(dend),wts=order(match(FB.meta %>%
#                                                       dplyr::arrange(diagnosis) %>%
#                                                       dplyr::mutate(sample_donor=gsub('pooled_FB_','',sample_donor)) %>%
#                                                       dplyr::pull(sample_donor),
#                                                     row.names(FB.X))),agglo.FUN=mean)

class_df <- data.frame(class=plyr::revalue(as.character(cutree(dend, k = 4)),
                                           c('2'='cluster3',
                                             '3'='cluster1',
                                             '1'='cluster2',
                                             '4'='cluster2')),
                       name=row.names(FB.X))

class_colors <- c('cluster3'=diagnosis_colors[['control']],
                  'cluster1'=cat_cols[['structural']],
                  'cluster2'=cat_cols[['diversity']])

hm1 <- Heatmap(t(FB.X),
               col=colorRamp2(c(-2,0, 2), c("blue",'gray90', "red")),
               cluster_columns=dend,
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               top_annotation=HeatmapAnnotation(df=FB.meta[,c('diagnosis','sex','age',#'pct_B','pct_IgA','pct_IgG',
                                                              #'splenomegaly','lymphoproliferation',
                                                              #'lung','PnPS','Tetanus','Diphteria',
                                                              #'complication','EUROclass',
                                                              'donor','timepoint'),drop=FALSE] %>% 
                                                  tibble::remove_rownames() %>%
                                                  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(donor,'_',timepoint)))) %>%
                                                  dplyr::left_join(class_df,by='name') %>%
                                                  dplyr::select(-c(donor,timepoint)) %>%
                                                  tibble::column_to_rownames('name'),
                                                show_annotation_name=TRUE,
                                                show_legend=TRUE,
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_legend_param=list(title_gp=gpar(fontsize=7),
                                                                             labels_gp=gpar(fontsize=7),
                                                                             grid_height = unit(3, "mm"),
                                                                             grid_width = unit(3, "mm")),
                                                annotation_name_gp=gpar(fontsize=8),
                                                col=list('diagnosis'=diagnosis_colors,
                                                         'dataset'=c('training'='gray60','testing'='gray20',
                                                                     'Charite'='red3','QP'='forestgreen'),
                                                         'class'=class_colors,
                                                         'EUROclass'=c('smB+21norm'='gray20',
                                                                       'smB+21low'='orange3',
                                                                       'smB-21norm'='lightblue3',
                                                                       'smB-21low'='darkseagreen2'),
                                                         'complication'=c('c'='coral3',
                                                                          'io'='olivedrab4'),
                                                         'age'=colorRamp2(c(20,50,80),c("yellow2","orange2","coral3")),
                                                         'sex'=c('male'='mistyrose2',
                                                                 'female'='lightseagreen'))),
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize = 7),
               row_dend_width = unit(7, "mm"),
               row_split = columns %>% tibble::column_to_rownames('col') %>% .[selected.cols, 'category'], 
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=8),
                                         legend_height=unit(10,'mm')),
               row_title=NULL,
               name='z-score')  

hm3 <- Heatmap(t(pca_hm[,c('PC1','PC2')]),
               cluster_columns=dend,
               cluster_rows=TRUE,
               col=colorRamp2(c(-5, 0, 5), c("darkorchid3","gray90","darkseagreen3")),
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize=7),
               row_dend_width = unit(7, "mm"),
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=8),
                                         legend_height = unit(10, "mm")),
               row_title=NULL,
               name='PC')

hm1 %v% hm3
```

show PCA loadings per category

```{r FigS4D, fig.width=3 * fig_scale, fig.height=4 * fig_scale}
tmp <- data.frame(pca$rotation) %>%
  tibble::rownames_to_column('col') %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::mutate(col=factor(col, levels=columns %>% 
                             dplyr::filter(col %in% selected.cols) %>% 
                             dplyr::pull(col)),
                col_name=gsub("_downsampled|_switch",'',col))

cols_shapes <- tmp %>% 
  dplyr::distinct(category, col,col_name) %>%
  dplyr::mutate(var_col=cat_cols[as.character(category)]) %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(var_ind=1:n()) %>%
  dplyr::mutate(col=factor(col, levels=levels(tmp$col))) %>% 
  dplyr::arrange(col)

ggplot(tmp,aes(x=abs(PC1),y=abs(PC2), color=col, shape=col)) +
  geom_point(size=1) +
  #geom_text_repel(aes(label=col), size=1) +
  my_theme() +
  scale_color_manual(values=cols_shapes %>%
                       dplyr::pull(var_col,col),
                     labels=cols_shapes %>%
                       dplyr::pull(col_name)) +
  scale_shape_manual(values=cols_shapes %>%
                       dplyr::pull(var_ind,col),
                     labels=cols_shapes %>%
                       dplyr::pull(col_name)) +
  theme(legend.key.size=unit(1.5,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
        legend.spacing=unit(0,'mm'),
        legend.key.spacing.y=unit(0,'mm'),
        legend.position='bottom',
        legend.box='horizontal') +
  guides(color=guide_legend(ncol=2),
         shape=guide_legend(ncol=2)) +
  labs(x='abs. PC 1', y='abs. PC 2',
       color='',shape='',fill='',alpha='')

# ggplot(tmp,aes(x=PC1,y=PC2, color=col, shape=col)) +
#   geom_point(size=1) +
#   #geom_text_repel(aes(label=col), size=1) +
#   my_theme() +
#   scale_color_manual(values=cols_shapes %>%
#                        dplyr::pull(var_col,col),
#                      labels=cols_shapes %>%
#                        dplyr::pull(col_name)) +
#   scale_shape_manual(values=cols_shapes %>%
#                        dplyr::pull(var_ind,col),
#                      labels=cols_shapes %>%
#                        dplyr::pull(col_name)) +
#   theme(legend.key.size=unit(1.5,'mm'),
#         aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2,
#         legend.spacing=unit(0,'mm'),
#         legend.key.spacing.y=unit(0,'mm'),
#         legend.position='bottom',
#         legend.box='horizontal') +
#   guides(color=guide_legend(ncol=2),
#          shape=guide_legend(ncol=2)) +
#   labs(x='PC 1', y='PC 2',
#        color='',shape='',fill='',alpha='')
```

categorize the samples

```{r FigS4G, fig.width=2.5 * fig_scale, fig.height=2 * fig_scale}
ggplot(pca_hm %>% tibble::rownames_to_column('name') %>%
         dplyr::filter(dataset!='validation') %>%
         dplyr::left_join(class_df, by='name'), 
       aes(x=PC1,y=PC2, color=class, shape=dataset)) +
  geom_point(size=1) +
  geom_point(data=pcadat.combined %>% dplyr::filter(dataset=='validation'),
             aes(color=diagnosis), size=1) +
  geom_line(aes(group=donor)) +
  my_theme() +
  scale_shape_manual(values=c('testing'=16,'training'=16,'validation'=15), guide='none') +
  scale_color_manual(values=c(class_colors,diagnosis_colors), 
                     breaks=c(names(class_colors),
                              names(diagnosis_colors))) +
  guides(color=guide_legend(override.aes = list(shape=c(rep(16,3),rep(15,7)),size=2), ncol=4)) +
  theme(legend.key.size=unit(1.5,'mm'),
        legend.position='bottom',
        legend.key.spacing.y=unit(0,'mm'),
        aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2) +
  labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'),
       y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'),
       color='')
```

box plots for selected features

```{r Fig4J, fig.width=5.25 * fig_scale, fig.height=1.7 * fig_scale}
features.here <- c('frac_clusters_SG1',
                   'frac_clusters_SG2',
                   'mean_cluster_size_downsampled',
                   'n_homology_switch',
                   'n_untemplated_switch',
                   'frac_breaks_intra',
                   'frac_breaks_duplications')

groups <- c('cluster1','cluster2','cluster3','control','BRCA1','ATM')

tmp <- cbind(rbind(FB.data.pooled[,features.here],
                   QP.data[,features.here]),
             human.meta[,c('diagnosis','dataset','donor','timepoint')]) %>%
  tibble::rownames_to_column('sample_donor') %>%
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(donor,'_',timepoint)))) %>%
  dplyr::left_join(class_df,by='name') %>%
  dplyr::mutate(group=ifelse(dataset=='validation',as.character(diagnosis),as.character(class))) %>%
  dplyr::filter(group %in% groups) %>%
  dplyr::mutate(group=factor(group,levels=groups)) %>%
  dplyr::group_by(group) %>%
  dplyr::mutate(ng=n_distinct(donor),
                group_new=paste0(group,' (n=',ng,')')) %>%
  gather(col,value,features.here) %>%
  dplyr::mutate(col=factor(gsub("_downsampled|_switch",'',col),
                           levels=gsub("_downsampled|_switch","",features.here)))

ggplot(tmp, aes(x=group,y=value,fill=group)) +
  geom_boxplot(outlier.shape=NA, lwd=.5) +
  geom_jitter(width=.1,height=0,size=.5) +
  stat_compare_means(comparisons=list(c('cluster1','cluster3'),
                                      c('cluster1','cluster2'),
                                      c('cluster2','cluster3')),
                     method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2, vjust=.5) +
  stat_compare_means(comparisons=list(c('control','ATM'),
                                      c('control','BRCA1')),
                     method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2, vjust=.5) +
  facet_wrap(.~col,ncol=length(features.here),scales='free_y') +
  my_theme() +
  coord_cartesian(clip='off') +
  scale_fill_manual(values=c(diagnosis_colors,class_colors)) +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5),
        legend.position='none') +
  scale_x_discrete(breaks=c('cluster1','cluster2','cluster3','control','ATM','BRCA1'),
                   labels=c('cluster1','cluster2','cluster3','control','ATM','BRCA1')) +
  labs(x='',y='', color='diagnosis')
```

check association of clinics with features

```{r FigS4E, fig.width=3 * fig_scale, fig.height=2.5 * fig_scale}
pvals <- list()
for (x in c("Tetanus",
            "Diphteria",
            "PnPS")) {
  for (f in selected.cols) {
    data <- FB.data.pooled[FB.samples,c(x,f)] %>%
      tidyr::drop_na()
    form <- as.formula(paste0(x,'=="negative" ~ ',f))
    res <- glm(form, family='binomial', data=data) %>% summary()
    pvals[[paste0(x,'_',f)]] <- data.frame(pval=coef(res)[2,4],
                                           vacc=x,
                                           col=f)
  }
}

FB.data.pooled[FB.samples, ] %>%
  dplyr::mutate(Tetanus=ifelse(status=='control','positive',Tetanus),
                Diphteria=ifelse(status=='control','positive',Diphteria),
                PnPS=ifelse(status=='control','positive',PnPS)) %>%
  gather(feature, value, c('frac_clusters_SA1','frac_clusters_SG3')) %>%
  gather(vacc, vacc_status, c('Tetanus','Diphteria','PnPS')) %>%
  dplyr::mutate(vacc_status=factor(ifelse(vacc_status=='','no data',vacc_status),
                                   c('negative','positive','no data'))) %>%
  ggplot(aes(x=vacc_status, y=value, fill=vacc_status)) +
  geom_boxplot(outlier.shape=NA) +
  geom_jitter(aes(color=status),width=.1, size=.5) +
  stat_compare_means(comparisons=list(c('negative','positive'),
                                      c('positive','no data')),
                     method='wilcox.test', label = "p.signif", hide.ns=TRUE, size=2, vjust=0) +
  facet_grid(feature ~ vacc, switch='y') + 
  my_theme() +
  scale_fill_manual(values=c('negative'='red3','positive'='gray')) +
  scale_color_manual(values=diagnosis_colors) +
  labs(x='',y='') +
  coord_cartesian(clip='off') +
  theme(legend.position='none',
        strip.clip='off',
        axis.text.x=element_text(angle=45, hjust=1, vjust=1))

# for (x in c("splenomegaly",
#             "lymphoproliferation",
#             "lung")) {
#   for (f in selected.cols) {
#     data <- FB.data.pooled[FB.samples,c(x,f,'status')] %>%
#       dplyr::filter(status!='control') %>%
#       tidyr::drop_na()
#     form <- as.formula(paste0(x,'!="" ~ ',f))
#     res <- glm(form, family='binomial', data=data) %>% summary()
#     pvals[[paste0(x,'_',f)]] <- data.frame(pval=coef(res)[2,4],
#                                            vacc=x,
#                                            col=f)
#   }
# }

# FB.data.pooled[FB.samples, ] %>%
#   gather(feature, value, c('frac_clusters_SG3','alpha_ratio_clusters')) %>%
#   gather(disease, disease_status, c('splenomegaly','lymphoproliferation','lung')) %>%
#   dplyr::mutate(disease_status=factor(ifelse(status=='control','control',
#                                              ifelse(disease_status=='','absent','present')),
#                                       levels=c('control','absent','present'))) %>%
#   ggplot(aes(x=disease_status, y=value, fill=disease_status)) +
#   geom_boxplot(outlier.shape=NA) +
#   geom_jitter(aes(color=status),width=.1, size=.5) +
#   stat_compare_means(comparisons=list(c('absent','present'),
#                                       c('absent','control')),
#                      method='wilcox.test', label = "p.signif", hide.ns=TRUE, size=2, vjust=0) +
#   facet_grid(feature ~ disease, switch='y', scales='free_y') + 
#   my_theme() +
#   scale_fill_manual(values=c('present'='red3','absent'='gray')) +
#   scale_color_manual(values=diagnosis_colors) +
#   labs(x='',y='') +
#   coord_cartesian(clip='off') +
#   theme(legend.position='none',
#         strip.clip='off',
#         axis.text.x=element_text(angle=45, hjust=1, vjust=1))
```

check aggregated scores

```{r FigS4F, fig.width=5.5 * fig_scale, fig.height=3.5 * fig_scale}
FB.X.aggregated <-  FB.data.pooled[FB.samples, selected.cols] %>%
  dplyr::mutate_if(is.numeric, scale) %>%
  tibble::rownames_to_column('sample') %>%
  gather(col, value, -sample) %>%
  dplyr::left_join(columns, by='col') %>%
  dplyr::group_by(sample, category) %>%
  dplyr::summarise(value=mean(abs(value))) %>%
  dplyr::left_join(FB.meta %>% tibble::rownames_to_column('sample'), by='sample') %>%
  #dplyr::filter(diagnosis!='control') %>%
  dplyr::group_by(category) %>%
  dplyr::mutate(#value=(value-mean(value[diagnosis=='control'])/sd(value[diagnosis=='control'])),
                name=gsub('t','',gsub('FB_','',paste0(donor,'_',timepoint)))) %>%
  dplyr::select(name,category,value) %>%
  spread(category,value) %>%
  tibble::column_to_rownames('name')

FB.meta.aggregated <- FB.meta[,c('diagnosis','sex','age',
                                 'donor','timepoint'),drop=FALSE] %>% 
  dplyr::filter(!(diagnosis %in% c('control','subclass'))) %>%
  tibble::remove_rownames() %>%
  dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(donor,'_',timepoint)))) %>%
  dplyr::left_join(class_df,by='name') %>%
  dplyr::select(-c(donor,timepoint)) %>%
  tibble::column_to_rownames('name')

samples <- row.names(FB.meta.aggregated)
dend <- hclust(dist(FB.X.aggregated[samples,]))
dend <- reorder(as.dendrogram(dend),wts=order(match(FB.meta.aggregated %>%
                                                      tibble::rownames_to_column('sample') %>%
                                                      dplyr::arrange(class) %>%
                                                      dplyr::pull(sample),
                                                    samples)),agglo.FUN=mean)

hm2 <- Heatmap(t(FB.X.aggregated[samples,]),
               cluster_columns=dend,
               col=colorRamp2(c(0, 2), c("lightyellow2","darkorchid3")),
               cluster_rows=TRUE,
               show_row_names=TRUE,
               show_column_names=TRUE,
               show_row_dend=TRUE,
               top_annotation=HeatmapAnnotation(df=FB.meta.aggregated,
                                                show_annotation_name=TRUE,
                                                show_legend=TRUE,
                                                simple_anno_size=unit(2,'mm'),
                                                annotation_legend_param=list(title_gp=gpar(fontsize=7),
                                                                             labels_gp=gpar(fontsize=7),
                                                                             grid_height = unit(3, "mm"),
                                                                             grid_width = unit(3, "mm")),
                                                annotation_name_gp=gpar(fontsize=8),
                                                col=list('diagnosis'=diagnosis_colors,
                                                         'dataset'=c('training'='gray60','testing'='gray20',
                                                                     'Charite'='red3','QP'='forestgreen'),
                                                         'class'=class_colors,
                                                         'EUROclass'=c('smB+21norm'='gray20',
                                                                       'smB+21low'='orange3',
                                                                       'smB-21norm'='lightblue3',
                                                                       'smB-21low'='darkseagreen2'),
                                                         'complication'=c('c'='coral3',
                                                                          'io'='olivedrab4'),
                                                         'age'=colorRamp2(c(20,50,80),c("yellow2","orange2","coral3")),
                                                         'sex'=c('male'='mistyrose2',
                                                                 'female'='lightseagreen'))),
               column_names_gp=gpar(fontsize=4),
               row_names_gp =gpar(fontsize = 7),
               row_dend_width = unit(7, "mm"),
               heatmap_legend_param=list(labels_gp=gpar(fontsize=7),
                                         title_gp=gpar(fontsize=8),
                                         legend_height=unit(10,'mm')),
               row_title=NULL,
               name='category score')  

hm2
```

<!-- ```{r FigS4X2, fig.width=4 * fig_scale, fig.height=1.75 * fig_scale, include=FALSE, eval=FALSE} -->
<!-- all.cols <- intersect(colnames(CH12.X), colnames(human.X)) -->
<!-- dpca <- dualPCA(CH12.X[,all.cols], human.X[,all.cols]) -->

<!-- dpcadat <- cbind(rbind(dpca$x,dpca$y), -->
<!--                 data.frame(species=c(rep('mouse',dim(dpca$x)[1]), -->
<!--                                      rep('human',dim(dpca$y)[1])), -->
<!--                            genotype=c(CH12.meta[,'genotype'], -->
<!--                                       human.meta[,'diagnosis']))) -->

<!-- ggplot(dpcadat,aes(x=dPC1,y=dPC2, color=genotype)) + -->
<!--   geom_point(aes(shape=species),size=.5) + -->
<!--   geom_polygon(data = dpcadat %>% -->
<!--                  dplyr::group_by(genotype) %>% -->
<!--                  dplyr::slice(chull(dPC1, dPC2)), alpha = 0.1, aes(fill=genotype), linewidth=.1) + -->
<!--   my_theme() + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,genotype_colors)) + -->
<!--   scale_color_manual(values=c(diagnosis_colors,genotype_colors)) + -->
<!--   scale_shape_manual(values=c('human'=15,'mouse'=16)) + -->
<!--   guides(fill=guide_legend(ncol=2, override.aes = list(pch=c(rep(16,5),rep(15,14)))), -->
<!--          color=guide_legend(ncol=2, override.aes = list(pch=c(rep(16,5),rep(15,14)))), -->
<!--          shape='none') + -->
<!--   labs(x=paste0('dPC 1 (',100*signif(dpca$sdev[1]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        y=paste0('dPC 2 (',100*signif(dpca$sdev[2]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='') -->

<!-- ggplot(dpcadat,aes(x=dPC1,y=dPC2, color=genotype)) + -->
<!--   geom_point(aes(shape=species),size=.5) + -->
<!--   geom_polygon(data = dpcadat %>% -->
<!--                  dplyr::group_by(genotype) %>% -->
<!--                  dplyr::slice(chull(dPC1, dPC2)), alpha = 0.1, aes(fill=genotype), linewidth=.1) + -->
<!--   my_theme() + -->
<!--   facet_wrap(~species, ncol=2) + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,genotype_colors)) + -->
<!--   scale_color_manual(values=c(diagnosis_colors,genotype_colors)) + -->
<!--   scale_shape_manual(values=c('human'=15,'mouse'=16)) + -->
<!--   theme(legend.key.size=unit(1.5,'mm'), -->
<!--         aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2) + -->
<!--   guides(fill=guide_legend(ncol=2, override.aes = list(pch=c(rep(16,5),rep(15,14)))), -->
<!--          color=guide_legend(ncol=2, override.aes = list(pch=c(rep(16,5),rep(15,14)))), -->
<!--          shape='none') + -->
<!--   labs(x=paste0('dPC 1 (',100*signif(dpca$sdev[1]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        y=paste0('dPC 2 (',100*signif(dpca$sdev[2]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='') -->

<!-- dpcadat[,'group'] <- as.character(dpcadat$genotype) -->
<!-- dpcadat[FB.meta$sample_donor,'group'] <- as.character(class_df$class) -->
<!-- dpcadat[,'dataset'] <- 'CH12' -->
<!-- dpcadat[dpcadat$species=='human','dataset'] <- human.meta$dataset -->
<!-- ggplot(dpcadat,# %>% dplyr::filter(dataset!='validation'), -->
<!--        aes(x=dPC1,y=dPC2, color=group)) + -->
<!--   geom_point(aes(shape=species),size=.5) + -->
<!--   geom_polygon(data = dpcadat %>% -->
<!--                  #dplyr::filter(dataset!='validation') %>% -->
<!--                  dplyr::group_by(group) %>% -->
<!--                  dplyr::slice(chull(dPC1, dPC2)), alpha = 0.1, aes(fill=group), linewidth=.1) + -->
<!--   my_theme() + -->
<!--   facet_wrap(~species, ncol=2) + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,genotype_colors,class_colors)) + -->
<!--   scale_color_manual(values=c(diagnosis_colors,genotype_colors,class_colors)) + -->
<!--   scale_shape_manual(values=c('human'=15,'mouse'=16)) + -->
<!--   theme(legend.key.size=unit(1.5,'mm'), -->
<!--         aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2) + -->
<!--   guides(fill=guide_legend(ncol=2), #, override.aes = list(pch=c(rep(16,5),rep(15,10)))), -->
<!--          color=guide_legend(ncol=2), #, override.aes = list(pch=c(rep(16,5),rep(15,10)))), -->
<!--          shape='none') + -->
<!--   labs(x=paste0('dPC 1 (',100*signif(dpca$sdev[1]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        y=paste0('dPC 2 (',100*signif(dpca$sdev[2]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='') -->
<!-- ``` -->

<!-- ```{r FigS4X3, fig.width=3 * fig_scale, fig.height=4 * fig_scale} -->
<!-- tmp <- t(t(t(as.matrix(rbind(CH12.X[,all.cols], human.X[,all.cols]))) %*% as.matrix(dpcadat[,c('dPC1','dPC2')])) / colSums(dpcadat[,c('dPC1','dPC2')]**2)) %>% -->
<!--   as.data.frame() %>% -->
<!--   tibble::rownames_to_column('col') %>% -->
<!--   dplyr::left_join(columns, by='col') %>% -->
<!--   dplyr::mutate(col=factor(col, levels=columns %>%  -->
<!--                              dplyr::filter(col %in% all.cols) %>% -->
<!--                              dplyr::pull(col)), -->
<!--                 col_name=gsub("_downsampled|_switch",'',col)) -->

<!-- cols_shapes <- tmp %>%  -->
<!--   dplyr::distinct(category, col, col_name) %>% -->
<!--   dplyr::mutate(var_col=cat_cols[as.character(category)]) %>% -->
<!--   dplyr::group_by(category) %>% -->
<!--   dplyr::mutate(var_ind=1:n()) %>% -->
<!--   dplyr::mutate(col=factor(col, levels=levels(tmp$col))) %>%  -->
<!--   dplyr::arrange(col) -->

<!-- ggplot(tmp,aes(x=dPC1,y=dPC2, color=col, shape=col)) + -->
<!--   geom_point(size=1) + -->
<!--   #geom_text_repel(aes(label=col), size=1) + -->
<!--   my_theme() + -->
<!--   scale_color_manual(values=cols_shapes %>% -->
<!--                        dplyr::pull(var_col,col), -->
<!--                      labels=cols_shapes %>% -->
<!--                        dplyr::pull(col_name)) + -->
<!--   scale_shape_manual(values=cols_shapes %>% -->
<!--                        dplyr::pull(var_ind,col), -->
<!--                      labels=cols_shapes %>% -->
<!--                        dplyr::pull(col_name)) + -->
<!--   theme(legend.key.size=unit(1.5,'mm'), -->
<!--         aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2, -->
<!--         legend.spacing=unit(0,'mm'), -->
<!--         legend.key.spacing.y=unit(0,'mm'), -->
<!--         legend.position='bottom', -->
<!--         legend.box='horizontal') + -->
<!--   guides(color=guide_legend(ncol=2), -->
<!--          shape=guide_legend(ncol=2)) + -->
<!--   labs(x='dPC 1', y='dPC 2', -->
<!--        color='',shape='',fill='',alpha='') -->
<!-- ``` -->

<!-- ```{r FigS4X4, fig.width=6.75 * fig_scale, fig.height=3.5 * fig_scale} -->
<!-- features.here <- c('cluster_entropy_downsampled', -->
<!--                    'top_clone_occupancy_downsampled', -->
<!--                    'donor_complexity', -->
<!--                    'donor_score_W', -->
<!--                    'donor_score_WGCW', -->
<!--                    'homology_fw', -->
<!--                    'n_homology_switch', -->
<!--                    'mean_length_SA', -->
<!--                    'frac_breaks_intra', -->
<!--                    'frac_breaks_inversions') -->

<!-- tmp <- cbind(rbind(CH12.data[,features.here], -->
<!--                    FB.data.pooled[,features.here], -->
<!--                    QP.data[,features.here]), -->
<!--              rbind(CH12.meta %>%  -->
<!--                      dplyr::mutate(diagnosis=genotype, -->
<!--                                    dataset='CH12') %>% -->
<!--                      dplyr::select(diagnosis, dataset), -->
<!--                    human.meta[,c('diagnosis','dataset')])) %>% -->
<!--   dplyr::filter(diagnosis %in% c('WT','Trp53bp1 KO','Brca1 KO','Rif1 KO','Lig4 KO','control','CVID','TNFRSF13B','ATM','BRCA1')) %>% -->
<!--   tibble::rownames_to_column('sample') %>% -->
<!--   gather(col,value,features.here) %>% -->
<!--   dplyr::mutate(col=factor(gsub("_downsampled|_switch",'',col), -->
<!--                            levels=gsub("_downsampled|_switch","",features.here)), -->
<!--                 group=droplevels(factor(diagnosis, levels=c(names(genotype_colors), names(diagnosis_colors))))) -->

<!-- ggplot(tmp, aes(x=ifelse(dataset=='CH12',as.numeric(group),as.numeric(group) + 1),y=value,fill=group)) + -->
<!--   geom_boxplot(outlier.shape=NA, lwd=.5) + -->
<!--   geom_jitter(width=.1,height=0,size=.1) + -->
<!--   stat_compare_means(comparisons=list(c(1,2), -->
<!--                                       c(1,3), -->
<!--                                       c(1,4), -->
<!--                                       c(1,5)), -->
<!--                      method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2, vjust=.5) + -->
<!--   stat_compare_means(comparisons=list(c(7,8), -->
<!--                                       c(7,9), -->
<!--                                       c(7,10), -->
<!--                                       c(7,11)), -->
<!--                      method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2, vjust=.5) + -->
<!--   facet_wrap(.~col,ncol=5,scales='free_y') + -->
<!--   my_theme() + -->
<!--   coord_cartesian(clip='off') + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,genotype_colors)) + -->
<!--   theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5), -->
<!--         legend.position='none') + -->
<!--   scale_x_continuous(breaks=c(seq(1,5), seq(1,5) + 6), -->
<!--                      labels=levels(tmp$group)) + -->
<!--   labs(x='',y='', color='diagnosis') -->
<!-- ``` -->

<!-- ```{r FigS4X5, fig.width=3.5 * fig_scale, fig.height=3 * fig_scale} -->
<!-- OC_sample_sheet <- read.csv('../sodar/2023_OC/a_2023_OC.txt', -->
<!--                             sep='\t',header=1) %>% -->
<!--   dplyr::left_join(read.csv('../sodar/2023_OC/s_2023_OC.txt',sep='\t',header=1), -->
<!--                    by='Sample.Name') %>% -->
<!--   dplyr::mutate(donor=as.character(Source.Name), -->
<!--                 sample=Extract.Name.1, -->
<!--                 batch=Parameter.Value.Batch., -->
<!--                 QC=Parameter.Value.QC., -->
<!--                 sequencing=Parameter.Value.Sequencing.method.) %>% -->
<!--   dplyr::distinct(donor) -->

<!-- OC.data <-read.csv('../human_samples/20240307_hg38_results/output/summary/20240307_human_samples_stats.csv', -->
<!--                    header=1,row.names=1) %>% -->
<!--   tibble::rownames_to_column('sample') %>% -->
<!--   dplyr::filter(grepl('pooled_',sample)) %>% -->
<!--   dplyr::mutate(sample_donor=gsub('pooled_','',sample)) %>% -->
<!--   dplyr::inner_join(OC_sample_sheet, by=c('sample_donor'='donor')) %>% -->
<!--   dplyr::filter(nreads_final > 500) %>% -->
<!--   dplyr::mutate_all(~replace_na(.,0)) %>% -->
<!--   dplyr::mutate(log10_nreads=log10(nreads_final), -->
<!--                 log10_nclusters=log10(nclusters_final)) %>% -->
<!--   tibble::column_to_rownames('sample_donor') -->

<!-- OC.meta <- OC.data %>% -->
<!--   tibble::rownames_to_column('donor') %>% -->
<!--   dplyr::mutate(status=ifelse(grepl('BIH',sample),'OC','control'), -->
<!--                 sample_donor=donor, -->
<!--                 diagnosis=ifelse(donor %in% c('BIH116','BIH141','BIH152'), 'BRCA1', status), -->
<!--                 timepoint='t1', -->
<!--                 dataset='OC', -->
<!--                 sex=NA, -->
<!--                 age=NA, -->
<!--                 complication=NA, -->
<!--                 EUROclass=NA, -->
<!--                 pct_B=NA, -->
<!--                 pct_IgA=NA, -->
<!--                 pct_IgG=NA) %>% -->
<!--   dplyr::select(sample_donor,donor,diagnosis,status,timepoint,dataset,sex,age,complication,EUROclass,pct_B,pct_IgA,pct_IgG) %>% -->
<!--   tibble::column_to_rownames('sample_donor') -->

<!-- OC.X <- OC.data[row.names(OC.meta), selected.cols] %>% -->
<!--   dplyr::mutate_if(is.numeric, scale) -->

<!-- pcadat.all <- rbind(cbind(as.matrix(human.X[train,]) %*% pca$rotation, human.meta[train,]), -->
<!--                     cbind(as.matrix(human.X[test,]) %*% pca$rotation, human.meta[test,]), -->
<!--                     cbind(as.matrix(human.X[validation,]) %*% pca$rotation, human.meta[validation,]), -->
<!--                     cbind(as.matrix(OC.X) %*% pca$rotation, OC.meta)) %>% -->
<!--   dplyr::mutate(dataset=factor(dataset, levels=c('training','testing','validation','OC'))) -->

<!-- ggplot(pcadat.all, aes(x=PC1,y=PC2)) + -->
<!--   geom_polygon(data = pcadat.all %>% dplyr::filter(dataset=='training') %>% -->
<!--                  dplyr::group_by(status) %>% -->
<!--                  dplyr::slice(chull(PC1, PC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) + -->
<!--   geom_point(aes(color=diagnosis,alpha=dataset, shape=dataset), size=1) + -->
<!--   geom_text_repel(data=pcadat.all %>% dplyr::filter(dataset=='OC'), aes(label=donor,color=diagnosis), size=2, segment.size=.25) + -->
<!--   geom_line(aes(group=donor, color=diagnosis,alpha=dataset, shape=dataset)) + -->
<!--   my_theme() + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,'OC'='deeppink'), guide='none') + -->
<!--   scale_color_manual(values=c(diagnosis_colors,'OC'='deeppink')) + -->
<!--   scale_alpha_manual(values=c('testing'=.25,'training'=.25,'validation'=.5,'OC'=1)) + -->
<!--   scale_shape_manual(values=c('testing'=15,'training'=16,'validation'=17,'OC'=18)) + -->
<!--   guides(color=guide_legend(override.aes = list(shape=15,size=2),ncol=3), -->
<!--          shape=guide_legend(ncol=1), -->
<!--          alpha=guide_legend(ncol=1)) + -->
<!--   theme(legend.key.height=unit(1.5,'mm'), -->
<!--         legend.key.spacing=unit(0,'mm'), -->
<!--         aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2, -->
<!--         legend.position='bottom', -->
<!--         legend.box='horizontal') + -->
<!--   coord_cartesian(xlim=c(-8,10), -->
<!--                   ylim=c(-5,8), -->
<!--                   clip='off') + -->
<!--   labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'), -->
<!--        y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='',alpha='') -->

<!-- ``` -->

<!-- ```{r FigS4X6, fig.width=3.5 * fig_scale, fig.height=3 * fig_scale} -->
<!-- dpca <- dualPCA(human.X, OC.X) -->
<!-- dpcadat <- cbind(rbind(dpca$x,dpca$y), -->
<!--                  pcadat.all) -->

<!-- ggplot(dpcadat, aes(x=dPC1,y=dPC2)) + -->
<!--   geom_polygon(data = dpcadat %>% dplyr::filter(dataset=='training') %>% -->
<!--                  dplyr::group_by(status) %>% -->
<!--                  dplyr::slice(chull(dPC1, dPC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) + -->
<!--   geom_point(aes(color=diagnosis,alpha=dataset, shape=dataset), size=1) + -->
<!--   geom_line(aes(group=donor, color=diagnosis,alpha=dataset, shape=dataset)) + -->
<!--   geom_text_repel(data=dpcadat %>% dplyr::filter(dataset=='OC'), aes(label=donor, color=diagnosis), size=2, segment.size=.25) + -->
<!--   my_theme() + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,'OC'='deeppink'), guide='none') + -->
<!--   scale_color_manual(values=c(diagnosis_colors,'OC'='deeppink')) + -->
<!--   scale_alpha_manual(values=c('testing'=.25,'training'=.25,'validation'=.5,'OC'=1)) + -->
<!--   scale_shape_manual(values=c('testing'=15,'training'=16,'validation'=17,'OC'=18)) + -->
<!--   guides(color=guide_legend(override.aes = list(shape=15,size=2),ncol=3), -->
<!--          shape=guide_legend(ncol=1), -->
<!--          alpha=guide_legend(ncol=1)) + -->
<!--   theme(legend.key.height=unit(1.5,'mm'), -->
<!--         legend.key.spacing=unit(0,'mm'), -->
<!--         #aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2, -->
<!--         legend.position='bottom', -->
<!--         legend.box='horizontal') + -->
<!--   coord_cartesian(clip='off') + -->
<!--   labs(x=paste0('dPC 1 (',100*signif(dpca$sdev[1]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        y=paste0('dPC 2 (',100*signif(dpca$sdev[2]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='',alpha='') -->

<!-- ``` -->

<!-- ```{r FigS4X7, fig.width=3.5 * fig_scale, fig.height=3 * fig_scale} -->
<!-- LB_sample_sheet <- read.csv('../sodar/2023_LB/a_2023_LB.txt', -->
<!--                             sep='\t',header=1) %>% -->
<!--   dplyr::left_join(read.csv('../sodar/2023_LB/s_2023_LB.txt',sep='\t',header=1), -->
<!--                    by='Sample.Name') %>% -->
<!--   dplyr::mutate(donor=as.character(Source.Name), -->
<!--                 sample=Extract.Name.1, -->
<!--                 batch=Parameter.Value.Batch., -->
<!--                 QC=Parameter.Value.QC., -->
<!--                 diagnosis=Characteristics.Diagnosis., -->
<!--                 sequencing=Parameter.Value.Sequencing.method.) %>% -->
<!--   dplyr::distinct(donor,diagnosis) -->

<!-- LB.data <- read.csv('../human_samples/20240307_hg38_results/output/summary/20240307_human_samples_stats.csv', -->
<!--                    header=1,row.names=1) %>% -->
<!--   tibble::rownames_to_column('sample') %>% -->
<!--   dplyr::filter(grepl('pooled_',sample)) %>% -->
<!--   dplyr::mutate(sample_donor=gsub('pooled_','',sample)) %>% -->
<!--   dplyr::inner_join(LB_sample_sheet, by=c('sample_donor'='donor')) %>% -->
<!--   dplyr::filter(nreads_final > 500) %>% -->
<!--   dplyr::mutate_all(~replace_na(.,0)) %>% -->
<!--   dplyr::mutate(log10_nreads=log10(nreads_final), -->
<!--                 log10_nclusters=log10(nclusters_final)) %>% -->
<!--   tibble::column_to_rownames('sample_donor') -->

<!-- LB.meta <- LB.data %>% -->
<!--   tibble::rownames_to_column('donor') %>% -->
<!--   dplyr::mutate(sample_donor=donor, -->
<!--                 status=diagnosis, -->
<!--                 timepoint='t1', -->
<!--                 dataset='LB', -->
<!--                 sex=NA, -->
<!--                 age=NA, -->
<!--                 complication=NA, -->
<!--                 EUROclass=NA, -->
<!--                 pct_B=NA, -->
<!--                 pct_IgA=NA, -->
<!--                 pct_IgG=NA) %>% -->
<!--   dplyr::select(sample_donor,donor,diagnosis,status,timepoint,dataset,sex,age,complication,EUROclass,pct_B,pct_IgA,pct_IgG) %>% -->
<!--   tibble::column_to_rownames('sample_donor') -->

<!-- LB.X <- LB.data[row.names(LB.meta), selected.cols] %>% -->
<!--   dplyr::mutate_if(is.numeric, scale) -->

<!-- pcadat.all <- rbind(cbind(as.matrix(human.X[train,]) %*% pca$rotation, human.meta[train,]), -->
<!--                     cbind(as.matrix(human.X[test,]) %*% pca$rotation, human.meta[test,]), -->
<!--                     cbind(as.matrix(human.X[validation,]) %*% pca$rotation, human.meta[validation,]), -->
<!--                     cbind(as.matrix(LB.X) %*% pca$rotation, LB.meta)) %>% -->
<!--   dplyr::mutate(dataset=factor(dataset, levels=c('training','testing','validation','LB'))) -->

<!-- ggplot(pcadat.all, aes(x=PC1,y=PC2)) + -->
<!--   geom_polygon(data = pcadat.all %>% dplyr::filter(dataset=='training') %>% -->
<!--                  dplyr::group_by(status) %>% -->
<!--                  dplyr::slice(chull(PC1, PC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) + -->
<!--   geom_point(aes(color=diagnosis,alpha=dataset, shape=dataset), size=1) + -->
<!--   geom_text_repel(data=pcadat.all %>% dplyr::filter(dataset=='LB'), aes(label=donor,color=diagnosis), size=2, segment.size=.25) + -->
<!--   geom_line(aes(group=donor, color=diagnosis,alpha=dataset, shape=dataset)) + -->
<!--   my_theme() + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,'unknown'='gray'), guide='none') + -->
<!--   scale_color_manual(values=c(diagnosis_colors,'unknown'='gray')) + -->
<!--   scale_alpha_manual(values=c('testing'=.25,'training'=.25,'validation'=.5,'LB'=1)) + -->
<!--   scale_shape_manual(values=c('testing'=15,'training'=16,'validation'=17,'LB'=18)) + -->
<!--   guides(color=guide_legend(override.aes = list(shape=15,size=2),ncol=3), -->
<!--          shape=guide_legend(ncol=1), -->
<!--          alpha=guide_legend(ncol=1)) + -->
<!--   theme(legend.key.height=unit(1.5,'mm'), -->
<!--         legend.key.spacing=unit(0,'mm'), -->
<!--         aspect.ratio=pca$sdev[2]**2/pca$sdev[1]**2, -->
<!--         legend.position='bottom', -->
<!--         legend.box='horizontal') + -->
<!--   coord_cartesian(xlim=c(-8,10), -->
<!--                   ylim=c(-5,8), -->
<!--                   clip='off') + -->
<!--   labs(x=paste0('PC 1 (',100*signif(pca$sdev[1]**2/sum(pca$sdev**2),3),'%)'), -->
<!--        y=paste0('PC 2 (',100*signif(pca$sdev[2]**2/sum(pca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='',alpha='') -->

<!-- ``` -->

<!-- ```{r FigS4X8, fig.width=3.5 * fig_scale, fig.height=3 * fig_scale} -->
<!-- dpca <- dualPCA(human.X, LB.X) -->
<!-- dpcadat <- cbind(rbind(dpca$x,dpca$y), -->
<!--                  pcadat.all) -->

<!-- ggplot(dpcadat, aes(x=dPC1,y=dPC2)) + -->
<!--   geom_polygon(data = dpcadat %>% dplyr::filter(dataset=='training') %>% -->
<!--                  dplyr::group_by(status) %>% -->
<!--                  dplyr::slice(chull(dPC1, dPC2)), alpha = 0.1, color=NA, aes(fill=status), show.legend=FALSE) + -->
<!--   geom_point(aes(color=diagnosis,alpha=dataset, shape=dataset), size=1) + -->
<!--   geom_line(aes(group=donor, color=diagnosis,alpha=dataset, shape=dataset)) + -->
<!--   geom_text_repel(data=dpcadat %>% dplyr::filter(dataset=='LB'), aes(label=donor, color=diagnosis), size=2, segment.size=.25) + -->
<!--   my_theme() + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,'unknown'='gray'), guide='none') + -->
<!--   scale_color_manual(values=c(diagnosis_colors,'unknown'='gray')) + -->
<!--   scale_alpha_manual(values=c('testing'=.25,'training'=.25,'validation'=.5,'LB'=1)) + -->
<!--   scale_shape_manual(values=c('testing'=15,'training'=16,'validation'=17,'LB'=18)) + -->
<!--   guides(color=guide_legend(override.aes = list(shape=15,size=2),ncol=3), -->
<!--          shape=guide_legend(ncol=1), -->
<!--          alpha=guide_legend(ncol=1)) + -->
<!--   theme(legend.key.height=unit(1.5,'mm'), -->
<!--         legend.key.spacing=unit(0,'mm'), -->
<!--         #aspect.ratio=dpca$sdev[2]**2/dpca$sdev[1]**2, -->
<!--         legend.position='bottom', -->
<!--         legend.box='horizontal') + -->
<!--   coord_cartesian(clip='off') + -->
<!--   labs(x=paste0('dPC 1 (',100*signif(dpca$sdev[1]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        y=paste0('dPC 2 (',100*signif(dpca$sdev[2]**2/sum(dpca$sdev**2),3),'%)'), -->
<!--        color='',shape='',fill='',alpha='') -->

<!-- ``` -->

<!-- ```{r aggregated_FB_boxplots, fig.width=4.5 * fig_scale, fig.height=1.7 * fig_scale} -->
<!-- tmp <- FB.X.aggregated %>% -->
<!--   tibble::rownames_to_column('sample') %>% -->
<!--   dplyr::left_join(FB.meta %>% tibble::rownames_to_column('sample'), by='sample') %>% -->
<!--   dplyr::mutate(name=gsub('t','',gsub('FB_','',paste0(donor,'_',timepoint)))) %>% -->
<!--   dplyr::left_join(class_df,by='name') %>% -->
<!--   gather(col,value,c('diversity','isotypes','breaks','context','structural')) -->

<!-- ggplot(tmp, aes(x=class,y=value,fill=class)) + -->
<!--   geom_boxplot(outlier.shape=NA, lwd=.5) + -->
<!--   geom_jitter(width=.1,height=0,size=.5) + -->
<!--   stat_compare_means(comparisons=list(c('cluster1','cluster3'), -->
<!--                                       c('cluster2','cluster3'), -->
<!--                                       c('cluster1','cluster2')), -->
<!--                      method='wilcox.test',label = "p.signif", hide.ns=TRUE, size=2, vjust=.5) + -->
<!--   facet_wrap(.~col,ncol=length(features.here),scales='free_y') + -->
<!--   my_theme() + -->
<!--   coord_cartesian(clip='off') + -->
<!--   scale_fill_manual(values=c(diagnosis_colors,class_colors)) + -->
<!--   theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1), -->
<!--         legend.position='none') + -->
<!--   labs(x='',y='category score', color='diagnosis') -->
<!-- ``` -->

<!-- check correlation with measured B cell proportion in FACS -->

<!-- ```{r pct_B_cor, fig.width=3 * fig_scale, fig.height=3 * fig_scale} -->
<!-- FB.data.pooled %>% -->
<!--   gather(feature, value, selected.cols) %>% -->
<!--   dplyr::group_by(feature) %>% -->
<!--   dplyr::summarise(R=cor(value, pct_B, use='na.or.complete')) %>% -->
<!--   dplyr::left_join(columns, by=c('feature'='col')) %>% -->
<!--   dplyr::mutate(feature=gsub('_downsampled|_switch','',feature)) %>% -->
<!--   dplyr::arrange(R) %>% -->
<!--   dplyr::mutate(feature=factor(feature, levels=feature)) %>% -->
<!--   ggplot(aes(y=R, x=feature, fill=category)) +  -->
<!--   geom_col() + -->
<!--   my_theme() + -->
<!--   coord_flip() + -->
<!--   scale_fill_manual(values=cat_cols) + -->
<!--   labs(x='',y='correlation with % B cells') + -->
<!--   theme(legend.position=c(.2,.9)) -->


<!-- FB.data.pooled %>% -->
<!--   gather(feature, value, selected.cols) %>% -->
<!--   dplyr::group_by(feature) %>% -->
<!--   dplyr::summarise(R=cor(value, pct_IgA, use='na.or.complete')) %>% -->
<!--   dplyr::left_join(columns, by=c('feature'='col')) %>% -->
<!--   dplyr::mutate(feature=gsub('_downsampled|_switch','',feature)) %>% -->
<!--   dplyr::arrange(R) %>% -->
<!--   dplyr::mutate(feature=factor(feature, levels=feature)) %>% -->
<!--   ggplot(aes(y=R, x=feature, fill=category)) +  -->
<!--   geom_col() + -->
<!--   my_theme() + -->
<!--   coord_flip() + -->
<!--   scale_fill_manual(values=cat_cols) + -->
<!--   labs(x='',y='correlation with % IgA cells') + -->
<!--   theme(legend.position=c(.2,.9)) -->

<!-- FB.data.pooled %>% -->
<!--   gather(feature, value, selected.cols) %>% -->
<!--   dplyr::group_by(feature) %>% -->
<!--   dplyr::summarise(R=cor(value, pct_IgG, use='na.or.complete')) %>% -->
<!--   dplyr::left_join(columns, by=c('feature'='col')) %>% -->
<!--   dplyr::mutate(feature=gsub('_downsampled|_switch','',feature)) %>% -->
<!--   dplyr::arrange(R) %>% -->
<!--   dplyr::mutate(feature=factor(feature, levels=feature)) %>% -->
<!--   ggplot(aes(y=R, x=feature, fill=category)) +  -->
<!--   geom_col() + -->
<!--   my_theme() + -->
<!--   coord_flip() + -->
<!--   scale_fill_manual(values=cat_cols) + -->
<!--   labs(x='',y='correlation with % IgG cells') + -->
<!--   theme(legend.position=c(.2,.9)) -->


<!-- ``` -->

# session info

```{r sessioninfo}
sessionInfo()
```
